// @backstage/frontend-plugin-api/dist/wiring/createFrontendFeatureLoader.esm.js
import { describeParentCallSite } from '../routing/describeParentCallSite.esm.js';

function createFrontendFeatureLoader(options) {
  const description = `created at '${describeParentCallSite()}'`;
  return {
    $$type: "@backstage/FrontendFeatureLoader",
    version: "v1",
    description,
    toString() {
      return `FeatureLoader{description=${description}}`;
    },
    async loader(deps) {
      const it = await options.loader(deps);
      const result = new Array();
      for await (const item of it) {
        if (isFeatureOrLoader(item)) {
          result.push(item);
        } else if ("default" in item) {
          result.push(item.default);
        } else {
          throw new Error(`Invalid item "${item}"`);
        }
      }
      return result;
    }
  };
}
function isFeatureOrLoader(obj) {
  if (obj !== null && typeof obj === "object" && "$$type" in obj) {
    return obj.$$type === "@backstage/FrontendPlugin" || obj.$$type === "@backstage/FrontendModule" || obj.$$type === "@backstage/FrontendFeatureLoader";
  }
  return false;
}

export { createFrontendFeatureLoader };
