// @backstage/plugin-scaffolder-backend/dist/lib/templating/filters/parseEntityRef/filter.cjs.js
'use strict';

var catalogModel = require('@backstage/catalog-model');
var alpha = require('@backstage/plugin-scaffolder-node/alpha');
var examples = require('./examples.cjs.js');

const parseEntityRef = alpha.createTemplateFilter({
  id: "parseEntityRef",
  description: "Extracts the parts of an entity reference, such as the kind, namespace, and name.",
  schema: (z) => z.function().args(
    z.union([
      z.string().describe("compact entity reference"),
      z.object({
        kind: z.string().optional(),
        namespace: z.string().optional(),
        name: z.string()
      }).describe("`CompoundEntityRef`")
    ]),
    z.object({
      defaultKind: z.string().describe("The default kind, if none is given in the reference"),
      defaultNamespace: z.string().describe(
        "The default namespace, if none is given in the reference"
      )
    }).partial().optional()
  ).returns(
    z.object({
      kind: z.string(),
      namespace: z.string(),
      name: z.string()
    }).describe("`CompoundEntityRef`")
  ),
  examples: examples.examples,
  filter: catalogModel.parseEntityRef
});

exports.parseEntityRef = parseEntityRef;
