// @backstage/plugin-catalog/dist/alpha/components/EntityLayout/EntityLayout.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import Alert from '@material-ui/lab/Alert';
import { useRouteRefParams, useElementFilter, attachComponentData } from '@backstage/core-plugin-api';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { Page, Progress, Content, WarningPanel, Link } from '@backstage/core-components';
import { entityRouteRef, useAsyncEntity } from '@backstage/plugin-catalog-react';
import { catalogTranslationRef } from '../../translation.esm.js';
import { EntityHeader } from '../EntityHeader/EntityHeader.esm.js';
import { EntityTabs } from '../EntityTabs/EntityTabs.esm.js';

const dataKey = "plugin.catalog.entityLayoutRoute";
const Route = () => null;
attachComponentData(Route, dataKey, true);
attachComponentData(Route, "core.gatherMountPoints", true);
const EntityLayout = (props) => {
  const {
    UNSTABLE_extraContextMenuItems,
    UNSTABLE_contextMenuOptions,
    contextMenuItems,
    children,
    NotFoundComponent,
    parentEntityRelations
  } = props;
  const { kind } = useRouteRefParams(entityRouteRef);
  const { entity, loading, error } = useAsyncEntity();
  const header = props.header ?? /* @__PURE__ */ jsx(
    EntityHeader,
    {
      parentEntityRelations,
      UNSTABLE_contextMenuOptions,
      UNSTABLE_extraContextMenuItems,
      contextMenuItems
    }
  );
  const routes = useElementFilter(
    children,
    (elements) => elements.selectByComponentData({
      key: dataKey,
      withStrictError: "Child of EntityLayout must be an EntityLayout.Route"
    }).getElements().flatMap(({ props: elementProps }) => {
      if (!entity) {
        return [];
      }
      if (elementProps.if && !elementProps.if(entity)) {
        return [];
      }
      return [
        {
          path: elementProps.path,
          title: elementProps.title,
          group: elementProps.group,
          children: elementProps.children
        }
      ];
    }),
    [entity]
  );
  const { t } = useTranslationRef(catalogTranslationRef);
  return /* @__PURE__ */ jsxs(Page, { themeId: entity?.spec?.type?.toString() ?? "home", children: [
    !loading && header,
    loading && /* @__PURE__ */ jsx(Progress, {}),
    entity && /* @__PURE__ */ jsx(EntityTabs, { routes }),
    error && /* @__PURE__ */ jsx(Content, { children: /* @__PURE__ */ jsx(Alert, { severity: "error", children: error.toString() }) }),
    !loading && !error && !entity && /* @__PURE__ */ jsx(Content, { children: NotFoundComponent ? NotFoundComponent : /* @__PURE__ */ jsxs(WarningPanel, { title: t("entityLabels.warningPanelTitle"), children: [
      "There is no ",
      kind,
      " with the requested",
      " ",
      /* @__PURE__ */ jsx(Link, { to: "https://backstage.io/docs/features/software-catalog/references", children: "kind, namespace, and name" }),
      "."
    ] }) })
  ] });
};
EntityLayout.Route = Route;

export { EntityLayout };
