// @backstage/core-components/dist/components/CodeSnippet/CodeSnippet.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import Box from '@material-ui/core/Box';
import { useTheme } from '@material-ui/core/styles';
import LightAsync from 'react-syntax-highlighter/dist/esm/light-async';
import dark from 'react-syntax-highlighter/dist/esm/styles/hljs/dark';
import docco from 'react-syntax-highlighter/dist/esm/styles/hljs/docco';
import { CopyTextButton } from '../CopyTextButton/CopyTextButton.esm.js';

function CodeSnippet(props) {
  const {
    text,
    language,
    showLineNumbers = false,
    highlightedNumbers,
    customStyle,
    showCopyCodeButton = false
  } = props;
  const theme = useTheme();
  const mode = theme.palette.type === "dark" ? dark : docco;
  const highlightColor = theme.palette.type === "dark" ? "#256bf3" : "#e6ffed";
  return /* @__PURE__ */ jsxs(Box, { position: "relative", children: [
    /* @__PURE__ */ jsx(
      LightAsync,
      {
        customStyle,
        language,
        style: mode,
        showLineNumbers,
        wrapLines: true,
        lineNumberStyle: { color: theme.palette.textVerySubtle },
        lineProps: (lineNumber) => highlightedNumbers?.includes(lineNumber) ? {
          style: {
            backgroundColor: highlightColor
          }
        } : {},
        children: text
      }
    ),
    showCopyCodeButton && /* @__PURE__ */ jsx(Box, { position: "absolute", top: 0, right: 0, children: /* @__PURE__ */ jsx(CopyTextButton, { text }) })
  ] });
}

export { CodeSnippet };
