// @backstage/plugin-search-react/dist/components/SearchFilter/SearchFilter.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { useRef } from 'react';
import { capitalize } from 'lodash';
import { v4 } from 'uuid';
import FormControl from '@material-ui/core/FormControl';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import Checkbox from '@material-ui/core/Checkbox';
import FormLabel from '@material-ui/core/FormLabel';
import { makeStyles } from '@material-ui/core/styles';
import { Select } from '@backstage/core-components';
import { useSearch } from '../../context/SearchContext.esm.js';
import { AutocompleteFilter } from './SearchFilter.Autocomplete.esm.js';
import { useDefaultFilterValue, useAsyncFilterValues } from './hooks.esm.js';
import { ensureFilterValueWithLabel } from './types.esm.js';
import { useTranslationRef } from '@backstage/frontend-plugin-api';
import { searchReactTranslationRef } from '../../translation.esm.js';

const useStyles = makeStyles({
  label: {
    textTransform: "capitalize"
  },
  checkboxWrapper: {
    display: "flex",
    alignItems: "center",
    width: "100%"
  },
  textWrapper: {
    overflow: "hidden",
    textOverflow: "ellipsis",
    whiteSpace: "nowrap"
  }
});
const CheckboxFilter = (props) => {
  const {
    className,
    defaultValue,
    label: formLabel,
    name,
    values: givenValues = [],
    valuesDebounceMs
  } = props;
  const classes = useStyles();
  const { filters, setFilters } = useSearch();
  useDefaultFilterValue(name, defaultValue);
  const asyncValues = typeof givenValues === "function" ? givenValues : void 0;
  const defaultValues = typeof givenValues === "function" ? void 0 : givenValues.map((v) => ensureFilterValueWithLabel(v));
  const { value: values = [], loading } = useAsyncFilterValues(
    asyncValues,
    "",
    defaultValues,
    valuesDebounceMs
  );
  const handleChange = (e) => {
    const {
      target: { value, checked }
    } = e;
    setFilters((prevFilters) => {
      const { [name]: filter, ...others } = prevFilters;
      const rest = (filter || []).filter((i) => i !== value);
      const items = checked ? [...rest, value] : rest;
      return items.length ? { ...others, [name]: items } : others;
    });
  };
  return /* @__PURE__ */ jsxs(
    FormControl,
    {
      className,
      disabled: loading,
      fullWidth: true,
      "data-testid": "search-checkboxfilter-next",
      children: [
        !!formLabel && /* @__PURE__ */ jsx(FormLabel, { className: classes.label, children: formLabel }),
        values.map(({ value, label }) => /* @__PURE__ */ jsx(
          FormControlLabel,
          {
            classes: {
              root: classes.checkboxWrapper,
              label: classes.textWrapper
            },
            label,
            control: /* @__PURE__ */ jsx(
              Checkbox,
              {
                color: "primary",
                inputProps: { "aria-labelledby": label },
                value,
                name: label,
                onChange: handleChange,
                checked: (filters[name] ?? []).includes(value)
              }
            )
          },
          value
        ))
      ]
    }
  );
};
const SelectFilter = (props) => {
  const {
    className,
    defaultValue,
    label,
    name,
    values: givenValues,
    valuesDebounceMs
  } = props;
  const { t } = useTranslationRef(searchReactTranslationRef);
  useDefaultFilterValue(name, defaultValue);
  const asyncValues = typeof givenValues === "function" ? givenValues : void 0;
  const defaultValues = typeof givenValues === "function" ? void 0 : givenValues?.map((v) => ensureFilterValueWithLabel(v));
  const { value: values = [], loading } = useAsyncFilterValues(
    asyncValues,
    "",
    defaultValues,
    valuesDebounceMs
  );
  const allOptionValue = useRef(v4());
  const allOption = {
    value: allOptionValue.current,
    label: t("searchFilter.allOptionTitle")
  };
  const { filters, setFilters } = useSearch();
  const handleChange = (value) => {
    setFilters((prevFilters) => {
      const { [name]: filter, ...others } = prevFilters;
      return value !== allOptionValue.current ? { ...others, [name]: value } : others;
    });
  };
  const items = [allOption, ...values];
  return /* @__PURE__ */ jsx(
    FormControl,
    {
      disabled: loading,
      className,
      variant: "filled",
      fullWidth: true,
      "data-testid": "search-selectfilter-next",
      children: /* @__PURE__ */ jsx(
        Select,
        {
          label: label ?? capitalize(name),
          selected: filters[name] || allOptionValue.current,
          onChange: handleChange,
          items
        }
      )
    }
  );
};
const SearchFilter = (props) => {
  const { component: Element, ...elementProps } = props;
  return /* @__PURE__ */ jsx(Element, { ...elementProps });
};
SearchFilter.Checkbox = (props) => /* @__PURE__ */ jsx(SearchFilter, { ...props, component: CheckboxFilter });
SearchFilter.Select = (props) => /* @__PURE__ */ jsx(SearchFilter, { ...props, component: SelectFilter });
SearchFilter.Autocomplete = (props) => /* @__PURE__ */ jsx(SearchFilter, { ...props, component: AutocompleteFilter });

export { CheckboxFilter, SearchFilter, SelectFilter };
