// @backstage/plugin-search-react/dist/components/SearchResult/SearchResult.esm.js
import { jsx } from 'react/jsx-runtime';
import useAsync from 'react-use/esm/useAsync';
import { isFunction } from 'lodash';
import { EmptyState, Progress, ResponseErrorPanel } from '@backstage/core-components';
import { AnalyticsContext, useApi } from '@backstage/core-plugin-api';
import { searchApiRef } from '../../api.esm.js';
import { useSearch } from '../../context/SearchContext.esm.js';
import { SearchResultListItemExtensions } from '../../extensions.esm.js';
import { useTranslationRef } from '@backstage/frontend-plugin-api';
import { searchReactTranslationRef } from '../../translation.esm.js';

const SearchResultContext = (props) => {
  const { children } = props;
  const context = useSearch();
  const { result: state, ...query } = context;
  return children(state, query);
};
const SearchResultApi = (props) => {
  const { query, children } = props;
  const searchApi = useApi(searchApiRef);
  const state = useAsync(() => {
    const { term = "", types = [], filters = {}, ...rest } = query;
    return searchApi.query({ ...rest, term, types, filters });
  }, [query]);
  return children(state, query);
};
const SearchResultState = (props) => {
  const { query, children } = props;
  return query ? /* @__PURE__ */ jsx(SearchResultApi, { query, children }) : /* @__PURE__ */ jsx(SearchResultContext, { children });
};
const SearchResultComponent = (props) => {
  const { t } = useTranslationRef(searchReactTranslationRef);
  const {
    query,
    children,
    noResultsComponent = /* @__PURE__ */ jsx(EmptyState, { missing: "data", title: t("noResultsDescription") }),
    ...rest
  } = props;
  return /* @__PURE__ */ jsx(SearchResultState, { query, children: ({ loading, error, value }) => {
    if (loading) {
      return /* @__PURE__ */ jsx(Progress, {});
    }
    if (error) {
      return /* @__PURE__ */ jsx(
        ResponseErrorPanel,
        {
          title: "Error encountered while fetching search results",
          error
        }
      );
    }
    if (!value?.results.length) {
      return noResultsComponent;
    }
    if (isFunction(children)) {
      return children(value);
    }
    return /* @__PURE__ */ jsx(SearchResultListItemExtensions, { ...rest, results: value.results, children });
  } });
};
const SearchResult = (props) => /* @__PURE__ */ jsx(
  AnalyticsContext,
  {
    attributes: {
      pluginId: "search",
      extension: "SearchResult"
    },
    children: /* @__PURE__ */ jsx(SearchResultComponent, { ...props })
  }
);

export { SearchResult, SearchResultApi, SearchResultComponent, SearchResultContext, SearchResultState };
