// @backstage/plugin-bitbucket-cloud-common/dist/models/index.esm.js
var Models;
((Models2) => {
  Models2.BaseCommitSummaryMarkupEnum = {
    Markdown: "markdown",
    Creole: "creole",
    Plaintext: "plaintext"
  };
  Models2.BranchMergeStrategiesEnum = {
    MergeCommit: "merge_commit",
    Squash: "squash",
    FastForward: "fast_forward",
    SquashFastForward: "squash_fast_forward",
    RebaseFastForward: "rebase_fast_forward",
    RebaseMerge: "rebase_merge"
  };
  Models2.CommitFileAttributesEnum = {
    Link: "link",
    Executable: "executable",
    Subrepository: "subrepository",
    Binary: "binary",
    Lfs: "lfs"
  };
  Models2.ParticipantRoleEnum = {
    Participant: "PARTICIPANT",
    Reviewer: "REVIEWER"
  };
  Models2.ParticipantStateEnum = {
    Approved: "approved",
    ChangesRequested: "changes_requested",
    Null: "null"
  };
  Models2.RepositoryForkPolicyEnum = {
    AllowForks: "allow_forks",
    NoPublicForks: "no_public_forks",
    NoForks: "no_forks"
  };
  Models2.RepositoryScmEnum = {
    Git: "git"
  };
  Models2.WorkspaceForkingModeEnum = {
    AllowForks: "allow_forks",
    InternalOnly: "internal_only"
  };
})(Models || (Models = {}));

export { Models };
