// @backstage/core-components/dist/components/Select/Select.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import Box from '@material-ui/core/Box';
import Checkbox from '@material-ui/core/Checkbox';
import Chip from '@material-ui/core/Chip';
import FormControl from '@material-ui/core/FormControl';
import InputBase from '@material-ui/core/InputBase';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import Select from '@material-ui/core/Select';
import { withStyles, createStyles, makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import CancelIcon from '@material-ui/icons/Cancel';
import { useState, useEffect } from 'react';
import ClosedDropdown from './static/ClosedDropdown.esm.js';
import OpenedDropdown from './static/OpenedDropdown.esm.js';

const BootstrapInput = withStyles(
  (theme) => createStyles({
    root: {
      "label + &": {
        marginTop: theme.spacing(3)
      },
      "&.Mui-focused > div[role=button]": {
        borderColor: theme.palette.primary.main
      }
    },
    input: {
      borderRadius: theme.shape.borderRadius,
      position: "relative",
      backgroundColor: theme.palette.background.paper,
      border: "1px solid #ced4da",
      fontSize: theme.typography.body1.fontSize,
      padding: theme.spacing(1.25, 3.25, 1.25, 1.5),
      transition: theme.transitions.create(["border-color", "box-shadow"]),
      "&:focus": {
        background: theme.palette.background.paper,
        borderRadius: theme.shape.borderRadius
      }
    }
  }),
  { name: "BackstageSelectInputBase" }
)(InputBase);
const useStyles = makeStyles(
  (theme) => createStyles({
    formControl: {
      margin: theme.spacing(1, 0)
    },
    label: {
      transform: "initial",
      fontWeight: "bold",
      fontSize: theme.typography.body2.fontSize,
      fontFamily: theme.typography.fontFamily,
      color: theme.palette.text.primary,
      "&.Mui-focused": {
        color: theme.palette.text.primary
      }
    },
    formLabel: {
      transform: "initial",
      fontWeight: "bold",
      fontSize: theme.typography.body2.fontSize,
      fontFamily: theme.typography.fontFamily,
      color: theme.palette.text.primary,
      "&.Mui-focused": {
        color: theme.palette.text.primary
      }
    },
    chips: {
      display: "flex",
      flexWrap: "wrap"
    },
    chip: {
      margin: 2
    },
    checkbox: {},
    root: {
      display: "flex",
      flexDirection: "column"
    }
  }),
  { name: "BackstageSelect" }
);
function SelectComponent(props) {
  const {
    multiple,
    items,
    label,
    placeholder,
    selected,
    onChange,
    triggerReset,
    native = false,
    disabled = false,
    margin,
    "data-testid": dataTestId = "select"
  } = props;
  const classes = useStyles();
  const [value, setValue] = useState(
    selected || (multiple ? [] : "")
  );
  const [isOpen, setOpen] = useState(false);
  useEffect(() => {
    setValue(multiple ? [] : "");
  }, [triggerReset, multiple]);
  useEffect(() => {
    setValue(selected || (multiple ? [] : ""));
  }, [selected, multiple]);
  const handleChange = (event) => {
    setValue(event.target.value);
    onChange(event.target.value);
  };
  const handleOpen = (event) => {
    if (disabled) {
      event.preventDefault();
      return;
    }
    setOpen((previous) => {
      if (multiple && !(event.target instanceof HTMLElement)) {
        return true;
      }
      return !previous;
    });
  };
  const handleClose = () => {
    setOpen(false);
  };
  const handleDelete = (selectedValue) => () => {
    const newValue = value.filter((chip) => chip !== selectedValue);
    setValue(newValue);
    onChange(newValue);
  };
  return /* @__PURE__ */ jsx(Box, { className: classes.root, children: /* @__PURE__ */ jsxs(FormControl, { className: classes.formControl, children: [
    /* @__PURE__ */ jsx(InputLabel, { className: classes.formLabel, children: label }),
    /* @__PURE__ */ jsxs(
      Select,
      {
        "aria-label": label,
        value,
        native,
        disabled,
        "data-testid": dataTestId,
        displayEmpty: true,
        multiple,
        margin,
        onChange: handleChange,
        open: isOpen,
        onOpen: handleOpen,
        onClose: handleClose,
        input: /* @__PURE__ */ jsx(BootstrapInput, {}),
        label,
        renderValue: (s) => multiple && value.length !== 0 ? /* @__PURE__ */ jsx(Box, { className: classes.chips, children: s.map((selectedValue) => {
          const item = items.find((el) => el.value === selectedValue);
          return item ? /* @__PURE__ */ jsx(
            Chip,
            {
              "data-testid": "chip",
              label: item?.label,
              clickable: true,
              deleteIcon: /* @__PURE__ */ jsx(
                CancelIcon,
                {
                  "data-testid": "cancel-icon",
                  onMouseDown: (event) => event.stopPropagation()
                }
              ),
              onDelete: handleDelete(selectedValue),
              className: classes.chip
            },
            item?.value
          ) : false;
        }) }) : /* @__PURE__ */ jsx(Typography, { children: value.length === 0 ? placeholder || "" : items.find((el) => el.value === s)?.label }),
        IconComponent: () => !isOpen ? /* @__PURE__ */ jsx(ClosedDropdown, {}) : /* @__PURE__ */ jsx(OpenedDropdown, {}),
        MenuProps: {
          anchorOrigin: {
            vertical: "bottom",
            horizontal: "left"
          },
          transformOrigin: {
            vertical: "top",
            horizontal: "left"
          },
          getContentAnchorEl: null
        },
        children: [
          !!placeholder && !multiple && /* @__PURE__ */ jsx(MenuItem, { value: [], children: placeholder }),
          native ? items && items.map((item) => /* @__PURE__ */ jsx("option", { value: item.value, children: item.label }, item.value)) : items && items.map((item) => /* @__PURE__ */ jsxs(MenuItem, { value: item.value, children: [
            multiple && /* @__PURE__ */ jsx(
              Checkbox,
              {
                color: "primary",
                checked: value.includes(item.value) || false,
                className: classes.checkbox
              }
            ),
            item.label
          ] }, item.value))
        ]
      }
    )
  ] }) });
}

export { SelectComponent };
