// @backstage/plugin-techdocs-module-addons-contrib/dist/alpha.esm.js
import { TechDocsAddonLocations } from '@backstage/plugin-techdocs-react';
import { AddonBlueprint } from '@backstage/plugin-techdocs-react/alpha';
import { TextSizeAddon } from './TextSize/TextSize.esm.js';
import { ReportIssueAddon } from './ReportIssue/ReportIssue.esm.js';
import { ExpandableNavigationAddon } from './ExpandableNavigation/ExpandableNavigation.esm.js';
import { LightBoxAddon } from './LightBox/LightBox.esm.js';
import { createFrontendModule } from '@backstage/frontend-plugin-api';

const techDocsExpandableNavigationAddon = AddonBlueprint.make({
  name: "expandable-navigation",
  params: {
    name: "ExpandableNavigation",
    location: TechDocsAddonLocations.PrimarySidebar,
    component: ExpandableNavigationAddon
  }
});
const techDocsExpandableNavigationAddonModule = createFrontendModule({
  pluginId: "techdocs",
  extensions: [techDocsExpandableNavigationAddon]
});
const techDocsReportIssueAddon = AddonBlueprint.make({
  name: "report-issue",
  params: {
    name: "ReportIssue",
    location: TechDocsAddonLocations.Content,
    component: ReportIssueAddon
  }
});
const techDocsReportIssueAddonModule = createFrontendModule({
  pluginId: "techdocs",
  extensions: [techDocsReportIssueAddon]
});
const techDocsTextSizeAddon = AddonBlueprint.make({
  name: "text-size",
  params: {
    name: "TextSize",
    location: TechDocsAddonLocations.Settings,
    component: TextSizeAddon
  }
});
const techDocsTextSizeAddonModule = createFrontendModule({
  pluginId: "techdocs",
  extensions: [techDocsTextSizeAddon]
});
const techDocsLightBoxAddon = AddonBlueprint.make({
  name: "light-box",
  params: {
    name: "LightBox",
    location: TechDocsAddonLocations.Content,
    component: LightBoxAddon
  }
});
const techDocsLightBoxAddonModule = createFrontendModule({
  pluginId: "techdocs",
  extensions: [techDocsLightBoxAddon]
});

export { techDocsExpandableNavigationAddonModule, techDocsLightBoxAddonModule, techDocsReportIssueAddonModule, techDocsTextSizeAddonModule };
