// @backstage/core-components/dist/layout/ContentHeader/ContentHeader.esm.js
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import Box from '@material-ui/core/Box';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import { Helmet } from 'react-helmet';

const useStyles = (props) => makeStyles(
  (theme) => ({
    container: {
      width: "100%",
      display: "flex",
      flexDirection: "row",
      flexWrap: "wrap",
      justifyContent: "flex-end",
      alignItems: "center",
      marginBottom: theme.spacing(2),
      textAlign: props.textAlign
    },
    leftItemsBox: {
      flex: "1 1 auto",
      minWidth: 0,
      overflow: "visible"
    },
    rightItemsBox: {
      flex: "0 1 auto",
      display: "flex",
      flexDirection: "row",
      flexWrap: "wrap",
      alignItems: "center",
      marginLeft: theme.spacing(1),
      minWidth: 0,
      overflow: "visible"
    },
    description: {},
    title: {
      display: "inline-flex",
      marginBottom: 0
    }
  }),
  { name: "BackstageContentHeader" }
);
const ContentHeaderTitle = ({ title, className }) => /* @__PURE__ */ jsx(
  Typography,
  {
    variant: "h4",
    component: "h2",
    className,
    "data-testid": "header-title",
    children: title
  }
);
const ContentHeaderDescription = ({
  description,
  className
}) => description ? /* @__PURE__ */ jsx(
  Typography,
  {
    variant: "body2",
    className,
    "data-testid": "header-description",
    children: description
  }
) : null;
function ContentHeader(props) {
  const {
    description,
    title,
    titleComponent: TitleComponent = void 0,
    children,
    descriptionComponent: DescriptionComponent = void 0,
    textAlign = "left"
  } = props;
  const classes = useStyles({ textAlign })();
  const renderedTitle = TitleComponent ? TitleComponent : /* @__PURE__ */ jsx(ContentHeaderTitle, { title, className: classes.title });
  const renderedDescription = DescriptionComponent ? DescriptionComponent : /* @__PURE__ */ jsx(
    ContentHeaderDescription,
    {
      description,
      className: classes.description
    }
  );
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(Helmet, { title }),
    /* @__PURE__ */ jsxs(Box, { className: classes.container, children: [
      /* @__PURE__ */ jsxs(Box, { className: classes.leftItemsBox, children: [
        renderedTitle,
        renderedDescription
      ] }),
      /* @__PURE__ */ jsx(Box, { className: classes.rightItemsBox, children })
    ] })
  ] });
}

export { ContentHeader };
