// @backstage/core-components/dist/components/EmptyState/EmptyState.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import Grid from '@material-ui/core/Grid';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import { EmptyStateImage } from './EmptyStateImage.esm.js';

const useStyles = makeStyles(
  (theme) => ({
    root: {
      backgroundColor: theme.palette.background.default,
      padding: theme.spacing(2, 0, 0, 0)
    },
    action: {
      marginTop: theme.spacing(2)
    },
    imageContainer: {
      position: "relative"
    }
  }),
  { name: "BackstageEmptyState" }
);
function EmptyState(props) {
  const { title, description, missing, action } = props;
  const classes = useStyles();
  return /* @__PURE__ */ jsxs(
    Grid,
    {
      container: true,
      direction: "row",
      justifyContent: "space-around",
      alignItems: "flex-start",
      className: classes.root,
      spacing: 2,
      children: [
        /* @__PURE__ */ jsx(Grid, { item: true, xs: 12, md: 6, children: /* @__PURE__ */ jsxs(Grid, { container: true, direction: "column", children: [
          /* @__PURE__ */ jsx(Grid, { item: true, xs: true, children: /* @__PURE__ */ jsx(Typography, { variant: "h5", children: title }) }),
          /* @__PURE__ */ jsx(Grid, { item: true, xs: true, children: /* @__PURE__ */ jsx(Typography, { variant: "body1", children: description }) }),
          /* @__PURE__ */ jsx(Grid, { item: true, xs: true, className: classes.action, children: action })
        ] }) }),
        /* @__PURE__ */ jsx(Grid, { item: true, xs: 12, md: 6, className: classes.imageContainer, children: typeof missing === "string" ? /* @__PURE__ */ jsx(EmptyStateImage, { missing }) : missing.customImage })
      ]
    }
  );
}

export { EmptyState };
