// @backstage/plugin-scaffolder/dist/alpha/components/TemplateEditorPage/CustomFieldExplorer.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { StreamLanguage } from '@codemirror/language';
import { yaml as yaml$1 } from '@codemirror/legacy-modes/mode/yaml';
import Button from '@material-ui/core/Button';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import CardHeader from '@material-ui/core/CardHeader';
import { makeStyles } from '@material-ui/core/styles';
import CodeMirror from '@uiw/react-codemirror';
import { useState, useMemo, useCallback } from 'react';
import yaml from 'yaml';
import { Form } from '@backstage/plugin-scaffolder-react/alpha';
import { TemplateEditorForm } from './TemplateEditorForm.esm.js';
import validator from '@rjsf/validator-ajv8';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { scaffolderTranslationRef } from '../../../translation.esm.js';
import InputAdornment from '@material-ui/core/InputAdornment';
import TextField from '@material-ui/core/TextField';
import SearchIcon from '@material-ui/icons/Search';
import Autocomplete from '@material-ui/lab/Autocomplete';

const useStyles = makeStyles(
  (theme) => ({
    root: {
      gridArea: "pageContent",
      display: "grid",
      gridGap: theme.spacing(2),
      gridTemplateAreas: `
      "controls"
      "fieldForm"
      "preview"
    `,
      [theme.breakpoints.up("md")]: {
        gridTemplateAreas: `
      "controls controls"
      "fieldForm preview"
    `,
        gridTemplateRows: "auto 1fr",
        gridTemplateColumns: "1fr 1fr"
      }
    },
    controls: {
      gridArea: "controls",
      display: "flex",
      flexFlow: "row nowrap",
      alignItems: "center"
    },
    fieldForm: {
      gridArea: "fieldForm"
    },
    preview: {
      gridArea: "preview",
      display: "grid",
      gridGap: theme.spacing(2),
      alignContent: "start"
    }
  }),
  { name: "ScaffolderCustomFieldExplorer" }
);
const CustomFieldExplorer = ({
  customFieldExtensions = []
}) => {
  const classes = useStyles();
  const { t } = useTranslationRef(scaffolderTranslationRef);
  const fieldOptions = customFieldExtensions.filter((field) => !!field.schema);
  const [selectedField, setSelectedField] = useState(fieldOptions?.[0]);
  const [fieldFormState, setFieldFormState] = useState({});
  const [refreshKey, setRefreshKey] = useState(Date.now());
  const sampleFieldTemplate = useMemo(() => {
    if (!selectedField) {
      return "";
    }
    return yaml.stringify({
      parameters: [
        {
          title: `${selectedField.name} Example`,
          properties: {
            [selectedField.name]: {
              type: selectedField.schema?.returnValue?.type,
              "ui:field": selectedField.name,
              "ui:options": fieldFormState
            }
          }
        }
      ]
    });
  }, [fieldFormState, selectedField]);
  const fieldComponents = useMemo(() => {
    return Object.fromEntries(
      customFieldExtensions.map(({ name, component }) => [name, component])
    );
  }, [customFieldExtensions]);
  const handleSelectionChange = useCallback(
    (selection) => {
      setSelectedField(selection);
      setFieldFormState({});
    },
    [setFieldFormState, setSelectedField]
  );
  const handleFieldConfigChange = useCallback(
    (state) => {
      setFieldFormState(state);
      setRefreshKey(Date.now());
    },
    [setFieldFormState, setRefreshKey]
  );
  return /* @__PURE__ */ jsxs("main", { className: classes.root, children: [
    /* @__PURE__ */ jsx("div", { className: classes.controls, children: /* @__PURE__ */ jsx(
      Autocomplete,
      {
        id: "custom-fields-autocomplete",
        value: selectedField,
        options: fieldOptions,
        getOptionLabel: (option) => option.name,
        renderInput: (params) => /* @__PURE__ */ jsx(
          TextField,
          {
            ...params,
            "aria-label": t(
              "templateEditorPage.customFieldExplorer.selectFieldLabel"
            ),
            placeholder: t(
              "templateEditorPage.customFieldExplorer.selectFieldLabel"
            ),
            variant: "outlined",
            InputProps: {
              ...params.InputProps,
              startAdornment: /* @__PURE__ */ jsx(InputAdornment, { position: "start", children: /* @__PURE__ */ jsx(SearchIcon, {}) })
            }
          }
        ),
        onChange: (_event, option) => {
          if (option) {
            handleSelectionChange(option);
          }
        },
        disableClearable: true,
        fullWidth: true
      }
    ) }),
    /* @__PURE__ */ jsx("div", { className: classes.fieldForm, children: /* @__PURE__ */ jsxs(Card, { children: [
      /* @__PURE__ */ jsx(
        CardHeader,
        {
          title: t("templateEditorPage.customFieldExplorer.fieldForm.title")
        }
      ),
      /* @__PURE__ */ jsx(CardContent, { children: /* @__PURE__ */ jsx(
        Form,
        {
          showErrorList: false,
          fields: { ...fieldComponents },
          noHtml5Validate: true,
          formData: fieldFormState,
          formContext: { fieldFormState },
          onSubmit: (e) => handleFieldConfigChange(e.formData),
          validator,
          schema: selectedField?.schema?.uiOptions || {},
          experimental_defaultFormStateBehavior: {
            allOf: "populateDefaults"
          },
          children: /* @__PURE__ */ jsx(
            Button,
            {
              variant: "contained",
              color: "primary",
              type: "submit",
              disabled: !selectedField?.schema?.uiOptions,
              children: t(
                "templateEditorPage.customFieldExplorer.fieldForm.applyButtonTitle"
              )
            }
          )
        }
      ) })
    ] }) }),
    /* @__PURE__ */ jsxs("div", { className: classes.preview, children: [
      /* @__PURE__ */ jsxs(Card, { children: [
        /* @__PURE__ */ jsx(
          CardHeader,
          {
            title: t("templateEditorPage.customFieldExplorer.preview.title")
          }
        ),
        /* @__PURE__ */ jsx(CardContent, { children: /* @__PURE__ */ jsx(
          CodeMirror,
          {
            readOnly: true,
            theme: "dark",
            height: "100%",
            extensions: [StreamLanguage.define(yaml$1)],
            value: sampleFieldTemplate
          }
        ) })
      ] }),
      /* @__PURE__ */ jsx(
        TemplateEditorForm,
        {
          content: sampleFieldTemplate,
          contentIsSpec: true,
          fieldExtensions: customFieldExtensions,
          setErrorText: () => null
        },
        refreshKey
      )
    ] })
  ] });
};

export { CustomFieldExplorer };
