// @backstage/plugin-scaffolder/dist/alpha/components/TemplateEditorPage/TemplateEditorPage.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { makeStyles } from '@material-ui/core/styles';
import { Page, Header, Content } from '@backstage/core-components';
import { useRouteRef } from '@backstage/core-plugin-api';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { scaffolderTranslationRef } from '../../../translation.esm.js';
import { editRouteRef } from '../../../routes.esm.js';
import { TemplateEditor } from './TemplateEditor.esm.js';

const useStyles = makeStyles(
  {
    content: {
      padding: 0
    }
  },
  { name: "ScaffolderTemplateEditorToolbar" }
);
function TemplateEditorPage(props) {
  const classes = useStyles();
  const editLink = useRouteRef(editRouteRef);
  const { t } = useTranslationRef(scaffolderTranslationRef);
  return /* @__PURE__ */ jsxs(Page, { themeId: "home", children: [
    /* @__PURE__ */ jsx(
      Header,
      {
        title: t("templateEditorPage.title"),
        subtitle: t("templateEditorPage.subtitle"),
        type: t("templateIntroPage.title"),
        typeLink: editLink()
      }
    ),
    /* @__PURE__ */ jsx(Content, { className: classes.content, children: /* @__PURE__ */ jsx(
      TemplateEditor,
      {
        layouts: props.layouts,
        formProps: props.formProps,
        fieldExtensions: props.fieldExtensions
      }
    ) })
  ] });
}

export { TemplateEditorPage };
