// @backstage/plugin-scaffolder/dist/alpha/components/TemplateEditorPage/TemplateEditorLayout.esm.js
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { withStyles, useTheme } from '@material-ui/core/styles';
import { PanelGroup, Panel, PanelResizeHandle } from 'react-resizable-panels';
import useMediaQuery from '@material-ui/core/useMediaQuery';

const TemplateEditorLayout = withStyles(
  (theme) => ({
    root: {
      height: "100%",
      gridArea: "pageContent",
      display: "grid",
      gridTemplateAreas: `
      "toolbar"
      "browser"
      "editor"
      "preview"
      "results"
    `,
      [theme.breakpoints.up("md")]: {
        gridTemplateAreas: `
      "toolbar toolbar toolbar"
      "browser editor preview"
      "results results results"
    `,
        gridTemplateColumns: "1fr 5fr",
        gridTemplateRows: "auto 1fr auto"
      }
    }
  }),
  { name: "ScaffolderTemplateEditorLayout" }
)(({ children, classes }) => /* @__PURE__ */ jsx("main", { className: classes.root, children }));
const TemplateEditorLayoutToolbar = withStyles(
  {
    root: {
      gridArea: "toolbar"
    }
  },
  { name: "ScaffolderTemplateEditorLayoutToolbar" }
)(({ children, classes }) => /* @__PURE__ */ jsx("section", { className: classes.root, children }));
const TemplateEditorLayoutBrowser = withStyles(
  (theme) => ({
    root: {
      gridArea: "browser",
      overflow: "auto",
      [theme.breakpoints.up("md")]: {
        borderRight: `1px solid ${theme.palette.divider}`
      }
    }
  }),
  { name: "ScaffolderTemplateEditorLayoutBrowser" }
)(({ children, classes }) => /* @__PURE__ */ jsx("section", { className: classes.root, children }));
const TemplateEditorLayoutFiles = withStyles(
  (theme) => ({
    root: {
      gridArea: "editor",
      overflow: "auto",
      [theme.breakpoints.up("md")]: {
        height: "100%"
      }
    }
  }),
  { name: "ScaffolderTemplateEditorLayoutFiles" }
)(({ children, classes }) => /* @__PURE__ */ jsx("section", { className: classes.root, children }));
const TemplateEditorLayoutPreview = withStyles(
  (theme) => ({
    root: {
      gridArea: "preview",
      position: "relative",
      backgroundColor: theme.palette.background.default,
      [theme.breakpoints.up("md")]: {
        height: "100%"
      }
    },
    scroll: {
      height: "100%",
      padding: theme.spacing(1),
      [theme.breakpoints.up("md")]: {
        position: "absolute",
        top: 0,
        left: 0,
        right: 0,
        bottom: 0,
        overflow: "auto"
      }
    }
  }),
  { name: "ScaffolderTemplateEditorLayoutPreview" }
)(({ children, classes }) => /* @__PURE__ */ jsx("section", { className: classes.root, children: /* @__PURE__ */ jsx("div", { className: classes.scroll, children }) }));
const TemplateEditorLayoutConsole = withStyles(
  {
    root: {
      gridArea: "results"
    }
  },
  { name: "ScaffolderTemplateEditorLayoutConsole" }
)(({ children, classes }) => /* @__PURE__ */ jsx("section", { className: classes.root, children }));
const TemplateEditorPanelResizeHandle = withStyles(
  {
    root: {
      width: 8,
      cursor: "col-resize",
      background: "rgba(0,0,0,0.04)"
    }
  },
  { name: "ScaffolderTemplateEditorPanelResizeHandle" }
)(({ classes }) => /* @__PURE__ */ jsx(PanelResizeHandle, { className: classes.root }));
function TemplateEditorPanels({
  files,
  preview,
  autoSaveId = "template-editor-panels"
}) {
  const theme = useTheme();
  const isMdUp = useMediaQuery(theme.breakpoints.up("md"));
  if (isMdUp) {
    return /* @__PURE__ */ jsxs(PanelGroup, { direction: "horizontal", autoSaveId, children: [
      /* @__PURE__ */ jsx(Panel, { minSize: 15, defaultSize: 50, children: files }),
      /* @__PURE__ */ jsx(TemplateEditorPanelResizeHandle, {}),
      /* @__PURE__ */ jsx(Panel, { minSize: 15, defaultSize: 50, children: preview })
    ] });
  }
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    files,
    preview
  ] });
}

export { TemplateEditorLayout, TemplateEditorLayoutBrowser, TemplateEditorLayoutConsole, TemplateEditorLayoutFiles, TemplateEditorLayoutPreview, TemplateEditorLayoutToolbar, TemplateEditorPanelResizeHandle, TemplateEditorPanels };
