// @backstage/plugin-scaffolder/dist/alpha/components/TemplateEditorPage/TemplateFormPage.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { useCallback } from 'react';
import { useNavigate } from 'react-router-dom';
import { makeStyles } from '@material-ui/core/styles';
import { Page, Header, Content } from '@backstage/core-components';
import { useRouteRef } from '@backstage/core-plugin-api';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { editRouteRef } from '../../../routes.esm.js';
import { scaffolderTranslationRef } from '../../../translation.esm.js';
import { TemplateFormPreviewer } from './TemplateFormPreviewer.esm.js';

const useStyles = makeStyles({
  root: {
    padding: 0
  }
});
function TemplateFormPage(props) {
  const classes = useStyles();
  const navigate = useNavigate();
  const editLink = useRouteRef(editRouteRef);
  const { t } = useTranslationRef(scaffolderTranslationRef);
  const handleClose = useCallback(() => {
    navigate(editLink());
  }, [navigate, editLink]);
  return /* @__PURE__ */ jsxs(Page, { themeId: "home", children: [
    /* @__PURE__ */ jsx(
      Header,
      {
        title: t("templateFormPage.title"),
        subtitle: t("templateFormPage.subtitle"),
        type: t("templateIntroPage.title"),
        typeLink: editLink()
      }
    ),
    /* @__PURE__ */ jsx(Content, { className: classes.root, children: /* @__PURE__ */ jsx(
      TemplateFormPreviewer,
      {
        layouts: props.layouts,
        formProps: props.formProps,
        customFieldExtensions: props.fieldExtensions,
        defaultPreviewTemplate: props.defaultPreviewTemplate,
        onClose: handleClose
      }
    ) })
  ] });
}

export { TemplateFormPage };
