// @backstage/plugin-scaffolder-backend/dist/scaffolder/actions/builtin/filesystem/read.cjs.js
'use strict';

var pluginScaffolderNode = require('@backstage/plugin-scaffolder-node');
var backendPluginApi = require('@backstage/backend-plugin-api');
var fs = require('fs/promises');
var path = require('path');
var read_examples = require('./read.examples.cjs.js');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var fs__default = /*#__PURE__*/_interopDefaultCompat(fs);
var path__default = /*#__PURE__*/_interopDefaultCompat(path);

const contentSchema = (z) => z.object({
  name: z.string().describe("Name of the file or directory"),
  path: z.string().describe("path to the file or directory relative to the workspace"),
  fullPath: z.string().describe("full path to the file or directory")
});
const createFilesystemReadDirAction = () => {
  return pluginScaffolderNode.createTemplateAction({
    id: "fs:readdir",
    description: "Reads files and directories from the workspace",
    supportsDryRun: true,
    examples: read_examples.examples,
    schema: {
      input: {
        paths: (z) => z.array(z.string().min(1)),
        recursive: (z) => z.boolean().default(false)
      },
      output: {
        files: (z) => z.array(contentSchema(z)),
        folders: (z) => z.array(contentSchema(z))
      }
    },
    async handler(ctx) {
      const files = [];
      const folders = [];
      for (const localPath of ctx.input.paths) {
        const fullWorkspacePath = backendPluginApi.resolveSafeChildPath(
          ctx.workspacePath,
          localPath
        );
        const content = await fs__default.default.readdir(fullWorkspacePath, {
          recursive: ctx.input.recursive,
          withFileTypes: true
        });
        for (const dirent of content) {
          const fullPath = path__default.default.join(dirent.parentPath, dirent.name);
          const element = {
            name: dirent.name,
            path: path__default.default.relative(ctx.workspacePath, fullPath),
            fullPath
          };
          if (dirent.isDirectory()) {
            folders.push(element);
          } else {
            files.push(element);
          }
        }
      }
      ctx.output("files", files);
      ctx.output("folders", folders);
    }
  });
};

exports.createFilesystemReadDirAction = createFilesystemReadDirAction;
