// @backstage/core-components/dist/components/Lifecycle/Lifecycle.esm.js
import { jsx, Fragment } from 'react/jsx-runtime';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';

const useStyles = makeStyles(
  (theme) => ({
    alpha: {
      color: theme.palette.common.white,
      fontFamily: "serif",
      fontWeight: "normal",
      fontStyle: "italic"
    },
    beta: {
      color: "#4d65cc",
      fontFamily: "serif",
      fontWeight: "normal",
      fontStyle: "italic"
    }
  }),
  { name: "BackstageLifecycle" }
);
function Lifecycle(props) {
  const classes = useStyles(props);
  const { shorthand, alpha } = props;
  return shorthand ? /* @__PURE__ */ jsx(
    Typography,
    {
      component: "span",
      className: classes[alpha ? "alpha" : "beta"],
      style: { fontSize: "120%" },
      children: alpha ? /* @__PURE__ */ jsx(Fragment, { children: "\u03B1" }) : /* @__PURE__ */ jsx(Fragment, { children: "\u03B2" })
    }
  ) : /* @__PURE__ */ jsx(Typography, { component: "span", className: classes[alpha ? "alpha" : "beta"], children: alpha ? "Alpha" : "Beta" });
}

export { Lifecycle };
