// @backstage/plugin-scaffolder-react/dist/next/components/Stepper/ErrorListTemplate/errorListTemplate.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import Paper from '@material-ui/core/Paper';
import { makeStyles, createStyles } from '@material-ui/core/styles';
import ErrorIcon from '@material-ui/icons/Error';
import startCase from 'lodash/startCase';

const useStyles = makeStyles(
  (_theme) => createStyles({
    list: {
      width: "100%"
    },
    text: {
      textWrap: "wrap"
    }
  })
);
const ErrorListTemplate = ({ errors, schema }) => {
  const classes = useStyles();
  function formatErrorMessage(error) {
    if (error.property && error.message) {
      const propertyName = error.property.startsWith(".") ? error.property.substring(1) : error.property;
      if (schema.properties && propertyName in schema.properties) {
        const property = schema.properties[propertyName];
        if (typeof property === "object" && "title" in property) {
          return `'${property.title}' ${error.message}`;
        }
      }
      return `'${startCase(propertyName)}' ${error.message}`;
    }
    return error.stack;
  }
  return /* @__PURE__ */ jsx(Paper, { children: /* @__PURE__ */ jsx(List, { dense: true, className: classes.list, children: errors.map((error, index) => /* @__PURE__ */ jsxs(ListItem, { children: [
    /* @__PURE__ */ jsx(ListItemIcon, { children: /* @__PURE__ */ jsx(ErrorIcon, { color: "error" }) }),
    /* @__PURE__ */ jsx(
      ListItemText,
      {
        classes: { primary: classes.text },
        primary: formatErrorMessage(error)
      }
    )
  ] }, index)) }) });
};

export { ErrorListTemplate };
