// @backstage/dev-utils/dist/devApp/render.esm.js
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { createApp } from '@backstage/app-defaults';
import { AppRouter, FlatRoutes } from '@backstage/core-app-api';
import { SidebarItem, SignInPage, AlertDisplay, OAuthRequestDialog, SidebarPage, Sidebar, SidebarSpacer, SidebarSpace, SidebarDivider } from '@backstage/core-components';
import { attachComponentData, createRouteRef, createApiFactory, configApiRef } from '@backstage/core-plugin-api';
import { scmIntegrationsApiRef, ScmIntegrationsApi } from '@backstage/integration-react';
import Box from '@material-ui/core/Box';
import BookmarkIcon from '@material-ui/icons/Bookmark';
import { createRoutesFromChildren, Route } from 'react-router-dom';
import { SidebarThemeSwitcher } from './SidebarThemeSwitcher.esm.js';
import 'react-dom';
import '../components/EntityGridItem/EntityGridItem.esm.js';
import { SidebarSignOutButton } from '../components/SidebarSignOutButton/SidebarSignOutButton.esm.js';
import { SidebarLanguageSwitcher } from '../components/SidebarLanguageSwitcher/SidebarLanguageSwitcher.esm.js';

let ReactDOMPromise;
if (process.env.HAS_REACT_DOM_CLIENT) {
  ReactDOMPromise = import('react-dom/client');
} else {
  ReactDOMPromise = import('react-dom');
}
function isReactRouterBeta() {
  const [obj] = createRoutesFromChildren(/* @__PURE__ */ jsx(Route, { index: true, element: /* @__PURE__ */ jsx("div", {}) }));
  return !obj.index;
}
const MaybeGatheringRoute = ({ element }) => element;
if (isReactRouterBeta()) {
  attachComponentData(MaybeGatheringRoute, "core.gatherMountPoints", true);
}
class DevAppBuilder {
  plugins = new Array();
  apis = new Array();
  rootChildren = new Array();
  routes = new Array();
  sidebarItems = new Array();
  signInProviders = new Array();
  translationResources = new Array();
  defaultPage;
  themes;
  languages;
  defaultLanguage;
  /**
   * Register one or more plugins to render in the dev app
   */
  registerPlugin(...plugins) {
    this.plugins.push(...plugins);
    return this;
  }
  /**
   * Register an API factory to add to the app
   */
  registerApi(factory) {
    this.apis.push(factory);
    return this;
  }
  /**
   * Adds a React node to place just inside the App Provider.
   *
   * Useful for adding more global components like the AlertDisplay.
   */
  addRootChild(node) {
    this.rootChildren.push(node);
    return this;
  }
  /**
   * Adds a new sidebar item to the dev app.
   *
   * Useful for adding only sidebar items without a corresponding page.
   */
  addSidebarItem(sidebarItem) {
    this.sidebarItems.push(sidebarItem);
    return this;
  }
  /**
   * Adds a page component along with accompanying sidebar item.
   *
   * If no path is provided one will be generated.
   * If no title is provided, no sidebar item will be created.
   */
  addPage(opts) {
    const path = opts.path ?? `/page-${this.routes.length + 1}`;
    if (!this.defaultPage || path === "/") {
      this.defaultPage = path;
    }
    if (opts.title) {
      this.sidebarItems.push(
        /* @__PURE__ */ jsx(
          SidebarItem,
          {
            to: path,
            text: opts.title,
            icon: opts.icon ?? BookmarkIcon
          },
          path
        )
      );
    }
    this.routes.push(
      /* @__PURE__ */ jsx(
        MaybeGatheringRoute,
        {
          path,
          element: opts.element,
          children: opts.children
        },
        path
      )
    );
    return this;
  }
  /**
   * Adds an array of themes to override the default theme.
   */
  addThemes(themes) {
    this.themes = themes;
    return this;
  }
  /**
   * Adds new sign in provider for the dev app
   */
  addSignInProvider(provider) {
    this.signInProviders.push(provider);
    return this;
  }
  /**
   * Set available languages to be shown in the dev app
   */
  setAvailableLanguages(languages) {
    this.languages = languages;
    return this;
  }
  /**
   * Add translation resource to the dev app
   */
  addTranslationResource(resource) {
    this.translationResources.push(resource);
    return this;
  }
  /**
   * Set default language for the dev app
   */
  setDefaultLanguage(language) {
    this.defaultLanguage = language;
    return this;
  }
  /**
   * Build a DevApp component using the resources registered so far
   */
  build() {
    const fakeRouteRef = createRouteRef({ id: "fake" });
    const FakePage = () => /* @__PURE__ */ jsx(Box, { p: 3, children: "Page belonging to another plugin." });
    attachComponentData(FakePage, "core.mountPoint", fakeRouteRef);
    const apis = [...this.apis];
    if (!apis.some((api) => api.api.id === scmIntegrationsApiRef.id)) {
      apis.push(
        createApiFactory({
          api: scmIntegrationsApiRef,
          deps: { configApi: configApiRef },
          factory: ({ configApi }) => ScmIntegrationsApi.fromConfig(configApi)
        })
      );
    }
    const app = createApp({
      apis,
      plugins: this.plugins,
      themes: this.themes,
      components: {
        SignInPage: (props) => {
          return /* @__PURE__ */ jsx(
            SignInPage,
            {
              ...props,
              providers: ["guest", ...this.signInProviders],
              title: "Select a sign-in method",
              align: "center"
            }
          );
        }
      },
      bindRoutes: ({ bind }) => {
        for (const plugin of this.plugins ?? []) {
          const targets = {};
          for (const routeKey of Object.keys(plugin.externalRoutes)) {
            targets[routeKey] = fakeRouteRef;
          }
          bind(plugin.externalRoutes, targets);
        }
      },
      __experimentalTranslations: {
        defaultLanguage: this.defaultLanguage,
        availableLanguages: this.languages,
        resources: this.translationResources
      }
    });
    const DevApp = /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx(AlertDisplay, {}),
      /* @__PURE__ */ jsx(OAuthRequestDialog, {}),
      this.rootChildren,
      /* @__PURE__ */ jsx(AppRouter, { children: /* @__PURE__ */ jsxs(SidebarPage, { children: [
        /* @__PURE__ */ jsxs(Sidebar, { children: [
          /* @__PURE__ */ jsx(SidebarSpacer, {}),
          this.sidebarItems,
          /* @__PURE__ */ jsx(SidebarSpace, {}),
          /* @__PURE__ */ jsx(SidebarDivider, {}),
          /* @__PURE__ */ jsx(SidebarThemeSwitcher, {}),
          /* @__PURE__ */ jsx(SidebarLanguageSwitcher, {}),
          /* @__PURE__ */ jsx(SidebarSignOutButton, {})
        ] }),
        /* @__PURE__ */ jsxs(FlatRoutes, { children: [
          this.routes,
          /* @__PURE__ */ jsx(Route, { path: "/_external_route", element: /* @__PURE__ */ jsx(FakePage, {}) })
        ] })
      ] }) })
    ] });
    return app.createRoot(DevApp);
  }
  /**
   * Build and render directory to #root element, with react hot loading.
   */
  render() {
    const DevApp = this.build();
    if (window.location.pathname === "/" && this.defaultPage && this.defaultPage !== "/") {
      window.location.pathname = this.defaultPage;
    }
    ReactDOMPromise.then((ReactDOM) => {
      if ("createRoot" in ReactDOM) {
        ReactDOM.createRoot(document.getElementById("root")).render(
          /* @__PURE__ */ jsx(DevApp, {})
        );
      } else {
        ReactDOM.render(/* @__PURE__ */ jsx(DevApp, {}), document.getElementById("root"));
      }
    });
  }
}
function createDevApp() {
  return new DevAppBuilder();
}

export { DevAppBuilder, createDevApp, isReactRouterBeta };
