// @backstage/frontend-test-utils/dist/frontend-app-api/src/tree/instantiateAppNodeTree.esm.js
import mapValues from 'lodash/mapValues';
import { toInternalExtension } from '../../../frontend-plugin-api/src/wiring/resolveExtensionDefinition.esm.js';
import { createExtensionDataContainer } from '../../../frontend-internal/src/wiring/createExtensionDataContainer.esm.js';
import '../../../frontend-internal/src/wiring/InternalSwappableComponentRef.esm.js';
import '../../../frontend-internal/src/wiring/InternalExtensionDefinition.esm.js';
import '../../../frontend-internal/src/wiring/InternalFrontendPlugin.esm.js';

const INSTANTIATION_FAILED = new Error("Instantiation failed");
function mapWithFailures(iterable, callback) {
  let failed = false;
  const results = Array.from(iterable).map((item) => {
    try {
      return callback(item);
    } catch (error) {
      if (error === INSTANTIATION_FAILED) {
        failed = true;
      } else {
        throw error;
      }
      return null;
    }
  });
  if (failed) {
    throw INSTANTIATION_FAILED;
  }
  return results;
}
function resolveV1InputDataMap(dataMap, attachment, inputName) {
  return Object.fromEntries(
    mapWithFailures(Object.entries(dataMap), ([key, ref]) => {
      const value = attachment.instance?.getData(ref);
      if (value === void 0 && !ref.config.optional) {
        const expected = Object.values(dataMap).filter((r) => !r.config.optional).map((r) => `'${r.id}'`).join(", ");
        const provided = [...attachment.instance?.getDataRefs() ?? []].map((r) => `'${r.id}'`).join(", ");
        throw new Error(
          `extension '${attachment.spec.id}' could not be attached because its output data (${provided}) does not match what the input '${inputName}' requires (${expected})`
        );
      }
      return [key, value];
    })
  );
}
function resolveInputDataContainer(extensionData, attachment, inputName, collector) {
  const dataMap = /* @__PURE__ */ new Map();
  mapWithFailures(extensionData, (ref) => {
    if (dataMap.has(ref.id)) {
      collector.report({
        code: "EXTENSION_INPUT_DATA_IGNORED",
        message: `Unexpected duplicate input data '${ref.id}'`
      });
      return;
    }
    const value = attachment.instance?.getData(ref);
    if (value === void 0 && !ref.config.optional) {
      const expected = extensionData.filter((r) => !r.config.optional).map((r) => `'${r.id}'`).join(", ");
      const provided = [...attachment.instance?.getDataRefs() ?? []].map((r) => `'${r.id}'`).join(", ");
      collector.report({
        code: "EXTENSION_INPUT_DATA_MISSING",
        message: `extension '${attachment.spec.id}' could not be attached because its output data (${provided}) does not match what the input '${inputName}' requires (${expected})`
      });
      throw INSTANTIATION_FAILED;
    }
    dataMap.set(ref.id, value);
  });
  return {
    node: attachment,
    get(ref) {
      return dataMap.get(ref.id);
    },
    *[Symbol.iterator]() {
      for (const [id, value] of dataMap) {
        yield {
          $$type: "@backstage/ExtensionDataValue",
          id,
          value
        };
      }
    }
  };
}
function reportUndeclaredAttachments(id, inputMap, attachments) {
  const undeclaredAttachments = Array.from(attachments.entries()).filter(
    ([inputName]) => inputMap[inputName] === void 0
  );
  const inputNames = Object.keys(inputMap);
  for (const [name, nodes] of undeclaredAttachments) {
    const pl = nodes.length > 1;
    console.warn(
      [
        `The extension${pl ? "s" : ""} '${nodes.map((n) => n.spec.id).join("', '")}' ${pl ? "are" : "is"}`,
        `attached to the input '${name}' of the extension '${id}', but it`,
        inputNames.length === 0 ? "has no inputs" : `has no such input (candidates are '${inputNames.join("', '")}')`
      ].join(" ")
    );
  }
}
function resolveV1Inputs(inputMap, attachments) {
  return Object.fromEntries(
    mapWithFailures(Object.entries(inputMap), ([inputName, input]) => {
      const attachedNodes = attachments.get(inputName) ?? [];
      if (input.config.singleton) {
        if (attachedNodes.length > 1) {
          const attachedNodeIds = attachedNodes.map((e) => e.spec.id);
          throw Error(
            `expected ${input.config.optional ? "at most" : "exactly"} one '${inputName}' input but received multiple: '${attachedNodeIds.join(
              "', '"
            )}'`
          );
        } else if (attachedNodes.length === 0) {
          if (input.config.optional) {
            return [inputName, void 0];
          }
          throw Error(`input '${inputName}' is required but was not received`);
        }
        return [
          inputName,
          {
            node: attachedNodes[0],
            output: resolveV1InputDataMap(
              input.extensionData,
              attachedNodes[0],
              inputName
            )
          }
        ];
      }
      return [
        inputName,
        attachedNodes.map((attachment) => ({
          node: attachment,
          output: resolveV1InputDataMap(
            input.extensionData,
            attachment,
            inputName
          )
        }))
      ];
    })
  );
}
function resolveV2Inputs(inputMap, attachments, parentCollector) {
  return mapValues(inputMap, (input, inputName) => {
    const attachedNodes = attachments.get(inputName) ?? [];
    const collector = parentCollector.child({ inputName });
    if (input.config.singleton) {
      if (attachedNodes.length > 1) {
        const attachedNodeIds = attachedNodes.map((e) => e.spec.id).join("', '");
        collector.report({
          code: "EXTENSION_ATTACHMENT_CONFLICT",
          message: `expected ${input.config.optional ? "at most" : "exactly"} one '${inputName}' input but received multiple: '${attachedNodeIds}'`
        });
        throw INSTANTIATION_FAILED;
      } else if (attachedNodes.length === 0) {
        if (!input.config.optional) {
          collector.report({
            code: "EXTENSION_ATTACHMENT_MISSING",
            message: `input '${inputName}' is required but was not received`
          });
          throw INSTANTIATION_FAILED;
        }
        return void 0;
      }
      return resolveInputDataContainer(
        input.extensionData,
        attachedNodes[0],
        inputName,
        collector
      );
    }
    return mapWithFailures(
      attachedNodes,
      (attachment) => resolveInputDataContainer(
        input.extensionData,
        attachment,
        inputName,
        collector
      )
    );
  });
}
function createAppNodeInstance(options) {
  const { node, apis, attachments } = options;
  const collector = options.collector.child({ node });
  const { id, extension, config } = node.spec;
  const extensionData = /* @__PURE__ */ new Map();
  const extensionDataRefs = /* @__PURE__ */ new Set();
  let parsedConfig;
  try {
    parsedConfig = extension.configSchema?.parse(config ?? {});
  } catch (e) {
    collector.report({
      code: "EXTENSION_CONFIGURATION_INVALID",
      message: `Invalid configuration for extension '${id}'; caused by ${e}`
    });
    return void 0;
  }
  try {
    const internalExtension = toInternalExtension(extension);
    if (process.env.NODE_ENV !== "production") {
      reportUndeclaredAttachments(id, internalExtension.inputs, attachments);
    }
    if (internalExtension.version === "v1") {
      const namedOutputs = internalExtension.factory({
        node,
        apis,
        config: parsedConfig,
        inputs: resolveV1Inputs(internalExtension.inputs, attachments)
      });
      for (const [name, output] of Object.entries(namedOutputs)) {
        const ref = internalExtension.output[name];
        if (!ref) {
          throw new Error(`unknown output provided via '${name}'`);
        }
        if (extensionData.has(ref.id)) {
          throw new Error(
            `duplicate extension data '${ref.id}' received via output '${name}'`
          );
        }
        extensionData.set(ref.id, output);
        extensionDataRefs.add(ref);
      }
    } else if (internalExtension.version === "v2") {
      const context = {
        node,
        apis,
        config: parsedConfig,
        inputs: resolveV2Inputs(
          internalExtension.inputs,
          attachments,
          collector
        )
      };
      const outputDataValues = options.extensionFactoryMiddleware ? createExtensionDataContainer(
        options.extensionFactoryMiddleware((overrideContext) => {
          return createExtensionDataContainer(
            internalExtension.factory({
              node: context.node,
              apis: context.apis,
              inputs: context.inputs,
              config: overrideContext?.config ?? context.config
            }),
            "extension factory"
          );
        }, context),
        "extension factory middleware"
      ) : internalExtension.factory(context);
      if (typeof outputDataValues !== "object" || !outputDataValues?.[Symbol.iterator]) {
        throw new Error("extension factory did not provide an iterable object");
      }
      const outputDataMap = /* @__PURE__ */ new Map();
      mapWithFailures(outputDataValues, (value) => {
        if (outputDataMap.has(value.id)) {
          collector.report({
            code: "EXTENSION_OUTPUT_CONFLICT",
            message: `extension factory output duplicate data '${value.id}'`,
            context: {
              dataRefId: value.id
            }
          });
          throw INSTANTIATION_FAILED;
        } else {
          outputDataMap.set(value.id, value.value);
        }
      });
      for (const ref of internalExtension.output) {
        const value = outputDataMap.get(ref.id);
        outputDataMap.delete(ref.id);
        if (value === void 0) {
          if (!ref.config.optional) {
            collector.report({
              code: "EXTENSION_OUTPUT_MISSING",
              message: `missing required extension data output '${ref.id}'`,
              context: {
                dataRefId: ref.id
              }
            });
            throw INSTANTIATION_FAILED;
          }
        } else {
          extensionData.set(ref.id, value);
          extensionDataRefs.add(ref);
        }
      }
      if (outputDataMap.size > 0) {
        for (const dataRefId of outputDataMap.keys()) {
          collector.report({
            code: "EXTENSION_OUTPUT_IGNORED",
            message: `unexpected output '${dataRefId}'`,
            context: {
              dataRefId
            }
          });
        }
      }
    } else {
      collector.report({
        code: "EXTENSION_INVALID",
        message: `unexpected extension version '${internalExtension.version}'`
      });
      throw INSTANTIATION_FAILED;
    }
  } catch (e) {
    if (e !== INSTANTIATION_FAILED) {
      collector.report({
        code: "EXTENSION_FACTORY_ERROR",
        message: `Failed to instantiate extension '${id}'${e.name === "Error" ? `, ${e.message}` : `; caused by ${e}`}`
      });
    }
    return void 0;
  }
  return {
    getDataRefs() {
      return extensionDataRefs.values();
    },
    getData(ref) {
      return extensionData.get(ref.id);
    }
  };
}
function instantiateAppNodeTree(rootNode, apis, collector, extensionFactoryMiddleware) {
  function createInstance(node) {
    if (node.instance) {
      return node.instance;
    }
    if (node.spec.disabled) {
      return void 0;
    }
    const instantiatedAttachments = /* @__PURE__ */ new Map();
    for (const [input, children] of node.edges.attachments) {
      const instantiatedChildren = children.flatMap((child) => {
        const childInstance = createInstance(child);
        if (!childInstance) {
          return [];
        }
        return [child];
      });
      if (instantiatedChildren.length > 0) {
        instantiatedAttachments.set(input, instantiatedChildren);
      }
    }
    node.instance = createAppNodeInstance({
      extensionFactoryMiddleware,
      node,
      apis,
      attachments: instantiatedAttachments,
      collector
    });
    return node.instance;
  }
  return createInstance(rootNode) !== void 0;
}

export { createAppNodeInstance, instantiateAppNodeTree };
