// @backstage/plugin-search/dist/package.json.esm.js
var name = "@backstage/plugin-search";
var version = "1.4.30";
var description = "The Backstage plugin that provides your backstage app with search";
var backstage = {
	role: "frontend-plugin",
	pluginId: "search",
	pluginPackages: [
		"@backstage/plugin-search",
		"@backstage/plugin-search-backend",
		"@backstage/plugin-search-backend-node",
		"@backstage/plugin-search-common",
		"@backstage/plugin-search-react"
	]
};
var publishConfig = {
	access: "public"
};
var keywords = [
	"backstage"
];
var homepage = "https://backstage.io";
var repository = {
	type: "git",
	url: "https://github.com/backstage/backstage",
	directory: "plugins/search"
};
var license = "Apache-2.0";
var sideEffects = false;
var exports = {
	".": "./src/index.ts",
	"./alpha": "./src/alpha.tsx",
	"./package.json": "./package.json"
};
var main = "src/index.ts";
var types = "src/index.ts";
var typesVersions = {
	"*": {
		alpha: [
			"src/alpha.tsx"
		],
		"package.json": [
			"package.json"
		]
	}
};
var files = [
	"dist",
	"config.d.ts"
];
var scripts = {
	build: "backstage-cli package build",
	clean: "backstage-cli package clean",
	lint: "backstage-cli package lint",
	prepack: "backstage-cli package prepack",
	postpack: "backstage-cli package postpack",
	start: "backstage-cli package start",
	test: "backstage-cli package test"
};
var dependencies = {
	"@backstage/core-compat-api": "workspace:^",
	"@backstage/core-components": "workspace:^",
	"@backstage/core-plugin-api": "workspace:^",
	"@backstage/errors": "workspace:^",
	"@backstage/frontend-plugin-api": "workspace:^",
	"@backstage/plugin-catalog-react": "workspace:^",
	"@backstage/plugin-search-common": "workspace:^",
	"@backstage/plugin-search-react": "workspace:^",
	"@backstage/types": "workspace:^",
	"@backstage/version-bridge": "workspace:^",
	"@material-ui/core": "^4.12.2",
	"@material-ui/icons": "^4.9.1",
	qs: "^6.9.4",
	"react-use": "^17.2.4"
};
var devDependencies = {
	"@backstage/cli": "workspace:^",
	"@backstage/core-app-api": "workspace:^",
	"@backstage/dev-utils": "workspace:^",
	"@backstage/test-utils": "workspace:^",
	"@testing-library/dom": "^10.0.0",
	"@testing-library/jest-dom": "^6.0.0",
	"@testing-library/react": "^16.0.0",
	"@testing-library/user-event": "^14.0.0",
	"@types/react": "^18.0.0",
	history: "^5.0.0",
	react: "^18.0.2",
	"react-dom": "^18.0.2",
	"react-router-dom": "^6.3.0"
};
var peerDependencies = {
	"@types/react": "^17.0.0 || ^18.0.0",
	react: "^17.0.0 || ^18.0.0",
	"react-dom": "^17.0.0 || ^18.0.0",
	"react-router-dom": "^6.3.0"
};
var peerDependenciesMeta = {
	"@types/react": {
		optional: true
	}
};
var configSchema = "config.d.ts";
var _package = {
	name: name,
	version: version,
	description: description,
	backstage: backstage,
	publishConfig: publishConfig,
	keywords: keywords,
	homepage: homepage,
	repository: repository,
	license: license,
	sideEffects: sideEffects,
	exports: exports,
	main: main,
	types: types,
	typesVersions: typesVersions,
	files: files,
	scripts: scripts,
	dependencies: dependencies,
	devDependencies: devDependencies,
	peerDependencies: peerDependencies,
	peerDependenciesMeta: peerDependenciesMeta,
	configSchema: configSchema
};

export { backstage, configSchema, _package as default, dependencies, description, devDependencies, exports, files, homepage, keywords, license, main, name, peerDependencies, peerDependenciesMeta, publishConfig, repository, scripts, sideEffects, types, typesVersions, version };
