// @janus-idp/shared-react/dist/types/pipeline/computedStatus.esm.js
var TerminatedReasons = /* @__PURE__ */ ((TerminatedReasons2) => {
  TerminatedReasons2["Completed"] = "Completed";
  return TerminatedReasons2;
})(TerminatedReasons || {});
var ComputedStatus = /* @__PURE__ */ ((ComputedStatus2) => {
  ComputedStatus2["All"] = "All";
  ComputedStatus2["Cancelling"] = "Cancelling";
  ComputedStatus2["Succeeded"] = "Succeeded";
  ComputedStatus2["Failed"] = "Failed";
  ComputedStatus2["Running"] = "Running";
  ComputedStatus2["In Progress"] = "In Progress";
  ComputedStatus2["FailedToStart"] = "FailedToStart";
  ComputedStatus2["PipelineNotStarted"] = "PipelineNotStarted";
  ComputedStatus2["Skipped"] = "Skipped";
  ComputedStatus2["Cancelled"] = "Cancelled";
  ComputedStatus2["Pending"] = "Pending";
  ComputedStatus2["Idle"] = "Idle";
  ComputedStatus2["Other"] = "Other";
  return ComputedStatus2;
})(ComputedStatus || {});
var SucceedConditionReason = /* @__PURE__ */ ((SucceedConditionReason2) => {
  SucceedConditionReason2["PipelineRunCancelled"] = "StoppedRunFinally";
  SucceedConditionReason2["PipelineRunStopped"] = "CancelledRunFinally";
  SucceedConditionReason2["TaskRunCancelled"] = "TaskRunCancelled";
  SucceedConditionReason2["Cancelled"] = "Cancelled";
  SucceedConditionReason2["PipelineRunStopping"] = "PipelineRunStopping";
  SucceedConditionReason2["PipelineRunPending"] = "PipelineRunPending";
  SucceedConditionReason2["TaskRunStopping"] = "TaskRunStopping";
  SucceedConditionReason2["CreateContainerConfigError"] = "CreateContainerConfigError";
  SucceedConditionReason2["ExceededNodeResources"] = "ExceededNodeResources";
  SucceedConditionReason2["ExceededResourceQuota"] = "ExceededResourceQuota";
  SucceedConditionReason2["ConditionCheckFailed"] = "ConditionCheckFailed";
  return SucceedConditionReason2;
})(SucceedConditionReason || {});

export { ComputedStatus, SucceedConditionReason, TerminatedReasons };
