// @backstage/repo-tools/dist/commands/generate-catalog-info/utils.cjs.js
'use strict';

var fs = require('fs');
var util = require('util');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var fs__default = /*#__PURE__*/_interopDefaultCompat(fs);

const readFile = util.promisify(fs__default.default.readFile);
const writeFile = util.promisify(fs__default.default.writeFile);
function isBackstagePackage(packageJson) {
  return packageJson && packageJson.hasOwnProperty("backstage") && packageJson?.backstage?.role !== "undefined";
}
const isRejected = (input) => input.status === "rejected";
const isFulfilled = (input) => input.status === "fulfilled";
const safeEntityName = (packageName) => {
  return packageName.replace(/^[^\w\s]|[^a-z0-9]$/g, "").replace(/[^A-Za-z0-9_\-.]+/g, "-").replace(
    /([a-z])([A-Z])/g,
    (_, a, b) => `${a}-${b.toLocaleLowerCase("en-US")}`
  ).replace(/^(.)/, (_, a) => a.toLocaleLowerCase("en-US"));
};

exports.isBackstagePackage = isBackstagePackage;
exports.isFulfilled = isFulfilled;
exports.isRejected = isRejected;
exports.readFile = readFile;
exports.safeEntityName = safeEntityName;
exports.writeFile = writeFile;
