// @backstage/plugin-api-docs/dist/components/GraphQlDefinitionWidget/GraphQlDefinition.esm.js
import { jsx } from 'react/jsx-runtime';
import { makeStyles } from '@material-ui/core/styles';
import { EditorContextProvider, SchemaContextProvider, ExplorerContextProvider, DocExplorer } from '@graphiql/react';
import 'graphiql/graphiql.css';
import { buildSchema } from 'graphql';

const useStyles = makeStyles({
  root: {
    height: "100%",
    display: "flex",
    flexFlow: "column nowrap"
  },
  graphiQlWrapper: {
    flex: 1,
    "@global": {
      ".graphiql-container": {
        boxSizing: "initial",
        height: "100%",
        minHeight: "600px",
        flex: "1 1 auto"
      },
      ".graphiql-sidebar": {
        width: "100%"
      }
    }
  }
});
const GraphQlDefinition = ({ definition }) => {
  const classes = useStyles();
  const schema = buildSchema(definition);
  return /* @__PURE__ */ jsx("div", { className: classes.root, children: /* @__PURE__ */ jsx("div", { className: classes.graphiQlWrapper, children: /* @__PURE__ */ jsx(EditorContextProvider, { children: /* @__PURE__ */ jsx(
    SchemaContextProvider,
    {
      schema,
      fetcher: () => Promise.resolve(null),
      children: /* @__PURE__ */ jsx("div", { className: "graphiql-container", children: /* @__PURE__ */ jsx("div", { className: "graphiql-sidebar", children: /* @__PURE__ */ jsx("div", { className: "graphiql-sidebar-section", children: /* @__PURE__ */ jsx(ExplorerContextProvider, { children: /* @__PURE__ */ jsx(DocExplorer, {}) }) }) }) })
    }
  ) }) }) });
};

export { GraphQlDefinition };
