// @backstage/core-components/dist/components/FavoriteToggle/FavoriteToggle.esm.js
import { jsx } from 'react/jsx-runtime';
import IconButton from '@material-ui/core/IconButton';
import Tooltip from '@material-ui/core/Tooltip';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';
import { StarIcon, UnstarredIcon } from '../../icons/icons.esm.js';

const useStyles = makeStyles(
  () => ({
    icon: {
      color: "#f3ba37",
      cursor: "pointer",
      display: "inline-flex"
    },
    iconBorder: {
      color: "inherit",
      cursor: "pointer",
      display: "inline-flex"
    }
  }),
  { name: "BackstageFavoriteToggleIcon" }
);
function FavoriteToggleIcon(props) {
  const { isFavorite } = props;
  const classes = useStyles();
  return /* @__PURE__ */ jsx(
    Typography,
    {
      component: "span",
      className: isFavorite ? classes.icon : classes.iconBorder,
      children: isFavorite ? /* @__PURE__ */ jsx(StarIcon, {}) : /* @__PURE__ */ jsx(UnstarredIcon, {})
    }
  );
}
function FavoriteToggle(props) {
  const {
    id,
    title,
    isFavorite: value,
    onToggle: onChange,
    ...iconButtonProps
  } = props;
  return /* @__PURE__ */ jsx(Tooltip, { id, title, children: /* @__PURE__ */ jsx(
    IconButton,
    {
      "aria-label": title,
      id,
      onClick: () => onChange(!value),
      color: "inherit",
      ...iconButtonProps,
      children: /* @__PURE__ */ jsx(FavoriteToggleIcon, { isFavorite: value })
    }
  ) });
}

export { FavoriteToggle, FavoriteToggleIcon };
