// @backstage-community/plugin-rbac/dist/components/RolesList/RolesListToolbar.esm.js
import React from 'react';
import { Link, LinkButton } from '@backstage/core-components';
import Alert from '@mui/material/Alert';
import AlertTitle from '@mui/material/AlertTitle';
import Typography from '@mui/material/Typography';

const RolesListToolbar = ({
  createRoleAllowed,
  createRoleLoading
}) => {
  return /* @__PURE__ */ React.createElement("div", null, !createRoleLoading && !createRoleAllowed && /* @__PURE__ */ React.createElement(Alert, { severity: "warning", "data-testid": "create-role-warning" }, /* @__PURE__ */ React.createElement(
    AlertTitle,
    {
      sx: {
        fontWeight: "bold"
      }
    },
    "Unable to create role."
  ), "To enable create/edit role button, make sure required users/groups are available in catalog as a role cannot be created without users/groups and also the role associated with your user should have the permission policies mentioned", " ", /* @__PURE__ */ React.createElement(
    Link,
    {
      to: "https://github.com/backstage/community-plugins/tree/main/workspaces/rbac/plugins/rbac#prerequisites",
      target: "_blank"
    },
    "here"
  ), /* @__PURE__ */ React.createElement(Typography, { sx: { mt: 1 }, fontSize: "small" }, /* @__PURE__ */ React.createElement(Typography, { component: "span", fontWeight: "bold", fontSize: "small" }, "Note"), ": Even after ingesting users/groups in catalog and applying above permissions if the create/edit button is still disabled then please contact your administrator as you might be conditionally restricted from accessing the create/edit button.")), /* @__PURE__ */ React.createElement("br", null), /* @__PURE__ */ React.createElement(
    Typography,
    {
      component: "span",
      sx: {
        display: "flex",
        justifyContent: "end",
        marginBottom: "24px !important"
      }
    },
    /* @__PURE__ */ React.createElement(
      LinkButton,
      {
        to: "role/new",
        color: "primary",
        variant: "contained",
        disabled: !createRoleAllowed,
        "data-testid": "create-role"
      },
      "Create"
    )
  ));
};

export { RolesListToolbar };
