// @backstage/repo-tools/dist/lib/openapi/constants.cjs.js
'use strict';

const YAML_SCHEMA_PATH = "src/schema/openapi.yaml";
const OUTPUT_PATH = "src/schema/openapi/generated";
const TS_MODULE = `${OUTPUT_PATH}/router`;
const OLD_SCHEMA_PATH = `src/schema/openapi.generated.ts`;
const TS_SCHEMA_PATH = `${TS_MODULE}.ts`;
const OPENAPI_IGNORE_FILES = [
  // Get rid of the default files.
  "*.md",
  "*.mustache",
  // The rest of these have to be explicit, otherwise they get added if this was a *.*
  "apis/baseapi.ts",
  "apis/exception.ts",
  "auth/*",
  "http/*",
  "middleware.ts",
  "servers.ts",
  "util.ts",
  "configuration.ts",
  "rxjsStub.ts",
  ".gitignore",
  // Override the created version.
  "apis/*.ts",
  "!apis/*.client.ts",
  "!apis/*.server.ts",
  "models/*.ts",
  "!models/*.model.ts",
  // Always include index.ts files.
  "!index.ts",
  "!**/index.ts",
  // Weird API typings.
  "types/ObjectParamAPI.ts",
  "types/ObservableAPI.ts",
  "types/PromiseAPI.ts",
  "git_push.sh",
  "package.json",
  "tsconfig.json"
];

exports.OLD_SCHEMA_PATH = OLD_SCHEMA_PATH;
exports.OPENAPI_IGNORE_FILES = OPENAPI_IGNORE_FILES;
exports.OUTPUT_PATH = OUTPUT_PATH;
exports.TS_MODULE = TS_MODULE;
exports.TS_SCHEMA_PATH = TS_SCHEMA_PATH;
exports.YAML_SCHEMA_PATH = YAML_SCHEMA_PATH;
