// @useoptic/openapi-utilities/build/openapi3/sdk/types/openApiKinds.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenApiParameterKind = exports.OpenApiKind = void 0;
var OpenApiKind;
(function (OpenApiKind) {
    OpenApiKind["Specification"] = "specification";
    OpenApiKind["Operation"] = "operation";
    OpenApiKind["Request"] = "request";
    OpenApiKind["QueryParameter"] = "query-parameter";
    OpenApiKind["PathParameter"] = "path-parameter";
    OpenApiKind["HeaderParameter"] = "header-parameter";
    OpenApiKind["CookieParameter"] = "cookie-parameter";
    OpenApiKind["ResponseHeader"] = "response-header";
    OpenApiKind["Response"] = "response";
    OpenApiKind["Body"] = "body";
    OpenApiKind["BodyExample"] = "body-example";
    OpenApiKind["Field"] = "field";
    OpenApiKind["Schema"] = "schema";
    OpenApiKind["ComponentSchemaExample"] = "component-schema-example";
})(OpenApiKind || (exports.OpenApiKind = OpenApiKind = {}));
// allow for iterations and `.includes` calls
exports.OpenApiParameterKind = [
    OpenApiKind.HeaderParameter,
    OpenApiKind.PathParameter,
    OpenApiKind.QueryParameter,
    OpenApiKind.CookieParameter,
];
