// @backstage/plugin-catalog-node/dist/alpha.cjs.js
'use strict';

var backendPluginApi = require('@backstage/backend-plugin-api');
var catalogClient = require('@backstage/catalog-client');
var alpha = require('@backstage/plugin-catalog-common/alpha');
var pluginPermissionNode = require('@backstage/plugin-permission-node');
var extensions = require('./extensions.cjs.js');

const catalogEntityPermissionResourceRef = pluginPermissionNode.createPermissionResourceRef().with({
  pluginId: "catalog",
  resourceType: alpha.RESOURCE_TYPE_CATALOG_ENTITY
});
const catalogServiceRef = backendPluginApi.createServiceRef({
  id: "catalog-client-legacy",
  defaultFactory: async (service) => backendPluginApi.createServiceFactory({
    service,
    deps: {
      discoveryApi: backendPluginApi.coreServices.discovery
    },
    async factory({ discoveryApi }) {
      return new catalogClient.CatalogClient({ discoveryApi });
    }
  })
});

exports.catalogAnalysisExtensionPoint = extensions.catalogAnalysisExtensionPoint;
exports.catalogLocationsExtensionPoint = extensions.catalogLocationsExtensionPoint;
exports.catalogModelExtensionPoint = extensions.catalogModelExtensionPoint;
exports.catalogPermissionExtensionPoint = extensions.catalogPermissionExtensionPoint;
exports.catalogProcessingExtensionPoint = extensions.catalogProcessingExtensionPoint;
exports.catalogEntityPermissionResourceRef = catalogEntityPermissionResourceRef;
exports.catalogServiceRef = catalogServiceRef;
