// @backstage/plugin-catalog/dist/alpha/components/EntityTabs/EntityTabsPanel.esm.js
import { jsx } from 'react/jsx-runtime';
import classnames from 'classnames';
import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles(
  (theme) => ({
    root: {
      gridArea: "pageContent",
      minWidth: 0,
      paddingTop: theme.spacing(3),
      paddingBottom: theme.spacing(3),
      paddingLeft: theme.spacing(2),
      paddingRight: theme.spacing(2),
      [theme.breakpoints.up("sm")]: {
        paddingLeft: theme.spacing(3),
        paddingRight: theme.spacing(3)
      }
    },
    stretch: {
      display: "flex",
      flexDirection: "column",
      flexGrow: 1
    },
    noPadding: {
      padding: 0
    }
  }),
  { name: "EntityTabsPanel" }
);
function EntityTabsPanel(props) {
  const { className, stretch, noPadding, children, ...restProps } = props;
  const classes = useStyles();
  return /* @__PURE__ */ jsx(
    "article",
    {
      ...restProps,
      className: classnames(classes.root, className, {
        [classes.stretch]: stretch,
        [classes.noPadding]: noPadding
      }),
      children
    }
  );
}

export { EntityTabsPanel };
