// @backstage/plugin-catalog/dist/components/EntityLabelsCard/EntityLabelsCard.esm.js
import { jsx } from 'react/jsx-runtime';
import { useEntity } from '@backstage/plugin-catalog-react';
import { InfoCard, Table } from '@backstage/core-components';
import { EntityLabelsEmptyState } from './EntityLabelsEmptyState.esm.js';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';
import { catalogTranslationRef } from '../../alpha/translation.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';

const useStyles = makeStyles((_) => ({
  key: {
    fontWeight: "bold"
  }
}));
const EntityLabelsCard = (props) => {
  const { variant, title } = props;
  const { entity } = useEntity();
  const classes = useStyles();
  const { t } = useTranslationRef(catalogTranslationRef);
  const columns = [
    {
      render: (row) => {
        return /* @__PURE__ */ jsx(Typography, { className: classes.key, variant: "body2", children: row.key });
      }
    },
    {
      field: "value"
    }
  ];
  const labels = entity?.metadata?.labels;
  return /* @__PURE__ */ jsx(InfoCard, { title: title || t("entityLabelsCard.title"), variant, children: !labels || Object.keys(labels).length === 0 ? /* @__PURE__ */ jsx(EntityLabelsEmptyState, {}) : /* @__PURE__ */ jsx(
    Table,
    {
      columns,
      data: Object.keys(labels).map((labelKey) => ({
        key: labelKey,
        value: labels[labelKey]
      })),
      options: {
        search: false,
        showTitle: true,
        loadingType: "linear",
        header: false,
        padding: "dense",
        pageSize: 5,
        toolbar: false,
        paging: Object.keys(labels).length > 5
      }
    }
  ) });
};

export { EntityLabelsCard };
