// @backstage/core-components/dist/layout/ErrorPage/ErrorPage.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import Grid from '@material-ui/core/Grid';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import { useNavigate } from 'react-router-dom';
import { Link } from '../../components/Link/Link.esm.js';
import 'lodash';
import 'qs';
import 'react';
import '@react-hookz/web';
import { useSupportConfig } from '../../hooks/useSupportConfig.esm.js';
import { MicDrop } from './MicDrop.esm.js';
import { StackDetails } from './StackDetails.esm.js';
import { coreComponentsTranslationRef } from '../../translation.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';

const useStyles = makeStyles(
  (theme) => ({
    container: {
      padding: theme.spacing(8),
      [theme.breakpoints.down("xs")]: {
        padding: theme.spacing(2)
      }
    },
    title: {
      paddingBottom: theme.spacing(5),
      [theme.breakpoints.down("xs")]: {
        paddingBottom: theme.spacing(4),
        fontSize: theme.typography.h3.fontSize
      }
    },
    subtitle: {
      color: theme.palette.textSubtle
    }
  }),
  { name: "BackstageErrorPage" }
);
function ErrorPage(props) {
  const {
    status = "",
    statusMessage,
    additionalInfo,
    supportUrl,
    stack
  } = props;
  const classes = useStyles();
  const navigate = useNavigate();
  const support = useSupportConfig();
  const { t } = useTranslationRef(coreComponentsTranslationRef);
  return /* @__PURE__ */ jsxs(Grid, { container: true, className: classes.container, children: [
    /* @__PURE__ */ jsxs(Grid, { item: true, xs: 12, sm: 8, md: 4, children: [
      /* @__PURE__ */ jsx(
        Typography,
        {
          "data-testid": "error",
          variant: "body1",
          className: classes.subtitle,
          children: t("errorPage.subtitle", { status, statusMessage })
        }
      ),
      /* @__PURE__ */ jsx(Typography, { variant: "body1", className: classes.subtitle, children: additionalInfo }),
      /* @__PURE__ */ jsx(Typography, { variant: "h2", className: classes.title, children: t("errorPage.title") }),
      /* @__PURE__ */ jsxs(Typography, { variant: "h6", className: classes.title, children: [
        /* @__PURE__ */ jsx(Link, { to: "#", "data-testid": "go-back-link", onClick: () => navigate(-1), children: t("errorPage.goBack") }),
        "... or please",
        " ",
        /* @__PURE__ */ jsx(Link, { to: supportUrl || support.url, children: "contact support" }),
        " if you think this is a bug."
      ] }),
      stack && /* @__PURE__ */ jsx(StackDetails, { stack })
    ] }),
    /* @__PURE__ */ jsx(MicDrop, {})
  ] });
}

export { ErrorPage };
