// @backstage/plugin-scaffolder/dist/alpha/components/TemplateListPage/RegisterExistingButton.esm.js
import { jsx } from 'react/jsx-runtime';
import Button from '@material-ui/core/Button';
import IconButton from '@material-ui/core/IconButton';
import useMediaQuery from '@material-ui/core/useMediaQuery';
import { Link } from 'react-router-dom';
import CreateComponentIcon from '@material-ui/icons/AddCircleOutline';
import { catalogEntityCreatePermission } from '@backstage/plugin-catalog-common/alpha';
import { usePermission } from '@backstage/plugin-permission-react';

const RegisterExistingButton = (props) => {
  const { title, to } = props;
  const { allowed } = usePermission({
    permission: catalogEntityCreatePermission
  });
  const isXSScreen = useMediaQuery(
    (theme) => theme.breakpoints.down("xs")
  );
  if (!to || !allowed) {
    return null;
  }
  return isXSScreen ? /* @__PURE__ */ jsx(
    IconButton,
    {
      component: Link,
      color: "primary",
      title,
      size: "small",
      to,
      children: /* @__PURE__ */ jsx(CreateComponentIcon, {})
    }
  ) : /* @__PURE__ */ jsx(Button, { component: Link, variant: "outlined", color: "primary", to, children: title });
};

export { RegisterExistingButton };
