// @backstage/cli/dist/cjs/index-Cw0lUK20.cjs.js
'use strict';

var createDistWorkspace = require('./createDistWorkspace-DxOWPD6z.cjs.js');
var role = require('./role-8b0z7P0n.cjs.js');
var cliNode = require('@backstage/cli-node');
var index = require('./index-DmUbBCFk.cjs.js');
var buildBackend = require('./buildBackend-CmtimF8a.cjs.js');
var chalk = require('chalk');
require('fs-extra');
require('path');
require('p-limit');
require('os');
require('tar');
require('lodash/partition');
require('./run-CSt1n0F1.cjs.js');
require('child_process');
require('util');
require('@backstage/errors');
require('rollup');
require('@rollup/plugin-commonjs');
require('@rollup/plugin-node-resolve');
require('rollup-plugin-postcss');
require('rollup-plugin-esbuild');
require('@svgr/rollup');
require('rollup-plugin-dts');
require('@rollup/plugin-json');
require('@rollup/plugin-yaml');
require('rollup-pluginutils');
require('./svgrTemplate-BTjBQ3by.cjs.js');
require('./entryPoints-CoHH4lBA.cjs.js');
require('./parallel-BszNaKyc.cjs.js');
require('worker_threads');
require('./productionPack-BWU8WkGs.cjs.js');
require('npm-packlist');
require('commander');
require('semver');
require('@backstage/cli-common');
require('webpack');
require('eslint-webpack-plugin');
require('fork-ts-checker-webpack-plugin');
require('html-webpack-plugin');
require('@module-federation/enhanced/webpack');
require('react-dev-utils/ModuleScopePlugin');
require('run-script-webpack-plugin');
require('@pmmmwh/react-refresh-webpack-plugin');
require('@manypkg/get-packages');
require('webpack-node-externals');
require('./moduleFederation-8XXecxLD.cjs.js');
require('lodash/pickBy');
require('mini-css-extract-plugin');
require('yn');
require('@backstage/config');
require('chokidar');
require('p-queue');
require('react-dev-utils/FileSizeReporter');
require('react-dev-utils/formatWebpackMessages');
require('react-dev-utils/openBrowser');
require('webpack-dev-server');
require('./config-BMsjTUVf.cjs.js');
require('@backstage/config-loader');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var chalk__default = /*#__PURE__*/_interopDefaultCompat(chalk);

function isValidUrl(url) {
  try {
    new URL(url);
    return true;
  } catch {
    return false;
  }
}

async function command(opts) {
  const role$1 = await role.findRoleFromCommand(opts);
  if (role$1 === "frontend" || role$1 === "backend") {
    const configPaths = opts.config.map((arg) => {
      if (isValidUrl(arg)) {
        return arg;
      }
      return index.paths.resolveTarget(arg);
    });
    if (role$1 === "frontend") {
      return buildBackend.buildFrontend({
        targetDir: index.paths.targetDir,
        configPaths,
        writeStats: Boolean(opts.stats)
      });
    }
    return buildBackend.buildBackend({
      targetDir: index.paths.targetDir,
      configPaths,
      skipBuildDependencies: Boolean(opts.skipBuildDependencies),
      minify: Boolean(opts.minify)
    });
  }
  if (role$1 === "frontend-dynamic-container") {
    console.log(
      chalk__default.default.yellow(
        `\u26A0\uFE0F  WARNING: The 'frontend-dynamic-container' package role is experimental and will receive immediate breaking changes in the future.`
      )
    );
    return buildBackend.buildFrontend({
      targetDir: index.paths.targetDir,
      configPaths: [],
      writeStats: Boolean(opts.stats),
      isModuleFederationRemote: true
    });
  }
  const roleInfo = cliNode.PackageRoles.getRoleInfo(role$1);
  const outputs = /* @__PURE__ */ new Set();
  if (roleInfo.output.includes("cjs")) {
    outputs.add(createDistWorkspace.Output.cjs);
  }
  if (roleInfo.output.includes("esm")) {
    outputs.add(createDistWorkspace.Output.esm);
  }
  if (roleInfo.output.includes("types")) {
    outputs.add(createDistWorkspace.Output.types);
  }
  return createDistWorkspace.buildPackage({
    outputs,
    minify: Boolean(opts.minify),
    workspacePackages: await cliNode.PackageGraph.listTargetPackages()
  });
}

exports.command = command;
