// @backstage/plugin-auth-react/dist/components/CookieAuthRedirect/CookieAuthRedirect.esm.js
import { jsxs, Fragment, jsx } from 'react/jsx-runtime';
import { useApi, identityApiRef } from '@backstage/core-plugin-api';
import { useAsync, useMountEffect } from '@react-hookz/web';

function CookieAuthRedirect() {
  const identityApi = useApi(identityApiRef);
  const [state, actions] = useAsync(async () => {
    const { token } = await identityApi.getCredentials();
    if (!token) {
      throw new Error("Expected Backstage token in sign-in response");
    }
    return token;
  });
  useMountEffect(actions.execute);
  if (state.status === "error" && state.error) {
    return /* @__PURE__ */ jsxs(Fragment, { children: [
      "An error occurred: ",
      state.error.message
    ] });
  }
  if (state.status === "success" && state.result) {
    return /* @__PURE__ */ jsxs(
      "form",
      {
        ref: (form) => form?.submit(),
        action: window.location.href,
        method: "POST",
        style: { visibility: "hidden" },
        children: [
          /* @__PURE__ */ jsx("input", { type: "hidden", name: "type", value: "sign-in" }),
          /* @__PURE__ */ jsx("input", { type: "hidden", name: "token", value: state.result }),
          /* @__PURE__ */ jsx("input", { type: "submit", value: "Continue" })
        ]
      }
    );
  }
  return null;
}

export { CookieAuthRedirect };
