// @backstage/backend-defaults/dist/entrypoints/urlReader/lib/tree/ReadTreeResponseFactory.cjs.js
'use strict';

var os = require('os');
var TarArchiveResponse = require('./TarArchiveResponse.cjs.js');
var ZipArchiveResponse = require('./ZipArchiveResponse.cjs.js');
var ReadableArrayResponse = require('./ReadableArrayResponse.cjs.js');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var os__default = /*#__PURE__*/_interopDefaultCompat(os);

class DefaultReadTreeResponseFactory {
  constructor(workDir) {
    this.workDir = workDir;
  }
  static create(options) {
    return new DefaultReadTreeResponseFactory(
      options.config.getOptionalString("backend.workingDirectory") ?? os__default.default.tmpdir()
    );
  }
  async fromTarArchive(options) {
    return new TarArchiveResponse.TarArchiveResponse(
      options.stream,
      options.subpath ?? "",
      this.workDir,
      options.etag,
      options.filter,
      options.stripFirstDirectory ?? true
    );
  }
  async fromZipArchive(options) {
    return new ZipArchiveResponse.ZipArchiveResponse(
      options.stream,
      options.subpath ?? "",
      this.workDir,
      options.etag,
      options.filter
    );
  }
  async fromReadableArray(options) {
    return new ReadableArrayResponse.ReadableArrayResponse(options, this.workDir, "");
  }
}

exports.DefaultReadTreeResponseFactory = DefaultReadTreeResponseFactory;
