// @backstage/plugin-scaffolder-backend/dist/service/alpha.cjs.js
'use strict';

var alpha = require('@backstage/plugin-scaffolder-common/alpha');
var pluginPermissionNode = require('@backstage/plugin-permission-node');
var rules = require('./rules.cjs.js');

const templateConditionExports = pluginPermissionNode.createConditionExports({
  pluginId: "scaffolder",
  resourceType: alpha.RESOURCE_TYPE_SCAFFOLDER_TEMPLATE,
  rules: rules.scaffolderTemplateRules
});
const actionsConditionExports = pluginPermissionNode.createConditionExports({
  pluginId: "scaffolder",
  resourceType: alpha.RESOURCE_TYPE_SCAFFOLDER_ACTION,
  rules: rules.scaffolderActionRules
});
const taskConditionExports = pluginPermissionNode.createConditionExports({
  pluginId: "scaffolder",
  resourceType: alpha.RESOURCE_TYPE_SCAFFOLDER_TASK,
  rules: rules.scaffolderTaskRules
});
const createScaffolderTemplateConditionalDecision = templateConditionExports.createConditionalDecision;
const scaffolderTemplateConditions = templateConditionExports.conditions;
const createScaffolderActionConditionalDecision = actionsConditionExports.createConditionalDecision;
const scaffolderActionConditions = actionsConditionExports.conditions;
const createScaffolderTaskConditionalDecision = taskConditionExports.createConditionalDecision;
const scaffolderTaskConditions = taskConditionExports.conditions;

exports.createScaffolderActionConditionalDecision = createScaffolderActionConditionalDecision;
exports.createScaffolderTaskConditionalDecision = createScaffolderTaskConditionalDecision;
exports.createScaffolderTemplateConditionalDecision = createScaffolderTemplateConditionalDecision;
exports.scaffolderActionConditions = scaffolderActionConditions;
exports.scaffolderTaskConditions = scaffolderTaskConditions;
exports.scaffolderTemplateConditions = scaffolderTemplateConditions;
