// @backstage/plugin-app/dist/extensions/AppRoutes.esm.js
import { jsx } from 'react/jsx-runtime';
import { createExtension, coreExtensionData, createExtensionInput, NotFoundErrorPage } from '@backstage/frontend-plugin-api';
import { useRoutes } from 'react-router-dom';

const AppRoutes = createExtension({
  name: "routes",
  attachTo: { id: "app/layout", input: "content" },
  inputs: {
    routes: createExtensionInput([
      coreExtensionData.routePath,
      coreExtensionData.routeRef.optional(),
      coreExtensionData.reactElement
    ])
  },
  output: [coreExtensionData.reactElement],
  factory({ inputs }) {
    const Routes = () => {
      const element = useRoutes([
        ...inputs.routes.map((route) => ({
          path: `${route.get(coreExtensionData.routePath).replace(/\/$/, "")}/*`,
          element: route.get(coreExtensionData.reactElement)
        })),
        {
          path: "*",
          element: /* @__PURE__ */ jsx(NotFoundErrorPage, {})
        }
      ]);
      return element;
    };
    return [coreExtensionData.reactElement(/* @__PURE__ */ jsx(Routes, {}))];
  }
});

export { AppRoutes };
