// @backstage/plugin-app/dist/extensions/components.esm.js
import { jsx } from 'react/jsx-runtime';
import { SwappableComponentBlueprint, Progress as Progress$2, NotFoundErrorPage as NotFoundErrorPage$1, ErrorDisplay as ErrorDisplay$1 } from '@backstage/frontend-plugin-api';
import { Progress as Progress$1, ErrorPage, ErrorPanel } from '@backstage/core-components';
import Button from '@material-ui/core/Button';

const Progress = SwappableComponentBlueprint.make({
  name: "core-progress",
  params: (define) => define({
    component: Progress$2,
    loader: () => Progress$1
  })
});
const NotFoundErrorPage = SwappableComponentBlueprint.make({
  name: "core-not-found-error-page",
  params: (define) => define({
    component: NotFoundErrorPage$1,
    loader: () => () => /* @__PURE__ */ jsx(ErrorPage, { status: "404", statusMessage: "PAGE NOT FOUND" })
  })
});
const ErrorDisplay = SwappableComponentBlueprint.make({
  name: "core-error-display",
  params: (define) => define({
    component: ErrorDisplay$1,
    loader: () => (props) => {
      const { plugin, error, resetError } = props;
      const title = `Error in ${plugin?.id}`;
      return /* @__PURE__ */ jsx(ErrorPanel, { title, error, defaultExpanded: true, children: /* @__PURE__ */ jsx(Button, { variant: "outlined", onClick: resetError, children: "Retry" }) });
    }
  })
});

export { ErrorDisplay, NotFoundErrorPage, Progress };
