// @backstage/plugin-app/dist/extensions/TranslationsApi.esm.js
import { ApiBlueprint, createExtensionInput, TranslationBlueprint } from '@backstage/frontend-plugin-api';
import { appLanguageApiRef, translationApiRef } from '@backstage/core-plugin-api/alpha';
import { I18nextTranslationApi } from '../packages/core-app-api/src/apis/implementations/TranslationApi/I18nextTranslationApi.esm.js';

const TranslationsApi = ApiBlueprint.makeWithOverrides({
  name: "translations",
  inputs: {
    translations: createExtensionInput(
      [TranslationBlueprint.dataRefs.translation],
      { replaces: [{ id: "app", input: "translations" }] }
    )
  },
  factory: (originalFactory, { inputs }) => {
    return originalFactory(
      (defineParams) => defineParams({
        api: translationApiRef,
        deps: { languageApi: appLanguageApiRef },
        factory: ({ languageApi }) => I18nextTranslationApi.create({
          languageApi,
          resources: inputs.translations.map(
            (i) => i.get(TranslationBlueprint.dataRefs.translation)
          )
        })
      })
    );
  }
});

export { TranslationsApi };
