// @backstage/plugin-catalog-graph/dist/components/CatalogGraphPage/CurveFilter.esm.js
import { jsx } from 'react/jsx-runtime';
import { Select } from '@backstage/core-components';
import Box from '@material-ui/core/Box';
import { useCallback } from 'react';
import { catalogGraphTranslationRef } from '../../translation.esm.js';
import { useTranslationRef } from '@backstage/frontend-plugin-api';

const curves = ["curveMonotoneX", "curveStepBefore"];
const CurveFilter = ({ value, onChange }) => {
  const { t } = useTranslationRef(catalogGraphTranslationRef);
  const CURVE_DISPLAY_NAMES = {
    curveMonotoneX: t("catalogGraphPage.curveFilter.curveMonotoneX"),
    curveStepBefore: t("catalogGraphPage.curveFilter.curveStepBefore")
  };
  const handleChange = useCallback(
    (v) => onChange(v),
    [onChange]
  );
  return /* @__PURE__ */ jsx(Box, { pb: 1, pt: 1, children: /* @__PURE__ */ jsx(
    Select,
    {
      label: t("catalogGraphPage.curveFilter.title"),
      selected: value,
      items: curves.map((v) => ({
        label: CURVE_DISPLAY_NAMES[v],
        value: v
      })),
      onChange: handleChange
    }
  ) });
};

export { CurveFilter };
