// @backstage/plugin-catalog-graph/dist/components/CatalogGraphPage/SwitchFilter.esm.js
import { jsx } from 'react/jsx-runtime';
import Box from '@material-ui/core/Box';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import Switch from '@material-ui/core/Switch';
import { makeStyles } from '@material-ui/core/styles';
import { useCallback } from 'react';

const useStyles = makeStyles(
  {
    root: {
      width: "100%",
      maxWidth: 300
    }
  },
  { name: "PluginCatalogGraphSwitchFilter" }
);
const SwitchFilter = ({ label, value, onChange }) => {
  const classes = useStyles();
  const handleChange = useCallback(
    (event) => {
      onChange(event.target.checked);
    },
    [onChange]
  );
  return /* @__PURE__ */ jsx(Box, { pb: 1, pt: 1, children: /* @__PURE__ */ jsx(
    FormControlLabel,
    {
      control: /* @__PURE__ */ jsx(
        Switch,
        {
          checked: value,
          onChange: handleChange,
          name: label,
          color: "primary"
        }
      ),
      label,
      className: classes.root
    }
  ) });
};

export { SwitchFilter };
