// @backstage/core-components/dist/layout/BottomLink/BottomLink.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import Box from '@material-ui/core/Box';
import Divider from '@material-ui/core/Divider';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import ArrowIcon from '@material-ui/icons/ArrowForward';
import { Link } from '../../components/Link/Link.esm.js';

const useStyles = makeStyles(
  (theme) => ({
    root: {
      maxWidth: "fit-content",
      padding: theme.spacing(2, 2, 2, 2.5)
    },
    boxTitle: {
      margin: 0,
      color: theme.palette.textSubtle
    },
    arrow: {
      color: theme.palette.textSubtle
    }
  }),
  { name: "BackstageBottomLink" }
);
function BottomLink(props) {
  const { link, title, onClick } = props;
  const classes = useStyles();
  return /* @__PURE__ */ jsxs(Box, { children: [
    /* @__PURE__ */ jsx(Divider, {}),
    /* @__PURE__ */ jsx(Link, { to: link, onClick, underline: "none", children: /* @__PURE__ */ jsxs(Box, { display: "flex", alignItems: "center", className: classes.root, children: [
      /* @__PURE__ */ jsx(Box, { className: classes.boxTitle, fontWeight: "fontWeightBold", m: 1, children: /* @__PURE__ */ jsx(Typography, { children: /* @__PURE__ */ jsx("strong", { children: title }) }) }),
      /* @__PURE__ */ jsx(ArrowIcon, { className: classes.arrow })
    ] }) })
  ] });
}

export { BottomLink };
