// @backstage/plugin-catalog-import/dist/components/ImportInfoCard/ImportInfoCard.esm.js
import { jsxs, jsx, Fragment } from 'react/jsx-runtime';
import { InfoCard } from '@backstage/core-components';
import { useApi, configApiRef } from '@backstage/core-plugin-api';
import { useTranslationRef } from '@backstage/frontend-plugin-api';
import Chip from '@material-ui/core/Chip';
import Typography from '@material-ui/core/Typography';
import { catalogImportApiRef } from '../../api/CatalogImportApi.esm.js';
import 'yaml';
import 'git-url-parse';
import '@octokit/rest';
import 'js-base64';
import { useCatalogFilename } from '../../hooks/useCatalogFilename.esm.js';
import { catalogImportTranslationRef } from '../../translation.esm.js';

const ImportInfoCard = (props) => {
  const {
    exampleLocationUrl = "https://github.com/backstage/backstage/blob/master/catalog-info.yaml",
    exampleRepositoryUrl = "https://github.com/backstage/backstage"
  } = props;
  const { t } = useTranslationRef(catalogImportTranslationRef);
  const configApi = useApi(configApiRef);
  const appTitle = configApi.getOptionalString("app.title") || "Backstage";
  const catalogImportApi = useApi(catalogImportApiRef);
  const hasGithubIntegration = configApi.has("integrations.github");
  const catalogFilename = useCatalogFilename();
  return /* @__PURE__ */ jsxs(
    InfoCard,
    {
      title: t("importInfoCard.title"),
      titleTypographyProps: { component: "h3" },
      deepLink: {
        title: t("importInfoCard.deepLinkTitle"),
        link: "https://backstage.io/docs/features/software-catalog/"
      },
      children: [
        /* @__PURE__ */ jsx(Typography, { variant: "body2", paragraph: true, children: t("importInfoCard.linkDescription", { appTitle }) }),
        /* @__PURE__ */ jsx(Typography, { component: "h4", variant: "h6", children: t("importInfoCard.fileLinkTitle") }),
        /* @__PURE__ */ jsxs(Typography, { variant: "subtitle2", color: "textSecondary", paragraph: true, children: [
          t("importInfoCard.examplePrefix"),
          /* @__PURE__ */ jsx("code", { children: exampleLocationUrl })
        ] }),
        /* @__PURE__ */ jsx(Typography, { variant: "body2", paragraph: true, children: t("importInfoCard.fileLinkDescription", { appTitle }) }),
        hasGithubIntegration && /* @__PURE__ */ jsxs(Fragment, { children: [
          /* @__PURE__ */ jsxs(Typography, { component: "h4", variant: "h6", children: [
            t("importInfoCard.githubIntegration.title"),
            /* @__PURE__ */ jsx(
              Chip,
              {
                label: t("importInfoCard.githubIntegration.label"),
                variant: "outlined",
                size: "small",
                style: { marginLeft: 8, marginBottom: 0 }
              }
            )
          ] }),
          /* @__PURE__ */ jsxs(Typography, { variant: "subtitle2", color: "textSecondary", paragraph: true, children: [
            t("importInfoCard.examplePrefix"),
            /* @__PURE__ */ jsx("code", { children: exampleRepositoryUrl })
          ] }),
          /* @__PURE__ */ jsx(Typography, { variant: "body2", paragraph: true, children: t("importInfoCard.exampleDescription", {
            catalogFilename: /* @__PURE__ */ jsx("code", { children: catalogFilename }),
            appTitle
          }) }),
          catalogImportApi.preparePullRequest && /* @__PURE__ */ jsx(Typography, { variant: "body2", paragraph: true, children: t("importInfoCard.preparePullRequestDescription", {
            catalogFilename: /* @__PURE__ */ jsx("code", { children: catalogFilename }),
            appTitle
          }) })
        ] })
      ]
    }
  );
};

export { ImportInfoCard };
