// @backstage/repo-tools/dist/commands/knip-reports/knip-reports.cjs.js
'use strict';

var knipExtractor = require('./knip-extractor.cjs.js');
var paths = require('../../lib/paths.cjs.js');

const buildKnipReports = async (paths$1 = [], opts) => {
  const isCiBuild = opts.ci;
  const isAllPackages = !paths$1?.length;
  const selectedPackageDirs = await paths.resolvePackagePaths({
    paths: paths$1,
    include: opts.include,
    exclude: opts.exclude
  });
  if (isAllPackages && !isCiBuild) {
    console.log("");
    console.log(
      "TIP: You can generate knip-reports for select packages by passing package paths:"
    );
    console.log("");
    console.log(
      "       yarn build:knip-reports packages/config packages/core-plugin-api plugins/*"
    );
    console.log("");
  }
  if (selectedPackageDirs.length > 0) {
    console.log("# Generating package knip reports");
    try {
      await knipExtractor.runKnipReports({
        packageDirs: selectedPackageDirs,
        isLocalBuild: !isCiBuild
      });
    } catch (e) {
      process.exit(1);
    }
  }
};

exports.buildKnipReports = buildKnipReports;
