// @backstage-community/plugin-rbac/dist/components/CreateRole/AddedMembersTableColumn.esm.js
import React from 'react';
import { parseEntityRef } from '@backstage/catalog-model';
import { Link } from '@backstage/core-components';
import Delete from '@mui/icons-material/Delete';
import Box from '@mui/material/Box';
import IconButton from '@mui/material/IconButton';

const reviewStepMemebersTableColumns = () => [
  {
    title: "Name",
    field: "label",
    type: "string"
  },
  {
    title: "Type",
    field: "type",
    type: "string"
  },
  {
    title: "Members",
    field: "members",
    type: "numeric",
    align: "left",
    render: (mem) => {
      if (mem || mem === 0) return mem;
      return "-";
    }
  }
];
const selectedMembersColumns = (selectedMembers, setFieldValue) => {
  const onRemove = (etag) => {
    const updatedMembers = selectedMembers.filter(
      (mem) => mem.etag !== etag
    );
    setFieldValue("selectedMembers", updatedMembers);
  };
  return [
    {
      title: "Name",
      field: "label",
      type: "string",
      render: (props) => {
        const { kind, namespace, name } = parseEntityRef(props.ref);
        return /* @__PURE__ */ React.createElement(Link, { to: `/catalog/${namespace}/${kind}/${name}`, target: "blank" }, props.label);
      }
    },
    {
      title: "Type",
      field: "type",
      type: "string"
    },
    {
      title: "Members",
      field: "members",
      type: "numeric",
      align: "left",
      emptyValue: "-"
    },
    {
      title: "Actions",
      sorting: false,
      render: (mem) => {
        return /* @__PURE__ */ React.createElement(Box, { key: mem.etag }, /* @__PURE__ */ React.createElement(
          IconButton,
          {
            onClick: () => onRemove(mem.etag),
            "aria-label": "Remove",
            title: "Remove member",
            style: { padding: "0.5rem", borderRadius: "50%" },
            sx: { "&:hover": { borderRadius: "50%" } }
          },
          /* @__PURE__ */ React.createElement(Delete, null)
        ));
      }
    }
  ];
};

export { reviewStepMemebersTableColumns, selectedMembersColumns };
