// @backstage-community/plugin-rbac/dist/components/CreateRole/MembersDropdownOption.esm.js
import React from 'react';
import Typography from '@mui/material/Typography';
import match from 'autosuggest-highlight/match';
import parse from 'autosuggest-highlight/parse';

const MembersDropdownOption = ({
  props,
  option,
  state
}) => {
  const { inputValue } = state;
  const { label, etag } = option;
  const matches = match(label, inputValue, { insideWords: true });
  const parts = parse(label, matches);
  return /* @__PURE__ */ React.createElement(
    "li",
    {
      ...props,
      key: `${etag}`,
      style: {
        display: "flex",
        flexDirection: "column",
        alignItems: "flex-start"
      }
    },
    /* @__PURE__ */ React.createElement(Typography, { component: "span" }, parts.map((part) => /* @__PURE__ */ React.createElement(
      Typography,
      {
        key: `${part.text}-${etag}`,
        component: "span",
        sx: {
          fontWeight: part.highlight ? 400 : 700,
          color: (theme) => theme.palette.text.primary
        },
        "data-testid": option.label
      },
      part.text
    ))),
    /* @__PURE__ */ React.createElement(Typography, { sx: { color: (theme) => theme.palette.text.secondary } }, option.description),
    " "
  );
};

export { MembersDropdownOption };
