// @backstage-community/plugin-rbac/dist/components/CreateRole/PoliciesCheckboxGroup.esm.js
import React from 'react';
import Checkbox from '@mui/material/Checkbox';
import FormControl from '@mui/material/FormControl';
import FormControlLabel from '@mui/material/FormControlLabel';
import FormGroup from '@mui/material/FormGroup';
import FormLabel from '@mui/material/FormLabel';

const PoliciesCheckboxGroup = ({
  permissionPoliciesRowData,
  rowName,
  onChangePolicy
}) => {
  return /* @__PURE__ */ React.createElement(
    FormControl,
    {
      required: true,
      style: {
        justifyContent: "flex-start",
        gap: "1px",
        width: "402px",
        flexGrow: "1",
        marginBottom: "25px"
      }
    },
    /* @__PURE__ */ React.createElement(
      FormLabel,
      {
        style: {
          fontWeight: 800,
          fontSize: "0.8rem"
        }
      },
      "What actions they can do?"
    ),
    /* @__PURE__ */ React.createElement(
      FormGroup,
      {
        style: {
          display: "flex",
          gap: "7px",
          flexDirection: "row",
          paddingLeft: "9px"
        }
      },
      permissionPoliciesRowData.policies.map(
        (p, index, self) => {
          const labelCheckedArray = self.filter(
            (val) => val.effect === "allow"
          );
          const labelCheckedCount = labelCheckedArray.length;
          return /* @__PURE__ */ React.createElement(
            FormControlLabel,
            {
              key: p.policy,
              disabled: !(permissionPoliciesRowData.plugin && permissionPoliciesRowData.permission) || permissionPoliciesRowData.policies.length === 1 || labelCheckedCount === 1 && labelCheckedArray[0].policy === p.policy,
              label: p.policy,
              name: `${rowName}.policies[${index}].policy`,
              control: /* @__PURE__ */ React.createElement(
                Checkbox,
                {
                  checked: p.effect === "allow",
                  name: `${rowName}.policies[${index}].policy-${p.policy}`,
                  onChange: (e) => onChangePolicy(e.target.checked, index),
                  color: "primary"
                }
              )
            }
          );
        }
      )
    )
  );
};

export { PoliciesCheckboxGroup };
