// @backstage-community/plugin-rbac/dist/components/CreateRole/ReviewStep.esm.js
import React from 'react';
import { StructuredMetadataTable } from '@backstage/core-components';
import Typography from '@mui/material/Typography';
import { getPermissionsNumber } from '../../utils/create-role-utils.esm.js';
import { getMembers } from '../../utils/rbac-utils.esm.js';
import { reviewStepMemebersTableColumns } from './AddedMembersTableColumn.esm.js';
import { ReviewStepTable } from './ReviewStepTable.esm.js';
import { selectedPermissionPoliciesColumn } from './SelectedPermissionPoliciesColumn.esm.js';

const tableMetadata = (values) => {
  const membersKey = values.selectedMembers.length > 0 ? `Users and groups (${getMembers(values.selectedMembers)})` : "Users and groups";
  const permissionPoliciesKey = `Permission policies (${getPermissionsNumber(
    values
  )})`;
  return {
    "Name and description of role": /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(Typography, { sx: { margin: "0px" } }, values.name), /* @__PURE__ */ React.createElement("br", null), /* @__PURE__ */ React.createElement(Typography, { sx: { margin: "0px" } }, values.description)),
    [membersKey]: /* @__PURE__ */ React.createElement(
      ReviewStepTable,
      {
        rows: values.selectedMembers,
        columns: reviewStepMemebersTableColumns(),
        tableWrapperWidth: 550
      }
    ),
    [permissionPoliciesKey]: /* @__PURE__ */ React.createElement(
      ReviewStepTable,
      {
        rows: values.permissionPoliciesRows,
        columns: selectedPermissionPoliciesColumn(),
        tableWrapperWidth: 700
      }
    )
  };
};
const ReviewStep = ({
  values,
  isEditing
}) => {
  return /* @__PURE__ */ React.createElement("div", { style: { overflow: "auto" } }, /* @__PURE__ */ React.createElement(Typography, { variant: "h6" }, isEditing ? "Review and save" : "Review and create"), /* @__PURE__ */ React.createElement(
    StructuredMetadataTable,
    {
      dense: true,
      metadata: tableMetadata(values),
      options: { titleFormat: (key) => key }
    }
  ));
};

export { ReviewStep };
