// @backstage/plugin-user-settings/dist/components/FeatureFlags/EmptyFlags.esm.js
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import { EmptyState, CodeSnippet } from '@backstage/core-components';
import { useTranslationRef } from '@backstage/frontend-plugin-api';
import { userSettingsTranslationRef } from '../../translation.esm.js';

const EXAMPLE = `import { createPlugin } from '@backstage/core-plugin-api';

export default createPlugin({
  id: 'plugin-name',
  featureFlags: [{ name: 'enable-example-feature' }],
});
`;
const EmptyFlags = () => {
  const { t } = useTranslationRef(userSettingsTranslationRef);
  return /* @__PURE__ */ jsx(
    EmptyState,
    {
      missing: "content",
      title: t("featureFlags.emptyFlags.title"),
      description: t("featureFlags.emptyFlags.description"),
      action: /* @__PURE__ */ jsxs(Fragment, { children: [
        /* @__PURE__ */ jsx(Typography, { variant: "body1", children: t("featureFlags.emptyFlags.action.title") }),
        /* @__PURE__ */ jsx(
          CodeSnippet,
          {
            text: EXAMPLE,
            language: "typescript",
            showLineNumbers: true,
            highlightedNumbers: [6],
            customStyle: { background: "inherit", fontSize: "115%" }
          }
        ),
        /* @__PURE__ */ jsx(
          Button,
          {
            variant: "contained",
            color: "primary",
            href: "https://backstage.io/docs/api/utility-apis",
            children: t("featureFlags.emptyFlags.action.readMoreButtonTitle")
          }
        )
      ] })
    }
  );
};

export { EmptyFlags };
