// @backstage/plugin-user-settings/dist/components/FeatureFlags/FeatureFlagsItem.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import Switch from '@material-ui/core/Switch';
import Tooltip from '@material-ui/core/Tooltip';
import { useTranslationRef } from '@backstage/frontend-plugin-api';
import { userSettingsTranslationRef } from '../../translation.esm.js';

const getSecondaryText = (flag, t) => {
  if (flag.description) {
    return flag.description;
  }
  return flag.pluginId ? t("featureFlags.flagItem.subtitle.registeredInPlugin", {
    pluginId: flag.pluginId
  }) : t("featureFlags.flagItem.subtitle.registeredInApplication");
};
const FlagItem = ({ flag, enabled, toggleHandler }) => {
  const { t } = useTranslationRef(userSettingsTranslationRef);
  return /* @__PURE__ */ jsxs(ListItem, { divider: true, button: true, onClick: () => toggleHandler(flag.name), children: [
    /* @__PURE__ */ jsx(ListItemIcon, { children: /* @__PURE__ */ jsx(
      Tooltip,
      {
        placement: "top",
        arrow: true,
        title: enabled ? t("featureFlags.flagItem.title.disable") : t("featureFlags.flagItem.title.enable"),
        children: /* @__PURE__ */ jsx(Switch, { color: "primary", checked: enabled, name: flag.name })
      }
    ) }),
    /* @__PURE__ */ jsx(ListItemText, { primary: flag.name, secondary: getSecondaryText(flag, t) })
  ] });
};

export { FlagItem };
