// @backstage/plugin-api-docs/dist/components/ApisCards/HasApisCard.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { RELATION_HAS_PART } from '@backstage/catalog-model';
import Typography from '@material-ui/core/Typography';
import { EntityTable, useEntity, useRelatedEntities } from '@backstage/plugin-catalog-react';
import { useMemo } from 'react';
import { createSpecApiTypeColumn } from './presets.esm.js';
import { InfoCard, Progress, WarningPanel, CodeSnippet, Link } from '@backstage/core-components';
import { useTranslationRef } from '@backstage/frontend-plugin-api';
import { apiDocsTranslationRef } from '../../translation.esm.js';

const HasApisCard = (props) => {
  const { t } = useTranslationRef(apiDocsTranslationRef);
  const presetColumns = useMemo(() => {
    return [
      EntityTable.columns.createEntityRefColumn({ defaultKind: "API" }),
      EntityTable.columns.createOwnerColumn(),
      createSpecApiTypeColumn(t),
      EntityTable.columns.createSpecLifecycleColumn(),
      EntityTable.columns.createMetadataDescriptionColumn()
    ];
  }, [t]);
  const {
    variant = "gridItem",
    title = t("hasApisCard.title"),
    columns = presetColumns,
    tableOptions = {}
  } = props;
  const { entity } = useEntity();
  const { entities, loading, error } = useRelatedEntities(entity, {
    type: RELATION_HAS_PART,
    kind: "API"
  });
  if (loading) {
    return /* @__PURE__ */ jsx(InfoCard, { variant, title, children: /* @__PURE__ */ jsx(Progress, {}) });
  }
  if (error || !entities) {
    return /* @__PURE__ */ jsx(InfoCard, { variant, title, children: /* @__PURE__ */ jsx(
      WarningPanel,
      {
        severity: "error",
        title: t("hasApisCard.error.title"),
        message: /* @__PURE__ */ jsx(CodeSnippet, { text: `${error}`, language: "text" })
      }
    ) });
  }
  return /* @__PURE__ */ jsx(
    EntityTable,
    {
      title,
      variant,
      emptyContent: /* @__PURE__ */ jsxs("div", { style: { textAlign: "center" }, children: [
        /* @__PURE__ */ jsx(Typography, { variant: "body1", children: t("hasApisCard.emptyContent.title", {
          entity: entity.kind.toLocaleLowerCase("en-US")
        }) }),
        /* @__PURE__ */ jsx(Typography, { variant: "body2", children: /* @__PURE__ */ jsx(Link, { to: "https://backstage.io/docs/features/software-catalog/descriptor-format#kind-api", children: t("apisCardHelpLinkTitle") }) })
      ] }),
      columns,
      tableOptions,
      entities
    }
  );
};

export { HasApisCard };
