// @backstage/plugin-scaffolder-react/dist/next/components/TaskSteps/TaskSteps.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import MuiStepper from '@material-ui/core/Stepper';
import MuiStep from '@material-ui/core/Step';
import MuiStepButton from '@material-ui/core/StepButton';
import MuiStepLabel from '@material-ui/core/StepLabel';
import Box from '@material-ui/core/Box';
import Paper from '@material-ui/core/Paper';
import { StepIcon } from './StepIcon.esm.js';
import { StepTime } from './StepTime.esm.js';
import { TaskBorder } from './TaskBorder.esm.js';

const TaskSteps = (props) => {
  return /* @__PURE__ */ jsxs(Paper, { style: { position: "relative", overflow: "hidden" }, children: [
    /* @__PURE__ */ jsx(
      TaskBorder,
      {
        isComplete: props.isComplete ?? false,
        isError: props.isError ?? false
      }
    ),
    /* @__PURE__ */ jsx(Box, { padding: 2, children: /* @__PURE__ */ jsx(
      MuiStepper,
      {
        activeStep: props.activeStep,
        alternativeLabel: true,
        variant: "elevation",
        style: { overflowX: "auto" },
        children: props.steps.map((step) => {
          const isCompleted = step.status === "completed";
          const isFailed = step.status === "failed";
          const isActive = step.status === "processing";
          const isSkipped = step.status === "skipped";
          const stepIconProps = {
            completed: isCompleted,
            error: isFailed,
            active: isActive,
            skipped: isSkipped
          };
          return /* @__PURE__ */ jsx(MuiStep, { children: /* @__PURE__ */ jsx(MuiStepButton, { children: /* @__PURE__ */ jsxs(
            MuiStepLabel,
            {
              StepIconProps: stepIconProps,
              StepIconComponent: StepIcon,
              "data-testid": "step-label",
              children: [
                /* @__PURE__ */ jsx(Box, { children: step.name }),
                !isSkipped && /* @__PURE__ */ jsx(StepTime, { step })
              ]
            }
          ) }) }, step.id);
        })
      }
    ) })
  ] });
};

export { TaskSteps };
