// @backstage/plugin-app/dist/packages/core-app-api/src/apis/implementations/auth/oauth2/OAuth2.esm.js
import { DefaultAuthConnector } from '../../../../lib/AuthConnector/DefaultAuthConnector.esm.js';
import { RefreshingAuthSessionManager } from '../../../../lib/AuthSessionManager/RefreshingAuthSessionManager.esm.js';
import 'zen-observable';
import '@backstage/core-plugin-api';

const DEFAULT_PROVIDER = {
  id: "oauth2",
  title: "Your Identity Provider",
  icon: () => null
};
class OAuth2 {
  static createAuthConnector(options) {
    if ("authConnector" in options) {
      return options.authConnector;
    }
    const {
      scopeTransform = (x) => x,
      configApi,
      discoveryApi,
      environment = "development",
      provider = DEFAULT_PROVIDER,
      oauthRequestApi,
      popupOptions
    } = options;
    return new DefaultAuthConnector({
      configApi,
      discoveryApi,
      environment,
      provider,
      oauthRequestApi,
      sessionTransform({
        backstageIdentity,
        ...res
      }) {
        const session = {
          ...res,
          providerInfo: {
            idToken: res.providerInfo.idToken,
            accessToken: res.providerInfo.accessToken,
            scopes: OAuth2.normalizeScopes(res.providerInfo.scope, {
              scopeTransform
            }),
            expiresAt: res.providerInfo.expiresInSeconds ? new Date(Date.now() + res.providerInfo.expiresInSeconds * 1e3) : void 0
          }
        };
        if (backstageIdentity) {
          session.backstageIdentity = {
            token: backstageIdentity.token,
            identity: backstageIdentity.identity,
            expiresAt: backstageIdentity.expiresInSeconds ? new Date(Date.now() + backstageIdentity.expiresInSeconds * 1e3) : void 0
          };
        }
        return session;
      },
      popupOptions
    });
  }
  static create(options) {
    const { defaultScopes = [], scopeTransform = (x) => x } = options;
    const connector = OAuth2.createAuthConnector(options);
    const sessionManager = new RefreshingAuthSessionManager({
      connector,
      defaultScopes: new Set(defaultScopes),
      sessionScopes: (session) => session.providerInfo.scopes,
      sessionShouldRefresh: (session) => {
        let min = Infinity;
        if (session.providerInfo?.expiresAt) {
          min = Math.min(
            min,
            (session.providerInfo.expiresAt.getTime() - Date.now()) / 1e3
          );
        }
        if (session.backstageIdentity?.expiresAt) {
          min = Math.min(
            min,
            (session.backstageIdentity.expiresAt.getTime() - Date.now()) / 1e3
          );
        }
        return min < 60 * 3;
      }
    });
    return new OAuth2({ sessionManager, scopeTransform });
  }
  sessionManager;
  scopeTransform;
  constructor(options) {
    this.sessionManager = options.sessionManager;
    this.scopeTransform = options.scopeTransform;
  }
  async signIn() {
    await this.getAccessToken();
  }
  async signOut() {
    await this.sessionManager.removeSession();
  }
  sessionState$() {
    return this.sessionManager.sessionState$();
  }
  async getAccessToken(scope, options) {
    const normalizedScopes = OAuth2.normalizeScopes(scope, {
      scopeTransform: this.scopeTransform
    });
    const session = await this.sessionManager.getSession({
      ...options,
      scopes: normalizedScopes
    });
    return session?.providerInfo.accessToken ?? "";
  }
  async getIdToken(options = {}) {
    const session = await this.sessionManager.getSession({
      ...options,
      scopes: /* @__PURE__ */ new Set(["openid"])
    });
    return session?.providerInfo.idToken ?? "";
  }
  async getBackstageIdentity(options = {}) {
    const session = await this.sessionManager.getSession(options);
    return session?.backstageIdentity;
  }
  async getProfile(options = {}) {
    const session = await this.sessionManager.getSession(options);
    return session?.profile;
  }
  /**
   * @public
   */
  static normalizeScopes(scopes, options) {
    if (!scopes) {
      return /* @__PURE__ */ new Set();
    }
    const scopeList = Array.isArray(scopes) ? scopes : scopes.split(/[\s|,]/).filter(Boolean);
    const transformedScopes = options ? options.scopeTransform(scopeList) : scopeList;
    return new Set(transformedScopes);
  }
}

export { OAuth2 as default };
