// @backstage/plugin-user-settings/dist/components/AuthProviders/DefaultProviderSettings.esm.js
import { jsxs, Fragment, jsx } from 'react/jsx-runtime';
import Star from '@material-ui/icons/Star';
import { ProviderSettingsItem } from './ProviderSettingsItem.esm.js';
import { googleAuthApiRef, microsoftAuthApiRef, githubAuthApiRef, gitlabAuthApiRef, oktaAuthApiRef, bitbucketAuthApiRef, oneloginAuthApiRef, atlassianAuthApiRef, bitbucketServerAuthApiRef, openshiftAuthApiRef } from '@backstage/core-plugin-api';
import { userSettingsTranslationRef } from '../../translation.esm.js';
import { useTranslationRef } from '@backstage/frontend-plugin-api';

const DefaultProviderSettings = (props) => {
  const { configuredProviders } = props;
  const { t } = useTranslationRef(userSettingsTranslationRef);
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    configuredProviders.includes("google") && /* @__PURE__ */ jsx(
      ProviderSettingsItem,
      {
        title: "Google",
        description: t("defaultProviderSettings.description", {
          provider: "Google"
        }),
        apiRef: googleAuthApiRef,
        icon: Star
      }
    ),
    configuredProviders.includes("microsoft") && /* @__PURE__ */ jsx(
      ProviderSettingsItem,
      {
        title: "Microsoft",
        description: t("defaultProviderSettings.description", {
          provider: "Microsoft"
        }),
        apiRef: microsoftAuthApiRef,
        icon: Star
      }
    ),
    configuredProviders.includes("github") && /* @__PURE__ */ jsx(
      ProviderSettingsItem,
      {
        title: "GitHub",
        description: t("defaultProviderSettings.description", {
          provider: "GitHub"
        }),
        apiRef: githubAuthApiRef,
        icon: Star
      }
    ),
    configuredProviders.includes("gitlab") && /* @__PURE__ */ jsx(
      ProviderSettingsItem,
      {
        title: "GitLab",
        description: t("defaultProviderSettings.description", {
          provider: "GitLab"
        }),
        apiRef: gitlabAuthApiRef,
        icon: Star
      }
    ),
    configuredProviders.includes("okta") && /* @__PURE__ */ jsx(
      ProviderSettingsItem,
      {
        title: "Okta",
        description: t("defaultProviderSettings.description", {
          provider: "Okta"
        }),
        apiRef: oktaAuthApiRef,
        icon: Star
      }
    ),
    configuredProviders.includes("bitbucket") && /* @__PURE__ */ jsx(
      ProviderSettingsItem,
      {
        title: "Bitbucket",
        description: t("defaultProviderSettings.description", {
          provider: "Bitbucket"
        }),
        apiRef: bitbucketAuthApiRef,
        icon: Star
      }
    ),
    configuredProviders.includes("onelogin") && /* @__PURE__ */ jsx(
      ProviderSettingsItem,
      {
        title: "OneLogin",
        description: t("defaultProviderSettings.description", {
          provider: "OneLogin"
        }),
        apiRef: oneloginAuthApiRef,
        icon: Star
      }
    ),
    configuredProviders.includes("atlassian") && /* @__PURE__ */ jsx(
      ProviderSettingsItem,
      {
        title: "Atlassian",
        description: t("defaultProviderSettings.description", {
          provider: "Atlassian"
        }),
        apiRef: atlassianAuthApiRef,
        icon: Star
      }
    ),
    configuredProviders.includes("bitbucketServer") && /* @__PURE__ */ jsx(
      ProviderSettingsItem,
      {
        title: "Bitbucket Server",
        description: t("defaultProviderSettings.description", {
          provider: "Bitbucket Server"
        }),
        apiRef: bitbucketServerAuthApiRef,
        icon: Star
      }
    ),
    configuredProviders.includes("openshift") && /* @__PURE__ */ jsx(
      ProviderSettingsItem,
      {
        title: "OpenShift",
        description: "Provides authentication towards OpenShift APIs and identities",
        apiRef: openshiftAuthApiRef,
        icon: Star
      }
    )
  ] });
};

export { DefaultProviderSettings };
