// @backstage/plugin-user-settings/dist/components/AuthProviders/EmptyProviders.esm.js
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import { EmptyState, CodeSnippet } from '@backstage/core-components';
import { useTranslationRef } from '@backstage/frontend-plugin-api';
import { userSettingsTranslationRef } from '../../translation.esm.js';

const EXAMPLE = `auth:
  providers:
    google:
      development:
        clientId: \${AUTH_GOOGLE_CLIENT_ID}
        clientSecret: \${AUTH_GOOGLE_CLIENT_SECRET}
`;
const EmptyProviders = () => {
  const { t } = useTranslationRef(userSettingsTranslationRef);
  return /* @__PURE__ */ jsx(
    EmptyState,
    {
      missing: "content",
      title: t("emptyProviders.title"),
      description: t("emptyProviders.description"),
      action: /* @__PURE__ */ jsxs(Fragment, { children: [
        /* @__PURE__ */ jsx(Typography, { variant: "body1", children: t("emptyProviders.action.title") }),
        /* @__PURE__ */ jsx(
          CodeSnippet,
          {
            text: EXAMPLE,
            language: "yaml",
            showLineNumbers: true,
            highlightedNumbers: [3, 4, 5, 6, 7, 8],
            customStyle: { background: "inherit", fontSize: "115%" }
          }
        ),
        /* @__PURE__ */ jsx(
          Button,
          {
            variant: "contained",
            color: "primary",
            href: "https://backstage.io/docs/auth/add-auth-provider",
            children: t("emptyProviders.action.readMoreButtonTitle")
          }
        )
      ] })
    }
  );
};

export { EmptyProviders };
