// @backstage/plugin-scaffolder/dist/alpha/components/TemplateEditorPage/DryRunResults/DryRunResultsSplitView.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { makeStyles } from '@material-ui/core/styles';
import Divider from '@material-ui/core/Divider';
import { Children } from 'react';
import classNames from 'classnames';

const useStyles = makeStyles((theme) => ({
  root: {
    display: "grid",
    gridTemplateColumns: "280px auto 3fr",
    gridTemplateRows: "1fr"
  },
  child: {
    overflowY: "auto",
    height: "100%",
    minHeight: 0
  },
  firstChild: {
    background: theme.palette.background.paper
  }
}));
function DryRunResultsSplitView(props) {
  const classes = useStyles();
  const childArray = Children.toArray(props.children);
  if (childArray.length !== 2) {
    throw new Error("must have exactly 2 children");
  }
  return /* @__PURE__ */ jsxs("div", { className: classes.root, children: [
    /* @__PURE__ */ jsx("div", { className: classNames(classes.child, classes.firstChild), children: childArray[0] }),
    /* @__PURE__ */ jsx(Divider, { orientation: "horizontal" }),
    /* @__PURE__ */ jsx("div", { className: classes.child, children: childArray[1] })
  ] });
}

export { DryRunResultsSplitView };
