// @backstage/plugin-scaffolder/dist/alpha/components/TemplateEditorPage/DryRunResults/DryRunResultsView.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { LogViewer } from '@backstage/core-components';
import { StreamLanguage } from '@codemirror/language';
import { yaml } from '@codemirror/legacy-modes/mode/yaml';
import Box from '@material-ui/core/Box';
import Divider from '@material-ui/core/Divider';
import { makeStyles } from '@material-ui/core/styles';
import Tab from '@material-ui/core/Tab';
import Tabs from '@material-ui/core/Tabs';
import CodeMirror from '@uiw/react-codemirror';
import { useState, useEffect, useMemo } from 'react';
import { useDryRun } from '../DryRunContext.esm.js';
import { DryRunResultsSplitView } from './DryRunResultsSplitView.esm.js';
import { FileBrowser } from '../../../../components/FileBrowser/FileBrowser.esm.js';
import { TaskPageLinks } from './TaskPageLinks.esm.js';
import { TaskStatusStepper } from './TaskStatusStepper.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { scaffolderTranslationRef } from '../../../../translation.esm.js';

const useStyles = makeStyles({
  root: {
    display: "flex",
    flexFlow: "column nowrap"
  },
  contentWrapper: {
    flex: 1,
    position: "relative"
  },
  content: {
    position: "absolute",
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    display: "flex",
    "& > *": {
      flex: 1
    }
  },
  codeMirror: {
    height: "100%",
    overflowY: "auto"
  }
});
function FilesContent() {
  const classes = useStyles();
  const { selectedResult } = useDryRun();
  const [selectedPath, setSelectedPath] = useState("");
  const selectedFile = selectedResult?.directoryContents.find(
    (f) => f.path === selectedPath
  );
  useEffect(() => {
    if (selectedResult) {
      const [firstFile] = selectedResult.directoryContents;
      if (firstFile) {
        setSelectedPath(firstFile.path);
      } else {
        setSelectedPath("");
      }
    }
    return void 0;
  }, [selectedResult]);
  if (!selectedResult) {
    return null;
  }
  return /* @__PURE__ */ jsxs(DryRunResultsSplitView, { children: [
    /* @__PURE__ */ jsx(
      FileBrowser,
      {
        selected: selectedPath,
        onSelect: setSelectedPath,
        filePaths: selectedResult.directoryContents.map((file) => file.path)
      }
    ),
    /* @__PURE__ */ jsx(
      CodeMirror,
      {
        className: classes.codeMirror,
        theme: "dark",
        height: "100%",
        extensions: [StreamLanguage.define(yaml)],
        readOnly: true,
        value: selectedFile?.base64Content ? atob(selectedFile.base64Content) : ""
      }
    )
  ] });
}
function LogContent() {
  const { selectedResult } = useDryRun();
  const [currentStepId, setUserSelectedStepId] = useState();
  const steps = useMemo(() => {
    if (!selectedResult) {
      return [];
    }
    return selectedResult.steps.map((step) => {
      const stepLog = selectedResult.log.filter(
        (l) => l.body.stepId === step.id
      );
      return {
        id: step.id,
        name: step.name,
        logString: stepLog.map((l) => l.body.message).join("\n"),
        status: stepLog[stepLog.length - 1]?.body.status ?? "completed"
      };
    }) ?? [];
  }, [selectedResult]);
  if (!selectedResult) {
    return null;
  }
  const selectedStep = steps.find((s) => s.id === currentStepId) ?? steps[0];
  return /* @__PURE__ */ jsxs(DryRunResultsSplitView, { children: [
    /* @__PURE__ */ jsx(
      TaskStatusStepper,
      {
        steps,
        currentStepId: selectedStep.id,
        onUserStepChange: setUserSelectedStepId
      }
    ),
    /* @__PURE__ */ jsx(LogViewer, { text: selectedStep?.logString ?? "" })
  ] });
}
function OutputContent() {
  const classes = useStyles();
  const { selectedResult } = useDryRun();
  if (!selectedResult) {
    return null;
  }
  return /* @__PURE__ */ jsxs(DryRunResultsSplitView, { children: [
    /* @__PURE__ */ jsx(Box, { pt: 2, children: selectedResult.output?.links?.length && /* @__PURE__ */ jsx(TaskPageLinks, { output: selectedResult.output }) }),
    /* @__PURE__ */ jsx(
      CodeMirror,
      {
        className: classes.codeMirror,
        theme: "dark",
        height: "100%",
        extensions: [StreamLanguage.define(yaml)],
        readOnly: true,
        value: JSON.stringify(selectedResult.output, null, 2)
      }
    )
  ] });
}
function DryRunResultsView() {
  const classes = useStyles();
  const [selectedTab, setSelectedTab] = useState(
    "files"
  );
  const { t } = useTranslationRef(scaffolderTranslationRef);
  return /* @__PURE__ */ jsxs("div", { className: classes.root, children: [
    /* @__PURE__ */ jsxs(Tabs, { value: selectedTab, onChange: (_, v) => setSelectedTab(v), children: [
      /* @__PURE__ */ jsx(
        Tab,
        {
          value: "files",
          label: t("templateEditorPage.dryRunResultsView.tab.files")
        }
      ),
      /* @__PURE__ */ jsx(
        Tab,
        {
          value: "log",
          label: t("templateEditorPage.dryRunResultsView.tab.log")
        }
      ),
      /* @__PURE__ */ jsx(
        Tab,
        {
          value: "output",
          label: t("templateEditorPage.dryRunResultsView.tab.output")
        }
      )
    ] }),
    /* @__PURE__ */ jsx(Divider, {}),
    /* @__PURE__ */ jsx("div", { className: classes.contentWrapper, children: /* @__PURE__ */ jsxs("div", { className: classes.content, children: [
      selectedTab === "files" && /* @__PURE__ */ jsx(FilesContent, {}),
      selectedTab === "log" && /* @__PURE__ */ jsx(LogContent, {}),
      selectedTab === "output" && /* @__PURE__ */ jsx(OutputContent, {})
    ] }) })
  ] });
}

export { DryRunResultsView };
