// @backstage/plugin-scaffolder/dist/alpha/components/TemplateEditorPage/DryRunResults/TaskStatusStepper.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { memo, useState } from 'react';
import { makeStyles, createStyles } from '@material-ui/core/styles';
import Step from '@material-ui/core/Step';
import StepLabel from '@material-ui/core/StepLabel';
import Stepper from '@material-ui/core/Stepper';
import StepButton from '@material-ui/core/StepButton';
import CircularProgress from '@material-ui/core/CircularProgress';
import Cancel from '@material-ui/icons/Cancel';
import CheckIcon from '@material-ui/icons/Check';
import FiberManualRecordIcon from '@material-ui/icons/FiberManualRecord';
import Typography from '@material-ui/core/Typography';
import { DateTime, Interval } from 'luxon';
import useInterval from 'react-use/esm/useInterval';
import humanizeDuration from 'humanize-duration';
import classNames from 'classnames';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { scaffolderTranslationRef } from '../../../../translation.esm.js';

const useStyles = makeStyles(
  (theme) => createStyles({
    root: {
      width: "100%"
    },
    button: {
      marginBottom: theme.spacing(2),
      marginLeft: theme.spacing(2)
    },
    actionsContainer: {
      marginBottom: theme.spacing(2)
    },
    resetContainer: {
      padding: theme.spacing(3)
    },
    labelWrapper: {
      display: "flex",
      flex: 1,
      flexDirection: "row",
      justifyContent: "space-between"
    },
    stepWrapper: {
      width: "100%"
    }
  })
);
const useStepIconStyles = makeStyles(
  (theme) => createStyles({
    root: {
      color: theme.palette.text.disabled,
      display: "flex",
      height: 22,
      alignItems: "center"
    },
    completed: {
      color: theme.palette.status.ok
    },
    error: {
      color: theme.palette.status.error
    }
  })
);
const StepTimeTicker = ({ step }) => {
  const [time, setTime] = useState("");
  useInterval(() => {
    if (!step.startedAt) {
      setTime("");
      return;
    }
    const end = step.endedAt ? DateTime.fromISO(step.endedAt) : DateTime.local();
    const startedAt = DateTime.fromISO(step.startedAt);
    const formatted = Interval.fromDateTimes(startedAt, end).toDuration().valueOf();
    setTime(humanizeDuration(formatted, { round: true }));
  }, 1e3);
  return /* @__PURE__ */ jsx(Typography, { variant: "caption", children: time });
};
function TaskStepIconComponent(props) {
  const classes = useStepIconStyles();
  const { active, completed, error } = props;
  const getMiddle = () => {
    if (active) {
      return /* @__PURE__ */ jsx(CircularProgress, { size: "24px" });
    }
    if (completed) {
      return /* @__PURE__ */ jsx(CheckIcon, {});
    }
    if (error) {
      return /* @__PURE__ */ jsx(Cancel, {});
    }
    return /* @__PURE__ */ jsx(FiberManualRecordIcon, {});
  };
  return /* @__PURE__ */ jsx(
    "div",
    {
      className: classNames(classes.root, {
        [classes.completed]: completed,
        [classes.error]: error
      }),
      children: getMiddle()
    }
  );
}
const TaskStatusStepper = memo(
  (props) => {
    const { steps, currentStepId, onUserStepChange } = props;
    const classes = useStyles(props);
    const { t } = useTranslationRef(scaffolderTranslationRef);
    return /* @__PURE__ */ jsx("div", { className: classes.root, children: /* @__PURE__ */ jsx(
      Stepper,
      {
        activeStep: steps.findIndex((s) => s.id === currentStepId),
        orientation: "vertical",
        nonLinear: true,
        children: steps.map((step, index) => {
          const isCancelled = step.status === "cancelled";
          const isActive = step.status === "processing";
          const isCompleted = step.status === "completed";
          const isFailed = step.status === "failed";
          const isSkipped = step.status === "skipped";
          return /* @__PURE__ */ jsx(Step, { expanded: true, children: /* @__PURE__ */ jsx(StepButton, { onClick: () => onUserStepChange(step.id), children: /* @__PURE__ */ jsx(
            StepLabel,
            {
              StepIconProps: {
                completed: isCompleted,
                error: isFailed || isCancelled,
                active: isActive
              },
              StepIconComponent: TaskStepIconComponent,
              className: classes.stepWrapper,
              children: /* @__PURE__ */ jsxs("div", { className: classes.labelWrapper, children: [
                /* @__PURE__ */ jsx(Typography, { variant: "subtitle2", children: step.name }),
                isSkipped ? /* @__PURE__ */ jsx(Typography, { variant: "caption", children: t(
                  "templateEditorPage.taskStatusStepper.skippedStepTitle"
                ) }) : /* @__PURE__ */ jsx(StepTimeTicker, { step })
              ] })
            }
          ) }) }, String(index));
        })
      }
    ) });
  }
);

export { TaskStatusStepper };
