// @backstage/plugin-techdocs/dist/search/components/TechDocsSearchResultListItem.esm.js
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import { Link } from '@backstage/core-components';
import { HighlightedSearchResultText } from '@backstage/plugin-search-react';

const useStyles = makeStyles({
  flexContainer: {
    flexWrap: "wrap"
  },
  itemText: {
    width: "100%",
    marginBottom: "1rem"
  }
});
const TechDocsSearchResultListItem = (props) => {
  const {
    result,
    highlight,
    lineClamp = 5,
    asListItem = true,
    asLink = true,
    title,
    icon
  } = props;
  const classes = useStyles();
  const LinkWrapper = ({ children }) => asLink ? /* @__PURE__ */ jsx(Link, { noTrack: true, to: result.location, children }) : /* @__PURE__ */ jsx(Fragment, { children });
  const TextItem = () => {
    const resultTitle = highlight?.fields.title ? /* @__PURE__ */ jsx(
      HighlightedSearchResultText,
      {
        text: highlight.fields.title,
        preTag: highlight.preTag,
        postTag: highlight.postTag
      }
    ) : result.title;
    const entityTitle = highlight?.fields.entityTitle ? /* @__PURE__ */ jsx(
      HighlightedSearchResultText,
      {
        text: highlight.fields.entityTitle,
        preTag: highlight.preTag,
        postTag: highlight.postTag
      }
    ) : result.entityTitle;
    const resultName = highlight?.fields.name ? /* @__PURE__ */ jsx(
      HighlightedSearchResultText,
      {
        text: highlight.fields.name,
        preTag: highlight.preTag,
        postTag: highlight.postTag
      }
    ) : result.name;
    if (!result) return null;
    return /* @__PURE__ */ jsx(
      ListItemText,
      {
        className: classes.itemText,
        primaryTypographyProps: { variant: "h6" },
        primary: /* @__PURE__ */ jsx(LinkWrapper, { children: title ? title : /* @__PURE__ */ jsxs(Fragment, { children: [
          resultTitle,
          " | ",
          entityTitle ?? resultName,
          " docs"
        ] }) }),
        secondary: /* @__PURE__ */ jsx(
          Typography,
          {
            component: "span",
            style: {
              display: "-webkit-box",
              WebkitBoxOrient: "vertical",
              WebkitLineClamp: lineClamp,
              overflow: "hidden"
            },
            color: "textSecondary",
            variant: "body2",
            children: highlight?.fields.text ? /* @__PURE__ */ jsx(
              HighlightedSearchResultText,
              {
                text: highlight.fields.text,
                preTag: highlight.preTag,
                postTag: highlight.postTag
              }
            ) : result.text
          }
        )
      }
    );
  };
  const ListItemWrapper = ({ children }) => asListItem ? /* @__PURE__ */ jsxs(Fragment, { children: [
    icon && /* @__PURE__ */ jsx(ListItemIcon, { children: typeof icon === "function" ? icon(result) : icon }),
    /* @__PURE__ */ jsx("div", { className: classes.flexContainer, children })
  ] }) : /* @__PURE__ */ jsx(Fragment, { children });
  return /* @__PURE__ */ jsx(ListItemWrapper, { children: /* @__PURE__ */ jsx(TextItem, {}) });
};

export { TechDocsSearchResultListItem };
