// @backstage/plugin-catalog/dist/alpha/filters.esm.js
import { jsx } from 'react/jsx-runtime';
import { CatalogFilterBlueprint } from '@backstage/plugin-catalog-react/alpha';

const catalogTagCatalogFilter = CatalogFilterBlueprint.make({
  name: "tag",
  params: {
    loader: async () => {
      const { EntityTagPicker } = await import('@backstage/plugin-catalog-react');
      return /* @__PURE__ */ jsx(EntityTagPicker, {});
    }
  }
});
const catalogKindCatalogFilter = CatalogFilterBlueprint.makeWithOverrides({
  name: "kind",
  config: {
    schema: {
      initialFilter: (z) => z.string().default("component")
    }
  },
  factory(originalFactory, { config }) {
    return originalFactory({
      loader: async () => {
        const { EntityKindPicker } = await import('@backstage/plugin-catalog-react');
        return /* @__PURE__ */ jsx(EntityKindPicker, { initialFilter: config.initialFilter });
      }
    });
  }
});
const catalogTypeCatalogFilter = CatalogFilterBlueprint.make({
  name: "type",
  params: {
    loader: async () => {
      const { EntityTypePicker } = await import('@backstage/plugin-catalog-react');
      return /* @__PURE__ */ jsx(EntityTypePicker, {});
    }
  }
});
const catalogModeCatalogFilter = CatalogFilterBlueprint.makeWithOverrides({
  name: "mode",
  config: {
    schema: {
      mode: (z) => z.enum(["owners-only", "all"]).optional()
    }
  },
  factory(originalFactory, { config }) {
    return originalFactory({
      loader: async () => {
        const { EntityOwnerPicker } = await import('@backstage/plugin-catalog-react');
        return /* @__PURE__ */ jsx(EntityOwnerPicker, { mode: config.mode });
      }
    });
  }
});
const catalogNamespaceCatalogFilter = CatalogFilterBlueprint.make({
  name: "namespace",
  params: {
    loader: async () => {
      const { EntityNamespacePicker } = await import('@backstage/plugin-catalog-react');
      return /* @__PURE__ */ jsx(EntityNamespacePicker, {});
    }
  }
});
const catalogLifecycleCatalogFilter = CatalogFilterBlueprint.make({
  name: "lifecycle",
  params: {
    loader: async () => {
      const { EntityLifecyclePicker } = await import('@backstage/plugin-catalog-react');
      return /* @__PURE__ */ jsx(EntityLifecyclePicker, {});
    }
  }
});
const catalogProcessingStatusCatalogFilter = CatalogFilterBlueprint.make({
  name: "processing-status",
  params: {
    loader: async () => {
      const { EntityProcessingStatusPicker } = await import('@backstage/plugin-catalog-react');
      return /* @__PURE__ */ jsx(EntityProcessingStatusPicker, {});
    }
  }
});
const catalogListCatalogFilter = CatalogFilterBlueprint.makeWithOverrides({
  name: "list",
  config: {
    schema: {
      initialFilter: (z) => z.enum(["owned", "starred", "all"]).default("owned")
    }
  },
  factory(originalFactory, { config }) {
    return originalFactory({
      loader: async () => {
        const { UserListPicker } = await import('@backstage/plugin-catalog-react');
        return /* @__PURE__ */ jsx(UserListPicker, { initialFilter: config.initialFilter });
      }
    });
  }
});
var filters = [
  catalogKindCatalogFilter,
  catalogTypeCatalogFilter,
  catalogListCatalogFilter,
  catalogModeCatalogFilter,
  catalogLifecycleCatalogFilter,
  catalogTagCatalogFilter,
  catalogProcessingStatusCatalogFilter,
  catalogNamespaceCatalogFilter
];

export { filters as default };
