// @backstage/core-components/dist/layout/ItemCard/ItemCardHeader.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import Box from '@material-ui/core/Box';
import { makeStyles, createStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';

const styles = (theme) => createStyles({
  root: {
    color: theme.palette.text.primary,
    padding: theme.spacing(2, 2, 3),
    backgroundImage: theme.getPageTheme({ themeId: "card" }).backgroundImage,
    backgroundPosition: 0,
    backgroundSize: "inherit"
  }
});
const useStyles = makeStyles(styles, { name: "BackstageItemCardHeader" });
function ItemCardHeader(props) {
  const { title, subtitle, children } = props;
  const classes = useStyles(props);
  return /* @__PURE__ */ jsxs(Box, { className: classes.root, children: [
    subtitle && /* @__PURE__ */ jsx(Typography, { variant: "subtitle2", component: "h3", children: subtitle }),
    title && /* @__PURE__ */ jsx(Typography, { variant: "h6", component: "h4", children: title }),
    children
  ] });
}

export { ItemCardHeader };
