// @backstage/plugin-catalog-import/dist/components/Buttons/index.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { LinkButton } from '@backstage/core-components';
import { useTranslationRef } from '@backstage/frontend-plugin-api';
import Button from '@material-ui/core/Button';
import CircularProgress from '@material-ui/core/CircularProgress';
import { makeStyles } from '@material-ui/core/styles';
import { catalogImportTranslationRef } from '../../translation.esm.js';

const useStyles = makeStyles((theme) => ({
  wrapper: {
    marginTop: theme.spacing(1),
    marginRight: theme.spacing(1),
    position: "relative"
  },
  buttonProgress: {
    position: "absolute",
    top: "50%",
    left: "50%",
    marginTop: -12,
    marginLeft: -12
  },
  button: {
    marginTop: theme.spacing(1),
    marginRight: theme.spacing(1)
  }
}));
const NextButton = (props) => {
  const { loading, ...buttonProps } = props;
  const classes = useStyles();
  return /* @__PURE__ */ jsxs("div", { className: classes.wrapper, children: [
    /* @__PURE__ */ jsx(
      Button,
      {
        color: "primary",
        variant: "contained",
        ...buttonProps,
        disabled: props.disabled || props.loading
      }
    ),
    props.loading && /* @__PURE__ */ jsx(CircularProgress, { size: "1.5rem", className: classes.buttonProgress }),
    props.loading
  ] });
};
const BackButton = (props) => {
  const { t } = useTranslationRef(catalogImportTranslationRef);
  const classes = useStyles();
  return /* @__PURE__ */ jsx(Button, { variant: "outlined", className: classes.button, ...props, children: props.children || t("buttons.back") });
};
const ViewComponentButton = (props) => {
  const classes = useStyles();
  return /* @__PURE__ */ jsx(
    LinkButton,
    {
      color: "primary",
      variant: "contained",
      className: classes.button,
      ...props,
      children: props.children
    }
  );
};

export { BackButton, NextButton, ViewComponentButton };
