// @backstage-community/plugin-rbac/dist/components/DownloadUserStatistics.esm.js
import React from 'react';
import { useApi } from '@backstage/core-plugin-api';
import Button from '@mui/material/Button';
import { licensedUsersApiRef } from '../api/LicensedUsersClient.esm.js';

function DownloadCSVLink() {
  const licensedUsersClient = useApi(licensedUsersApiRef);
  const handleDownload = async (event) => {
    event.preventDefault();
    try {
      const response = await licensedUsersClient.downloadStatistics();
      if (response.ok) {
        const csvData = await response.text();
        const blob = new Blob([csvData], { type: "text/csv" });
        const url = window.URL.createObjectURL(blob);
        const a = document.createElement("a");
        a.href = url;
        a.download = "licensed-users.csv";
        document.body.appendChild(a);
        a.click();
        document.body.removeChild(a);
        window.URL.revokeObjectURL(url);
      } else {
        throw new Error(
          `Failed to download the csv file with list licensed users  ${response.statusText}`
        );
      }
    } catch (error) {
      throw new Error(`Error during the download: ${error}`);
    }
  };
  return /* @__PURE__ */ React.createElement(
    Button,
    {
      href: "/download-csv",
      onClick: handleDownload,
      sx: {
        color: (theme) => theme.palette.link,
        textDecoration: "underline",
        marginTop: "1rem"
      },
      size: "small"
    },
    "Download User List"
  );
}

export { DownloadCSVLink as default };
