// @backstage-community/plugin-rbac/dist/components/EditRole.esm.js
import React from 'react';
import { parseEntityRef } from '@backstage/catalog-model';
import { Link } from '@backstage/core-components';
import EditIcon from '@mui/icons-material/Edit';
import IconButton from '@mui/material/IconButton';
import Tooltip from '@mui/material/Tooltip';
import Typography from '@mui/material/Typography';

const EditRole = ({
  roleName,
  tooltip,
  disable,
  dataTestId,
  to
}) => {
  const { name, namespace, kind } = parseEntityRef(roleName);
  return /* @__PURE__ */ React.createElement(Tooltip, { title: tooltip ?? "" }, /* @__PURE__ */ React.createElement(Typography, { component: "span", "data-testid": dataTestId }, /* @__PURE__ */ React.createElement(
    IconButton,
    {
      component: Link,
      "aria-label": "Update",
      disabled: disable,
      title: tooltip ?? "Edit Role",
      to: to ?? `../role/${kind}/${namespace}/${name}`,
      style: { padding: "0.5rem", color: "inherit", borderRadius: "50%" },
      sx: { "&:hover": { borderRadius: "50%" } }
    },
    /* @__PURE__ */ React.createElement(EditIcon, null)
  )));
};

export { EditRole as default };
