// @backstage-community/plugin-rbac-common/dist/permissions.esm.js
import { createPermission } from '@backstage/plugin-permission-common';

const RESOURCE_TYPE_POLICY_ENTITY = "policy-entity";
const policyEntityReadPermission = createPermission({
  name: "policy.entity.read",
  attributes: {
    action: "read"
  },
  resourceType: RESOURCE_TYPE_POLICY_ENTITY
});
const policyEntityCreatePermission = createPermission({
  name: "policy.entity.create",
  attributes: {
    action: "create"
  },
  resourceType: RESOURCE_TYPE_POLICY_ENTITY
});
const policyEntityDeletePermission = createPermission({
  name: "policy.entity.delete",
  attributes: {
    action: "delete"
  },
  resourceType: RESOURCE_TYPE_POLICY_ENTITY
});
const policyEntityUpdatePermission = createPermission({
  name: "policy.entity.update",
  attributes: {
    action: "update"
  },
  resourceType: RESOURCE_TYPE_POLICY_ENTITY
});
const policyEntityPermissions = [
  policyEntityReadPermission,
  policyEntityCreatePermission,
  policyEntityDeletePermission,
  policyEntityUpdatePermission
];

export { RESOURCE_TYPE_POLICY_ENTITY, policyEntityCreatePermission, policyEntityDeletePermission, policyEntityPermissions, policyEntityReadPermission, policyEntityUpdatePermission };
