// @backstage-community/plugin-rbac-common/dist/permissions.cjs.js
'use strict';

var pluginPermissionCommon = require('@backstage/plugin-permission-common');

const RESOURCE_TYPE_POLICY_ENTITY = "policy-entity";
const policyEntityReadPermission = pluginPermissionCommon.createPermission({
  name: "policy.entity.read",
  attributes: {
    action: "read"
  },
  resourceType: RESOURCE_TYPE_POLICY_ENTITY
});
const policyEntityCreatePermission = pluginPermissionCommon.createPermission({
  name: "policy.entity.create",
  attributes: {
    action: "create"
  },
  resourceType: RESOURCE_TYPE_POLICY_ENTITY
});
const policyEntityDeletePermission = pluginPermissionCommon.createPermission({
  name: "policy.entity.delete",
  attributes: {
    action: "delete"
  },
  resourceType: RESOURCE_TYPE_POLICY_ENTITY
});
const policyEntityUpdatePermission = pluginPermissionCommon.createPermission({
  name: "policy.entity.update",
  attributes: {
    action: "update"
  },
  resourceType: RESOURCE_TYPE_POLICY_ENTITY
});
const policyEntityPermissions = [
  policyEntityReadPermission,
  policyEntityCreatePermission,
  policyEntityDeletePermission,
  policyEntityUpdatePermission
];

exports.RESOURCE_TYPE_POLICY_ENTITY = RESOURCE_TYPE_POLICY_ENTITY;
exports.policyEntityCreatePermission = policyEntityCreatePermission;
exports.policyEntityDeletePermission = policyEntityDeletePermission;
exports.policyEntityPermissions = policyEntityPermissions;
exports.policyEntityReadPermission = policyEntityReadPermission;
exports.policyEntityUpdatePermission = policyEntityUpdatePermission;
