// @backstage/plugin-scaffolder/dist/components/fields/RepoUrlPicker/GerritRepoPicker.esm.js
import { jsxs, Fragment, jsx } from 'react/jsx-runtime';
import FormControl from '@material-ui/core/FormControl';
import TextField from '@material-ui/core/TextField';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { scaffolderTranslationRef } from '../../../translation.esm.js';

const GerritRepoPicker = (props) => {
  const { onChange, rawErrors, state, isDisabled } = props;
  const { t } = useTranslationRef(scaffolderTranslationRef);
  const { workspace, owner } = state;
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(FormControl, { margin: "normal", error: rawErrors?.length > 0 && !workspace, children: /* @__PURE__ */ jsx(
      TextField,
      {
        id: "ownerInput",
        label: t("fields.gerritRepoPicker.owner.title"),
        onChange: (e) => onChange({ owner: e.target.value }),
        helperText: t("fields.gerritRepoPicker.owner.description"),
        disabled: isDisabled,
        value: owner
      }
    ) }),
    /* @__PURE__ */ jsx(
      FormControl,
      {
        margin: "normal",
        required: true,
        error: rawErrors?.length > 0 && !workspace,
        children: /* @__PURE__ */ jsx(
          TextField,
          {
            id: "parentInput",
            label: t("fields.gerritRepoPicker.parent.title"),
            onChange: (e) => onChange({ workspace: e.target.value }),
            disabled: isDisabled,
            value: workspace,
            helperText: t("fields.gerritRepoPicker.parent.description")
          }
        )
      }
    )
  ] });
};

export { GerritRepoPicker };
