// @backstage/plugin-scaffolder/dist/components/fields/RepoUrlPicker/GiteaRepoPicker.esm.js
import { jsx, Fragment, jsxs } from 'react/jsx-runtime';
import FormControl from '@material-ui/core/FormControl';
import FormHelperText from '@material-ui/core/FormHelperText';
import TextField from '@material-ui/core/TextField';
import { Select } from '@backstage/core-components';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { scaffolderTranslationRef } from '../../../translation.esm.js';

const GiteaRepoPicker = (props) => {
  const { allowedOwners = [], state, onChange, rawErrors, isDisabled } = props;
  const { t } = useTranslationRef(scaffolderTranslationRef);
  const ownerItems = allowedOwners ? allowedOwners.map((i) => ({ label: i, value: i })) : [{ label: "Loading...", value: "loading" }];
  const { owner } = state;
  return /* @__PURE__ */ jsx(Fragment, { children: /* @__PURE__ */ jsx(
    FormControl,
    {
      margin: "normal",
      required: true,
      error: rawErrors?.length > 0 && !owner,
      children: allowedOwners?.length ? /* @__PURE__ */ jsxs(Fragment, { children: [
        /* @__PURE__ */ jsx(
          Select,
          {
            native: true,
            label: t("fields.giteaRepoPicker.owner.title"),
            onChange: (selected) => onChange({
              owner: String(
                Array.isArray(selected) ? selected[0] : selected
              )
            }),
            disabled: isDisabled || allowedOwners.length === 1,
            selected: owner,
            items: ownerItems
          }
        ),
        /* @__PURE__ */ jsx(FormHelperText, { children: t("fields.giteaRepoPicker.owner.description") })
      ] }) : /* @__PURE__ */ jsx(Fragment, { children: /* @__PURE__ */ jsx(
        TextField,
        {
          id: "ownerInput",
          label: t("fields.giteaRepoPicker.owner.inputTitle"),
          onChange: (e) => onChange({ owner: e.target.value }),
          helperText: t("fields.giteaRepoPicker.owner.description"),
          disabled: isDisabled,
          value: owner
        }
      ) })
    }
  ) });
};

export { GiteaRepoPicker };
