// @backstage/plugin-api-docs/dist/components/ComponentsCards/ProvidingComponentsCard.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { RELATION_API_PROVIDED_BY } from '@backstage/catalog-model';
import Typography from '@material-ui/core/Typography';
import { EntityTable, useEntity, useRelatedEntities } from '@backstage/plugin-catalog-react';
import { InfoCard, Progress, WarningPanel, CodeSnippet, Link } from '@backstage/core-components';
import { useTranslationRef } from '@backstage/frontend-plugin-api';
import { apiDocsTranslationRef } from '../../translation.esm.js';

const ProvidingComponentsCard = (props) => {
  const { variant = "gridItem", columns = EntityTable.componentEntityColumns } = props;
  const { entity } = useEntity();
  const { entities, loading, error } = useRelatedEntities(entity, {
    type: RELATION_API_PROVIDED_BY
  });
  const { t } = useTranslationRef(apiDocsTranslationRef);
  if (loading) {
    return /* @__PURE__ */ jsx(InfoCard, { variant, title: t("providingComponentsCard.title"), children: /* @__PURE__ */ jsx(Progress, {}) });
  }
  if (error || !entities) {
    return /* @__PURE__ */ jsx(InfoCard, { variant, title: t("providingComponentsCard.title"), children: /* @__PURE__ */ jsx(
      WarningPanel,
      {
        severity: "error",
        title: t("providingComponentsCard.error.title"),
        message: /* @__PURE__ */ jsx(CodeSnippet, { text: `${error}`, language: "text" })
      }
    ) });
  }
  return /* @__PURE__ */ jsx(
    EntityTable,
    {
      title: t("providingComponentsCard.title"),
      variant,
      emptyContent: /* @__PURE__ */ jsxs("div", { style: { textAlign: "center" }, children: [
        /* @__PURE__ */ jsx(Typography, { variant: "body1", children: t("providingComponentsCard.emptyContent.title") }),
        /* @__PURE__ */ jsx(Typography, { variant: "body2", children: /* @__PURE__ */ jsx(Link, { to: "https://backstage.io/docs/features/software-catalog/descriptor-format#specprovidesapis-optional", children: t("apisCardHelpLinkTitle") }) })
      ] }),
      columns,
      entities
    }
  );
};

export { ProvidingComponentsCard };
