// @backstage/backend-defaults/dist/alpha/entrypoints/actionsRegistry/actionsRegistryServiceFactory.cjs.js
'use strict';

var backendPluginApi = require('@backstage/backend-plugin-api');
var DefaultActionsRegistryService = require('./DefaultActionsRegistryService.cjs.js');
var alpha = require('@backstage/backend-plugin-api/alpha');

const actionsRegistryServiceFactory = backendPluginApi.createServiceFactory({
  service: alpha.actionsRegistryServiceRef,
  deps: {
    metadata: backendPluginApi.coreServices.pluginMetadata,
    httpRouter: backendPluginApi.coreServices.httpRouter,
    httpAuth: backendPluginApi.coreServices.httpAuth,
    logger: backendPluginApi.coreServices.logger,
    auth: backendPluginApi.coreServices.auth
  },
  factory: ({ metadata, httpRouter, httpAuth, logger, auth }) => {
    const actionsRegistryService = DefaultActionsRegistryService.DefaultActionsRegistryService.create({
      httpAuth,
      logger,
      auth,
      metadata
    });
    httpRouter.use(actionsRegistryService.createRouter());
    return actionsRegistryService;
  }
});

exports.actionsRegistryServiceFactory = actionsRegistryServiceFactory;
