// @backstage/plugin-scaffolder/dist/components/TemplatingExtensionsPage/TemplateGlobals.esm.js
import { jsx, Fragment, jsxs } from 'react/jsx-runtime';
import { MarkdownContent, CodeSnippet } from '@backstage/core-components';
import Accordion from '@material-ui/core/Accordion';
import AccordionDetails from '@material-ui/core/AccordionDetails';
import AccordionSummary from '@material-ui/core/AccordionSummary';
import Box from '@material-ui/core/Box';
import Typography from '@material-ui/core/Typography';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import classNames from 'classnames';
import { cloneElement, useState, Fragment as Fragment$1 } from 'react';
import { RenderSchema } from '../RenderSchema/RenderSchema.esm.js';
import { ScaffolderUsageExamplesTable } from '../ScaffolderUsageExamplesTable/ScaffolderUsageExamplesTable.esm.js';
import { inspectFunctionArgSchema } from './functionArgs.esm.js';
import { renderFragment } from './navigation.esm.js';

const FunctionDetailContent = ({
  classes,
  name,
  fn,
  t
}) => {
  const expanded = useState({});
  if (!Object.keys(fn).length) {
    return /* @__PURE__ */ jsx(
      Typography,
      {
        style: { fontStyle: "italic" },
        "data-testid": `${name}.metadataAbsent`,
        children: t("templatingExtensions.content.functions.metadataAbsent")
      }
    );
  }
  const schema = fn.schema;
  const partialSchemaRenderContext = {
    classes,
    expanded,
    headings: [/* @__PURE__ */ jsx(Typography, { variant: "h6", component: "h4" })]
  };
  return /* @__PURE__ */ jsxs(Fragment$1, { children: [
    fn.description && /* @__PURE__ */ jsx(MarkdownContent, { content: fn.description }),
    schema?.arguments?.length && /* @__PURE__ */ jsxs(Box, { pb: 2, children: [
      /* @__PURE__ */ jsx(Typography, { variant: "h5", component: "h3", children: t("templatingExtensions.content.functions.schema.arguments") }),
      schema.arguments.map((arg, i) => {
        const [argSchema, required] = inspectFunctionArgSchema(arg);
        return /* @__PURE__ */ jsxs(Fragment$1, { children: [
          /* @__PURE__ */ jsx(
            "div",
            {
              className: classNames({ [classes.argRequired]: required }),
              children: /* @__PURE__ */ jsx(
                Typography,
                {
                  variant: "h6",
                  component: "h4",
                  children: `[${i}]`
                }
              )
            }
          ),
          /* @__PURE__ */ jsx(
            RenderSchema,
            {
              strategy: "root",
              context: {
                parentId: `${name}.arg${i}`,
                ...partialSchemaRenderContext,
                headings: [/* @__PURE__ */ jsx(Typography, { variant: "h6", component: "h5" })]
              },
              schema: argSchema
            }
          )
        ] }, i);
      })
    ] }, `${name}.args`),
    /* @__PURE__ */ jsxs(Box, { pb: 2, children: [
      /* @__PURE__ */ jsx(Typography, { variant: "h5", component: "h3", children: t("templatingExtensions.content.functions.schema.output") }),
      /* @__PURE__ */ jsx(
        RenderSchema,
        {
          strategy: "root",
          context: {
            parentId: `${name}.output`,
            ...partialSchemaRenderContext
          },
          schema: schema?.output ?? {}
        }
      )
    ] }),
    fn.examples && /* @__PURE__ */ jsxs(Accordion, { children: [
      /* @__PURE__ */ jsx(AccordionSummary, { expandIcon: /* @__PURE__ */ jsx(ExpandMoreIcon, {}), children: /* @__PURE__ */ jsx(Typography, { variant: "h5", component: "h3", children: t("templatingExtensions.content.functions.examples") }) }),
      /* @__PURE__ */ jsx(AccordionDetails, { children: /* @__PURE__ */ jsx(Box, { pb: 2, children: /* @__PURE__ */ jsx(ScaffolderUsageExamplesTable, { examples: fn.examples }) }) })
    ] })
  ] }, `${name}.detail`);
};
const TemplateGlobalFunctions = ({
  classes,
  functions,
  t,
  baseLink,
  selectedItem
}) => {
  if (selectedItem && selectedItem.kind !== "function") {
    return /* @__PURE__ */ jsx(Fragment, {});
  }
  if (!Object.keys(functions).length) {
    return /* @__PURE__ */ jsx("div", { "data-testid": "no-functions", children: t("templatingExtensions.content.functions.notAvailable") });
  }
  return /* @__PURE__ */ jsx("div", { "data-testid": "functions", children: Object.entries(
    selectedItem ? { [selectedItem.name]: functions[selectedItem.name] } : functions
  ).map(([name, fn]) => {
    const fragment = renderFragment({ kind: "function", name });
    return /* @__PURE__ */ jsxs(Box, { pb: 4, "data-testid": name, children: [
      /* @__PURE__ */ jsx(
        Typography,
        {
          id: fragment,
          variant: "h4",
          component: "h2",
          className: classes.code,
          children: name
        }
      ),
      cloneElement(baseLink, {
        to: `${baseLink.props.to}#${fragment}`
      }),
      /* @__PURE__ */ jsx(FunctionDetailContent, { ...{ classes, name, fn, t } })
    ] }, name);
  }) });
};
const TemplateGlobalValues = ({
  classes,
  t,
  values,
  baseLink,
  selectedItem
}) => {
  if (selectedItem && selectedItem.kind !== "value") {
    return /* @__PURE__ */ jsx(Fragment, {});
  }
  if (!Object.keys(values).length) {
    return /* @__PURE__ */ jsx("div", { "data-testid": "no-values", children: t("templatingExtensions.content.values.notAvailable") });
  }
  return /* @__PURE__ */ jsx("div", { "data-testid": "values", children: Object.entries(
    selectedItem ? { [selectedItem.name]: values[selectedItem.name] } : values
  ).map(([name, gv]) => {
    const fragment = renderFragment({ kind: "value", name });
    return /* @__PURE__ */ jsxs(Box, { pb: 4, "data-testid": name, children: [
      /* @__PURE__ */ jsx(
        Typography,
        {
          id: fragment,
          variant: "h4",
          component: "h2",
          className: classes.code,
          children: name
        }
      ),
      cloneElement(baseLink, {
        to: `${baseLink.props.to}#${fragment}`
      }),
      gv.description && /* @__PURE__ */ jsx(MarkdownContent, { content: gv.description }),
      /* @__PURE__ */ jsx(Box, { padding: 1, "data-testid": `${name}.value`, children: /* @__PURE__ */ jsx(
        CodeSnippet,
        {
          text: JSON.stringify(gv.value, null, 2),
          showCopyCodeButton: true,
          language: "json"
        }
      ) })
    ] }, name);
  }) });
};

export { TemplateGlobalFunctions, TemplateGlobalValues };
