// @backstage/frontend-defaults/dist/createApp.esm.js
import { jsx } from 'react/jsx-runtime';
import { lazy, Suspense } from 'react';
import { coreExtensionData } from '@backstage/frontend-plugin-api';
import { Progress } from '@backstage/core-components';
import { defaultConfigLoaderSync } from './core-app-api/src/app/defaultConfigLoader.esm.js';
import { overrideBaseUrlConfigs } from './core-app-api/src/app/overrideBaseUrlConfigs.esm.js';
import { ConfigReader } from '@backstage/config';
import { createSpecializedApp } from '@backstage/frontend-app-api';
import appPlugin from '@backstage/plugin-app';
import { discoverAvailableFeatures } from './discovery.esm.js';
import { resolveAsyncFeatures } from './resolution.esm.js';
import { maybeCreateErrorPage } from './maybeCreateErrorPage.esm.js';

function createApp(options) {
  let suspenseFallback = options?.advanced?.loadingElement;
  if (suspenseFallback === void 0) {
    suspenseFallback = /* @__PURE__ */ jsx(Progress, {});
  }
  async function appLoader() {
    const config = await options?.advanced?.configLoader?.().then((c) => c.config) ?? ConfigReader.fromConfigs(
      overrideBaseUrlConfigs(defaultConfigLoaderSync())
    );
    const { features: discoveredFeaturesAndLoaders } = discoverAvailableFeatures(config);
    const { features: loadedFeatures } = await resolveAsyncFeatures({
      config,
      features: [...discoveredFeaturesAndLoaders, ...options?.features ?? []]
    });
    const app = createSpecializedApp({
      features: [appPlugin, ...loadedFeatures],
      config,
      bindRoutes: options?.bindRoutes,
      advanced: options?.advanced
    });
    const errorPage = maybeCreateErrorPage(app);
    if (errorPage) {
      return { default: () => errorPage };
    }
    const rootEl = app.tree.root.instance.getData(
      coreExtensionData.reactElement
    );
    return { default: () => rootEl };
  }
  const LazyApp = lazy(appLoader);
  return {
    createRoot() {
      return /* @__PURE__ */ jsx(Suspense, { fallback: suspenseFallback, children: /* @__PURE__ */ jsx(LazyApp, {}) });
    }
  };
}

export { createApp };
