// @backstage/plugin-catalog-import/dist/components/StepPrepareCreatePullRequest/AutocompleteTextField.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import CircularProgress from '@material-ui/core/CircularProgress';
import TextField from '@material-ui/core/TextField';
import Autocomplete from '@material-ui/lab/Autocomplete';
import { Fragment } from 'react';
import { Controller } from 'react-hook-form';

const AutocompleteTextField = (props) => {
  const {
    name,
    options,
    required,
    errors,
    rules,
    loading = false,
    loadingText,
    helperText,
    errorHelperText,
    textFieldProps = {}
  } = props;
  return /* @__PURE__ */ jsx(
    Controller,
    {
      name,
      rules,
      render: ({ field: { onChange } }) => /* @__PURE__ */ jsx(
        Autocomplete,
        {
          loading,
          loadingText,
          options: options || [],
          autoSelect: true,
          freeSolo: true,
          onChange: (_event, value) => onChange(value),
          renderInput: (params) => /* @__PURE__ */ jsx(
            TextField,
            {
              ...params,
              helperText: errors?.[name] && errorHelperText || helperText,
              error: Boolean(errors?.[name]),
              margin: "normal",
              variant: "outlined",
              required,
              InputProps: {
                ...params.InputProps,
                endAdornment: /* @__PURE__ */ jsxs(Fragment, { children: [
                  loading ? /* @__PURE__ */ jsx(CircularProgress, { color: "inherit", size: "1em" }) : null,
                  params.InputProps.endAdornment
                ] })
              },
              ...textFieldProps
            }
          )
        }
      )
    }
  );
};

export { AutocompleteTextField };
