// @backstage/plugin-search-react/dist/components/DefaultResultListItem/DefaultResultListItem.esm.js
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { AnalyticsContext } from '@backstage/core-plugin-api';
import { HighlightedSearchResultText } from '../HighlightedSearchResultText/HighlightedSearchResultText.esm.js';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import Box from '@material-ui/core/Box';
import Typography from '@material-ui/core/Typography';
import { Link } from '@backstage/core-components';

const DefaultResultListItemComponent = ({
  result,
  highlight,
  icon,
  secondaryAction,
  lineClamp = 5
}) => {
  if (!result) return null;
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    icon && /* @__PURE__ */ jsx(ListItemIcon, { children: icon }),
    /* @__PURE__ */ jsx(
      ListItemText,
      {
        primaryTypographyProps: { variant: "h6" },
        primary: /* @__PURE__ */ jsx(Link, { noTrack: true, to: result.location, children: highlight?.fields.title ? /* @__PURE__ */ jsx(
          HighlightedSearchResultText,
          {
            text: highlight?.fields.title || "",
            preTag: highlight?.preTag || "",
            postTag: highlight?.postTag || ""
          }
        ) : result.title }),
        secondary: /* @__PURE__ */ jsx(
          Typography,
          {
            component: "span",
            style: {
              display: "-webkit-box",
              WebkitBoxOrient: "vertical",
              WebkitLineClamp: lineClamp,
              overflow: "hidden"
            },
            color: "textSecondary",
            variant: "body2",
            children: highlight?.fields.text ? /* @__PURE__ */ jsx(
              HighlightedSearchResultText,
              {
                text: highlight.fields.text,
                preTag: highlight.preTag,
                postTag: highlight.postTag
              }
            ) : result.text
          }
        )
      }
    ),
    secondaryAction && /* @__PURE__ */ jsx(Box, { alignItems: "flex-end", children: secondaryAction })
  ] });
};
const HigherOrderDefaultResultListItem = (props) => {
  return /* @__PURE__ */ jsx(
    AnalyticsContext,
    {
      attributes: {
        pluginId: "search",
        extension: "DefaultResultListItem"
      },
      children: /* @__PURE__ */ jsx(DefaultResultListItemComponent, { ...props })
    }
  );
};

export { HigherOrderDefaultResultListItem as DefaultResultListItem, DefaultResultListItemComponent };
