// @backstage/plugin-signals/dist/package.json.esm.js
var name = "@backstage/plugin-signals";
var version = "0.0.22";
var backstage = {
	role: "frontend-plugin",
	pluginId: "signals",
	pluginPackages: [
		"@backstage/plugin-signals",
		"@backstage/plugin-signals-backend",
		"@backstage/plugin-signals-node",
		"@backstage/plugin-signals-react"
	]
};
var publishConfig = {
	access: "public"
};
var repository = {
	type: "git",
	url: "https://github.com/backstage/backstage",
	directory: "plugins/signals"
};
var license = "Apache-2.0";
var sideEffects = false;
var exports = {
	".": "./src/index.ts",
	"./alpha": "./src/alpha.tsx",
	"./package.json": "./package.json"
};
var main = "src/index.ts";
var types = "src/index.ts";
var typesVersions = {
	"*": {
		alpha: [
			"src/alpha.tsx"
		],
		"package.json": [
			"package.json"
		]
	}
};
var files = [
	"dist"
];
var scripts = {
	build: "backstage-cli package build",
	clean: "backstage-cli package clean",
	lint: "backstage-cli package lint",
	prepack: "backstage-cli package prepack",
	postpack: "backstage-cli package postpack",
	start: "backstage-cli package start",
	test: "backstage-cli package test"
};
var dependencies = {
	"@backstage/core-compat-api": "workspace:^",
	"@backstage/core-components": "workspace:^",
	"@backstage/core-plugin-api": "workspace:^",
	"@backstage/frontend-plugin-api": "workspace:^",
	"@backstage/plugin-signals-react": "workspace:^",
	"@backstage/theme": "workspace:^",
	"@backstage/types": "workspace:^",
	"@material-ui/core": "^4.12.4",
	"@material-ui/icons": "^4.9.1",
	"@material-ui/lab": "^4.0.0-alpha.61",
	"react-use": "^17.2.4",
	uuid: "^11.0.0"
};
var devDependencies = {
	"@backstage/cli": "workspace:^",
	"@backstage/core-app-api": "workspace:^",
	"@backstage/dev-utils": "workspace:^",
	"@backstage/test-utils": "workspace:^",
	"@testing-library/jest-dom": "^6.0.0",
	"@testing-library/react": "^16.0.0",
	"@testing-library/user-event": "^14.0.0",
	"@types/react": "^18.0.0",
	"jest-websocket-mock": "^2.5.0",
	msw: "^1.0.0",
	react: "^18.0.2",
	"react-dom": "^18.0.2",
	"react-router-dom": "^6.3.0"
};
var peerDependencies = {
	"@types/react": "^17.0.0 || ^18.0.0",
	react: "^17.0.0 || ^18.0.0",
	"react-dom": "^17.0.0 || ^18.0.0",
	"react-router-dom": "^6.3.0"
};
var peerDependenciesMeta = {
	"@types/react": {
		optional: true
	}
};
var _package = {
	name: name,
	version: version,
	backstage: backstage,
	publishConfig: publishConfig,
	repository: repository,
	license: license,
	sideEffects: sideEffects,
	exports: exports,
	main: main,
	types: types,
	typesVersions: typesVersions,
	files: files,
	scripts: scripts,
	dependencies: dependencies,
	devDependencies: devDependencies,
	peerDependencies: peerDependencies,
	peerDependenciesMeta: peerDependenciesMeta
};

export { backstage, _package as default, dependencies, devDependencies, exports, files, license, main, name, peerDependencies, peerDependenciesMeta, publishConfig, repository, scripts, sideEffects, types, typesVersions, version };
