// @backstage/repo-tools/dist/commands/package/schema/openapi/generate/server.cjs.js
'use strict';

var chalk = require('chalk');
var path = require('path');
var YAML = require('js-yaml');
var constants = require('../../../../../lib/openapi/constants.cjs.js');
var paths = require('../../../../../lib/paths.cjs.js');
var fs = require('fs-extra');
var exec = require('../../../../../lib/exec.cjs.js');
var backendPluginApi = require('@backstage/backend-plugin-api');
var helpers = require('../../../../../lib/openapi/helpers.cjs.js');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var chalk__default = /*#__PURE__*/_interopDefaultCompat(chalk);
var YAML__default = /*#__PURE__*/_interopDefaultCompat(YAML);
var fs__default = /*#__PURE__*/_interopDefaultCompat(fs);

async function generateSpecFile() {
  const openapiPath = await helpers.getPathToCurrentOpenApiSpec();
  const yaml = YAML__default.default.load(await fs__default.default.readFile(openapiPath, "utf8"));
  const tsPath = paths.paths.resolveTarget(constants.TS_SCHEMA_PATH);
  const schemaDir = path.dirname(tsPath);
  await fs__default.default.mkdirp(schemaDir);
  const oldTsPath = paths.paths.resolveTarget(constants.OLD_SCHEMA_PATH);
  if (fs__default.default.existsSync(oldTsPath)) {
    console.warn(`Removing old schema file at ${oldTsPath}`);
    fs__default.default.removeSync(oldTsPath);
  }
  await fs__default.default.writeFile(
    tsPath,
    `//

// ******************************************************************
// * THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY. *
// ******************************************************************
import {createValidatedOpenApiRouterFromGeneratedEndpointMap} from '@backstage/backend-openapi-utils';
import {EndpointMap} from './apis';
export const spec = ${JSON.stringify(yaml, null, 2)} as const;
export const createOpenApiRouter = async (
  options?: Parameters<typeof createValidatedOpenApiRouterFromGeneratedEndpointMap>['1'],
) => createValidatedOpenApiRouterFromGeneratedEndpointMap<EndpointMap>(spec, options);
`
  );
  const indexFile = path.join(schemaDir, "..", "index.ts");
  await fs__default.default.writeFile(
    indexFile,
    `//
    export * from './generated';`
  );
  await exec.exec(`yarn backstage-cli package lint`, ["--fix", tsPath, indexFile]);
  if (await paths.paths.resolveTargetRoot("node_modules/.bin/prettier")) {
    await exec.exec(`yarn prettier`, ["--write", tsPath, indexFile], {
      cwd: paths.paths.targetRoot
    });
  }
}
async function generate(serverAdditionalProperties, abortSignal) {
  const resolvedOpenapiPath = await helpers.getPathToCurrentOpenApiSpec();
  const resolvedOutputDirectory = await helpers.getRelativePathToFile(constants.OUTPUT_PATH);
  await fs__default.default.emptyDir(resolvedOutputDirectory);
  await fs__default.default.writeFile(
    path.resolve(resolvedOutputDirectory, ".openapi-generator-ignore"),
    constants.OPENAPI_IGNORE_FILES.join("\n")
  );
  const additionalProperties = helpers.toGeneratorAdditionalProperties({
    initialValue: serverAdditionalProperties
  });
  await exec.exec(
    "node",
    [
      backendPluginApi.resolvePackagePath("@openapitools/openapi-generator-cli", "main.js"),
      "generate",
      "-i",
      resolvedOpenapiPath,
      "-o",
      resolvedOutputDirectory,
      "-g",
      "typescript",
      "-c",
      backendPluginApi.resolvePackagePath(
        "@backstage/repo-tools",
        "templates/typescript-backstage-server.yaml"
      ),
      "--generator-key",
      "v3.0",
      additionalProperties ? `--additional-properties=${additionalProperties}` : ""
    ],
    {
      maxBuffer: Number.MAX_VALUE,
      cwd: backendPluginApi.resolvePackagePath("@backstage/repo-tools"),
      env: {
        ...process.env
      },
      signal: abortSignal?.signal
    }
  );
  await exec.exec(
    `yarn backstage-cli package lint --fix ${resolvedOutputDirectory}`,
    [],
    {
      signal: abortSignal?.signal
    }
  );
  const prettier = paths.paths.resolveTargetRoot("node_modules/.bin/prettier");
  if (prettier) {
    await exec.exec(`${prettier} --write ${resolvedOutputDirectory}`, [], {
      signal: abortSignal?.signal
    });
  }
  fs__default.default.removeSync(path.resolve(resolvedOutputDirectory, ".openapi-generator-ignore"));
  fs__default.default.rmSync(path.resolve(resolvedOutputDirectory, ".openapi-generator"), {
    recursive: true,
    force: true
  });
  await generateSpecFile();
}
async function command({
  abortSignal,
  isWatch = false
}, serverAdditionalProperties) {
  try {
    await generate(serverAdditionalProperties, abortSignal);
    console.log(chalk__default.default.green("Generated server files."));
  } catch (err) {
    if (err.name === "AbortError") {
      console.debug("Server generation aborted.");
      return;
    }
    if (isWatch) {
      console.log(chalk__default.default.red(`Server generation failed:`));
      console.group();
      console.log(chalk__default.default.red(err.message));
      console.groupEnd();
    } else {
      console.log(chalk__default.default.red(err.message));
      console.log(chalk__default.default.red(`OpenAPI server stub generation failed.`));
    }
  }
}

exports.command = command;
