// @backstage/plugin-catalog/dist/components/CatalogTable/CatalogTableToolbar.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { EntitySearchBar } from '@backstage/plugin-catalog-react';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';

const useToolbarStyles = makeStyles(
  (theme) => ({
    root: {
      paddingTop: theme.spacing(1.25),
      paddingLeft: theme.spacing(2.5),
      paddingBottom: theme.spacing(0.75),
      display: "flex",
      justifyContent: "space-between"
    },
    text: {
      whiteSpace: "nowrap",
      overflow: "hidden",
      textOverflow: "ellipsis"
    }
  }),
  { name: "PluginCatalogTableToolbar" }
);
function CatalogTableToolbar(props) {
  const styles = useToolbarStyles();
  return /* @__PURE__ */ jsxs(Toolbar, { className: styles.root, children: [
    /* @__PURE__ */ jsx(Typography, { variant: "h5", className: styles.text, children: props.title }),
    /* @__PURE__ */ jsx(EntitySearchBar, {})
  ] });
}

export { CatalogTableToolbar };
