// @janus-idp/shared-react/dist/components/common/DeleteDialogContext.cjs.js
'use strict';

var React = require('react');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefaultCompat(React);

const DeleteDialogContext = React.createContext({
  deleteComponent: {},
  setDeleteComponent: () => {
  },
  openDialog: false,
  setOpenDialog: () => {
  }
});
const DeleteDialogContextProvider = (props) => {
  const [openDialog, setOpenDialog] = React__default.default.useState(false);
  const [deleteComponent, setDeleteComponent] = React__default.default.useState({});
  const deleteDialogContextProviderValue = React__default.default.useMemo(
    () => ({
      openDialog,
      setOpenDialog,
      deleteComponent,
      setDeleteComponent
    }),
    [openDialog, setOpenDialog, deleteComponent, setDeleteComponent]
  );
  return /* @__PURE__ */ React__default.default.createElement(DeleteDialogContext.Provider, { value: deleteDialogContextProviderValue }, props.children);
};
const useDeleteDialog = () => React.useContext(DeleteDialogContext);

exports.DeleteDialogContext = DeleteDialogContext;
exports.DeleteDialogContextProvider = DeleteDialogContextProvider;
exports.useDeleteDialog = useDeleteDialog;
