// @janus-idp/shared-react/dist/components/common/DrawerContext.cjs.js
'use strict';

var React = require('react');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefaultCompat(React);

const DrawerContext = React.createContext({
  drawerData: {},
  setDrawerData: () => {
  },
  openDrawer: false,
  setOpenDrawer: () => {
  }
});
const DrawerContextProvider = ({
  children
}) => {
  const [openDrawer, setOpenDrawer] = React__default.default.useState(false);
  const [drawerData, setDrawerData] = React__default.default.useState({});
  const drawerContextProviderValue = React__default.default.useMemo(
    () => ({
      openDrawer,
      setOpenDrawer,
      drawerData,
      setDrawerData
    }),
    [openDrawer, setOpenDrawer, drawerData, setDrawerData]
  );
  return /* @__PURE__ */ React__default.default.createElement(DrawerContext.Provider, { value: drawerContextProviderValue }, children);
};
const useDrawer = () => React.useContext(DrawerContext);

exports.DrawerContext = DrawerContext;
exports.DrawerContextProvider = DrawerContextProvider;
exports.useDrawer = useDrawer;
