// @backstage/plugin-catalog-backend/dist/permissions/rules/createPropertyRule.cjs.js
'use strict';

var alpha = require('@backstage/plugin-catalog-node/alpha');
var pluginPermissionNode = require('@backstage/plugin-permission-node');
var lodash = require('lodash');
var zod = require('zod');

const createPropertyRule = (propertyType) => pluginPermissionNode.createPermissionRule({
  name: `HAS_${propertyType.toUpperCase()}`,
  description: `Allow entities with the specified ${propertyType} subfield`,
  resourceRef: alpha.catalogEntityPermissionResourceRef,
  paramsSchema: zod.z.object({
    key: zod.z.string().describe(`Property within the entities ${propertyType} to match on`),
    value: zod.z.string().optional().describe(`Value of the given property to match on`)
  }),
  apply: (resource, { key, value }) => {
    const foundValue = lodash.get(resource[propertyType], key);
    if (Array.isArray(foundValue)) {
      if (value !== void 0) {
        return foundValue.includes(value);
      }
      return foundValue.length > 0;
    }
    if (value !== void 0) {
      return value === foundValue;
    }
    return !!foundValue;
  },
  toQuery: ({ key, value }) => ({
    key: `${propertyType}.${key}`,
    ...value !== void 0 && { values: [value] }
  })
});

exports.createPropertyRule = createPropertyRule;
