// @backstage/plugin-catalog-backend/dist/permissions/rules/isEntityOwner.cjs.js
'use strict';

var catalogModel = require('@backstage/catalog-model');
var pluginPermissionNode = require('@backstage/plugin-permission-node');
var zod = require('zod');
var alpha = require('@backstage/plugin-catalog-node/alpha');

const isEntityOwner = pluginPermissionNode.createPermissionRule({
  name: "IS_ENTITY_OWNER",
  description: "Allow entities owned by a specified claim",
  resourceRef: alpha.catalogEntityPermissionResourceRef,
  paramsSchema: zod.z.object({
    claims: zod.z.array(zod.z.string()).describe(
      `List of claims to match at least one on within ${catalogModel.RELATION_OWNED_BY}`
    )
  }),
  apply: (resource, { claims }) => {
    if (!resource.relations) {
      return false;
    }
    return resource.relations.filter((relation) => relation.type === catalogModel.RELATION_OWNED_BY).some((relation) => claims.includes(relation.targetRef));
  },
  toQuery: ({ claims }) => ({
    key: "relations.ownedBy",
    values: claims
  })
});

exports.isEntityOwner = isEntityOwner;
