// @backstage/plugin-scaffolder-backend-module-bitbucket-cloud/dist/actions/bitbucketCloudBranchRestriction.examples.cjs.js
'use strict';

var yaml = require('yaml');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var yaml__default = /*#__PURE__*/_interopDefaultCompat(yaml);

const examples = [
  {
    description: "restrict push to the main branch, except for alice, bob, and the admins group",
    example: yaml__default.default.stringify({
      steps: [
        {
          id: "createBranchRestriction",
          action: "bitbucketCloud:branchRestriction:create",
          name: "Create Bitbucket Cloud branch restriction",
          input: {
            repoUrl: "bitbucket.org?repo=repo&workspace=workspace&project=project",
            kind: "push",
            users: [
              {
                uuid: "{a-b-c-d}"
              },
              {
                uuid: "{e-f-g-h}"
              }
            ],
            groups: [
              {
                slug: "admins"
              }
            ]
          }
        }
      ]
    })
  },
  {
    description: "restrict push to the main branch, except for the admins group",
    example: yaml__default.default.stringify({
      steps: [
        {
          id: "restrictPushToFeatureBranches",
          action: "bitbucketCloud:branchRestriction:create",
          name: "Create Bitbucket Cloud branch restriction by branch type",
          input: {
            repoUrl: "bitbucket.org?repo=repo&workspace=workspace&project=project",
            kind: "push",
            groups: [
              {
                slug: "admins"
              }
            ]
          }
        }
      ]
    })
  },
  {
    description: "require passing builds to merge to branches matching a pattern test-feature/*",
    example: yaml__default.default.stringify({
      steps: [
        {
          id: "requirePassingBuildsToMergeByPattern",
          action: "bitbucketCloud:branchRestriction:create",
          name: "Create Bitbucket Cloud require passing build to merge to branches matching a pattern",
          input: {
            repoUrl: "bitbucket.org?repo=repo&workspace=workspace&project=project",
            kind: "require_passing_builds_to_merge",
            branchMatchKind: "glob",
            pattern: "test-feature/*"
          }
        }
      ]
    })
  },
  {
    description: "require approvals to merge to branches matching a pattern test-feature/*",
    example: yaml__default.default.stringify({
      steps: [
        {
          id: "requireApprovalsToMergeByPattern",
          action: "bitbucketCloud:branchRestriction:create",
          name: "Create Bitbucket Cloud require approvals to merge branch restriction",
          input: {
            repoUrl: "bitbucket.org?repo=repo&workspace=workspace&project=project",
            kind: "require_approvals_to_merge",
            branchMatchKind: "glob",
            pattern: "test-feature/*"
          }
        }
      ]
    })
  }
];

exports.examples = examples;
