// @backstage/plugin-scaffolder-backend-module-bitbucket-cloud/dist/actions/helpers.cjs.js
'use strict';

var bitbucket = require('bitbucket');

const getBitbucketClient = (config) => {
  if (config.username && config.appPassword) {
    return new bitbucket.Bitbucket({
      auth: {
        username: config.username,
        password: config.appPassword
      }
    });
  } else if (config.token) {
    return new bitbucket.Bitbucket({
      auth: {
        token: config.token
      }
    });
  }
  throw new Error(
    `Authorization has not been provided for Bitbucket Cloud. Please add either username + appPassword to the Integrations config or a user login auth token`
  );
};
const getAuthorizationHeader = (config) => {
  if (config.username && config.appPassword) {
    const buffer = Buffer.from(
      `${config.username}:${config.appPassword}`,
      "utf8"
    );
    return `Basic ${buffer.toString("base64")}`;
  }
  if (config.token) {
    return `Bearer ${config.token}`;
  }
  throw new Error(
    `Authorization has not been provided for Bitbucket Cloud. Please add either username + appPassword to the Integrations config or a user login auth token`
  );
};

exports.getAuthorizationHeader = getAuthorizationHeader;
exports.getBitbucketClient = getBitbucketClient;
