// @backstage/core-components/dist/components/DependencyGraph/DefaultNode.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { useState, useRef, useLayoutEffect } from 'react';
import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles(
  (theme) => ({
    node: {
      fill: theme.palette.primary.light,
      stroke: theme.palette.primary.light
    },
    text: {
      fill: theme.palette.primary.contrastText
    }
  }),
  { name: "BackstageDependencyGraphDefaultNode" }
);
function DefaultNode({ node: { id } }) {
  const classes = useStyles();
  const [width, setWidth] = useState(0);
  const [height, setHeight] = useState(0);
  const idRef = useRef(null);
  useLayoutEffect(() => {
    if (idRef.current) {
      let { height: renderedHeight, width: renderedWidth } = idRef.current.getBBox();
      renderedHeight = Math.round(renderedHeight);
      renderedWidth = Math.round(renderedWidth);
      if (renderedHeight !== height || renderedWidth !== width) {
        setWidth(renderedWidth);
        setHeight(renderedHeight);
      }
    }
  }, [width, height]);
  const padding = 10;
  const paddedWidth = width + padding * 2;
  const paddedHeight = height + padding * 2;
  return /* @__PURE__ */ jsxs("g", { children: [
    /* @__PURE__ */ jsx(
      "rect",
      {
        className: classes.node,
        width: paddedWidth,
        height: paddedHeight,
        rx: 10
      }
    ),
    /* @__PURE__ */ jsx(
      "text",
      {
        ref: idRef,
        className: classes.text,
        y: paddedHeight / 2,
        x: paddedWidth / 2,
        textAnchor: "middle",
        alignmentBaseline: "middle",
        children: id
      }
    )
  ] });
}

export { DefaultNode };
