// @backstage/plugin-scaffolder/dist/components/TemplateTypePicker/TemplateTypePicker.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import capitalize from 'lodash/capitalize';
import { Progress } from '@backstage/core-components';
import Box from '@material-ui/core/Box';
import Checkbox from '@material-ui/core/Checkbox';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import CheckBoxIcon from '@material-ui/icons/CheckBox';
import CheckBoxOutlineBlankIcon from '@material-ui/icons/CheckBoxOutlineBlank';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import Autocomplete from '@material-ui/lab/Autocomplete';
import { useEntityTypeFilter } from '@backstage/plugin-catalog-react';
import { useApi, alertApiRef } from '@backstage/core-plugin-api';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { scaffolderTranslationRef } from '../../translation.esm.js';

const icon = /* @__PURE__ */ jsx(CheckBoxOutlineBlankIcon, { fontSize: "small" });
const checkedIcon = /* @__PURE__ */ jsx(CheckBoxIcon, { fontSize: "small" });
const TemplateTypePicker = () => {
  const alertApi = useApi(alertApiRef);
  const { error, loading, availableTypes, selectedTypes, setSelectedTypes } = useEntityTypeFilter();
  const { t } = useTranslationRef(scaffolderTranslationRef);
  if (loading) return /* @__PURE__ */ jsx(Progress, {});
  if (!availableTypes) return null;
  if (error) {
    alertApi.post({
      message: `Failed to load entity types`,
      severity: "error"
    });
    return null;
  }
  return /* @__PURE__ */ jsxs(Box, { pb: 1, pt: 1, children: [
    /* @__PURE__ */ jsx(
      Typography,
      {
        variant: "button",
        component: "label",
        htmlFor: "categories-picker",
        children: t("templateTypePicker.title")
      }
    ),
    /* @__PURE__ */ jsx(
      Autocomplete,
      {
        id: "categories-picker",
        multiple: true,
        options: availableTypes,
        value: selectedTypes,
        onChange: (_, value) => setSelectedTypes(value),
        renderOption: (option, { selected }) => /* @__PURE__ */ jsx(
          FormControlLabel,
          {
            control: /* @__PURE__ */ jsx(
              Checkbox,
              {
                icon,
                checkedIcon,
                checked: selected
              }
            ),
            label: capitalize(option)
          }
        ),
        size: "small",
        popupIcon: /* @__PURE__ */ jsx(ExpandMoreIcon, { "data-testid": "categories-picker-expand" }),
        renderInput: (params) => /* @__PURE__ */ jsx(TextField, { ...params, variant: "outlined" })
      }
    )
  ] });
};

export { TemplateTypePicker };
