// @backstage/plugin-catalog-react/dist/alpha/blueprints/EntityContextMenuItemBlueprint.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { createExtensionBlueprint, coreExtensionData, ExtensionBoundary } from '@backstage/frontend-plugin-api';
import MenuItem from '@material-ui/core/MenuItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import { useEntityContextMenu } from '../../hooks/useEntityContextMenu.esm.js';
import { entityPredicateToFilterFunction } from '../predicates/entityPredicateToFilterFunction.esm.js';
import { entityFilterFunctionDataRef } from './extensionData.esm.js';
import { createEntityPredicateSchema } from '../predicates/createEntityPredicateSchema.esm.js';

const EntityContextMenuItemBlueprint = createExtensionBlueprint({
  kind: "entity-context-menu-item",
  attachTo: { id: "page:catalog/entity", input: "contextMenuItems" },
  output: [
    coreExtensionData.reactElement,
    entityFilterFunctionDataRef.optional()
  ],
  dataRefs: {
    filterFunction: entityFilterFunctionDataRef
  },
  config: {
    schema: {
      filter: (z) => createEntityPredicateSchema(z).optional()
    }
  },
  *factory(params, { node, config }) {
    const loader = async () => {
      const Component = () => {
        const { onMenuClose } = useEntityContextMenu();
        const { title, ...menuItemProps } = params.useProps();
        let handleClick = void 0;
        if ("onClick" in menuItemProps) {
          handleClick = () => {
            const result = menuItemProps.onClick();
            if (result && "then" in result) {
              result.then(onMenuClose, onMenuClose);
            } else {
              onMenuClose();
            }
          };
        }
        return /* @__PURE__ */ jsxs(MenuItem, { ...menuItemProps, onClick: handleClick, children: [
          /* @__PURE__ */ jsx(ListItemIcon, { children: params.icon }),
          /* @__PURE__ */ jsx(ListItemText, { primary: title })
        ] });
      };
      return /* @__PURE__ */ jsx(Component, {});
    };
    yield coreExtensionData.reactElement(ExtensionBoundary.lazy(node, loader));
    if (config.filter) {
      yield entityFilterFunctionDataRef(
        entityPredicateToFilterFunction(config.filter)
      );
    } else if (typeof params.filter === "function") {
      yield entityFilterFunctionDataRef(params.filter);
    } else if (params.filter) {
      yield entityFilterFunctionDataRef(
        entityPredicateToFilterFunction(params.filter)
      );
    }
  }
});

export { EntityContextMenuItemBlueprint };
