// @backstage/plugin-catalog-react/dist/alpha/blueprints/extensionData.esm.js
import { createExtensionDataRef } from '@backstage/frontend-plugin-api';

const entityContentTitleDataRef = createExtensionDataRef().with({
  id: "catalog.entity-content-title"
});
const entityFilterFunctionDataRef = createExtensionDataRef().with({ id: "catalog.entity-filter-function" });
const entityFilterExpressionDataRef = createExtensionDataRef().with({
  id: "catalog.entity-filter-expression"
});
const defaultEntityContentGroups = {
  overview: "Overview",
  documentation: "Documentation",
  development: "Development",
  deployment: "Deployment",
  operation: "Operation",
  observability: "Observability"
};
const entityContentGroupDataRef = createExtensionDataRef().with({
  id: "catalog.entity-content-group"
});
const entityCardTypes = [
  "summary",
  "info",
  "content"
];
const entityCardTypeDataRef = createExtensionDataRef().with({
  id: "catalog.entity-card-type"
});

export { defaultEntityContentGroups, entityCardTypeDataRef, entityCardTypes, entityContentGroupDataRef, entityContentTitleDataRef, entityFilterExpressionDataRef, entityFilterFunctionDataRef };
