// @backstage/plugin-org/dist/components/Cards/Group/MembersList/MembersListCard.esm.js
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { stringifyEntityRef, DEFAULT_NAMESPACE } from '@backstage/catalog-model';
import { useEntity, catalogApiRef, EntityRefLink } from '@backstage/plugin-catalog-react';
import Box from '@material-ui/core/Box';
import Grid from '@material-ui/core/Grid';
import Switch from '@material-ui/core/Switch';
import Typography from '@material-ui/core/Typography';
import { makeStyles, createStyles } from '@material-ui/core/styles';
import Pagination from '@material-ui/lab/Pagination';
import { useState } from 'react';
import useAsync from 'react-use/esm/useAsync';
import { Progress, ResponseErrorPanel, InfoCard, Avatar, Link, OverflowTooltip } from '@backstage/core-components';
import { useApi } from '@backstage/core-plugin-api';
import { getAllDesendantMembersForGroupEntity, removeDuplicateEntitiesFrom } from '../../../../helpers/helpers.esm.js';
import { useTranslationRef } from '@backstage/frontend-plugin-api';
import { orgTranslationRef } from '../../../../translation.esm.js';

const useStyles = makeStyles(
  (theme) => createStyles({
    card: {
      border: `1px solid ${theme.palette.divider}`,
      boxShadow: theme.shadows[2],
      borderRadius: "4px",
      overflow: "visible",
      position: "relative",
      margin: theme.spacing(4, 1, 1),
      flex: "1",
      minWidth: "0px"
    },
    avatar: {
      position: "absolute",
      top: "-2rem"
    }
  }),
  { name: "PluginOrgMemberComponent" }
);
const MemberComponent = (props) => {
  const classes = useStyles();
  const {
    metadata: { name: metaName, description },
    spec: { profile }
  } = props.member;
  const displayName = profile?.displayName ?? metaName;
  return /* @__PURE__ */ jsx(Box, { className: classes.card, children: /* @__PURE__ */ jsxs(
    Box,
    {
      display: "flex",
      flexDirection: "column",
      m: 3,
      alignItems: "center",
      justifyContent: "center",
      children: [
        /* @__PURE__ */ jsx(
          Avatar,
          {
            displayName,
            picture: profile?.picture,
            classes
          }
        ),
        /* @__PURE__ */ jsxs(
          Box,
          {
            pt: 2,
            sx: {
              width: "100%"
            },
            textAlign: "center",
            children: [
              /* @__PURE__ */ jsx(Typography, { variant: "h6", children: /* @__PURE__ */ jsx(
                EntityRefLink,
                {
                  "data-testid": "user-link",
                  entityRef: props.member,
                  title: displayName
                }
              ) }),
              profile?.email && /* @__PURE__ */ jsx(Link, { to: `mailto:${profile.email}`, children: /* @__PURE__ */ jsx(OverflowTooltip, { text: profile.email }) }),
              description && /* @__PURE__ */ jsx(Typography, { variant: "subtitle2", children: /* @__PURE__ */ jsx(OverflowTooltip, { text: description, line: 5 }) })
            ]
          }
        )
      ]
    }
  ) });
};
const useListStyles = makeStyles(
  (theme) => ({
    root: {
      height: "100%"
    },
    cardContent: {
      overflow: "auto"
    },
    memberList: {
      display: "grid",
      gap: theme.spacing(1.5),
      gridTemplateColumns: `repeat(auto-fit, minmax(auto, ${theme.spacing(
        34
      )}px))`
    }
  }),
  { name: "PluginOrgMembersListCardComponent" }
);
const MembersListCard = (props) => {
  const { t } = useTranslationRef(orgTranslationRef);
  const {
    memberDisplayTitle = t("membersListCard.title"),
    pageSize = 50,
    showAggregateMembersToggle,
    relationType = "memberof"
  } = props;
  const relationAggregation = props.relationAggregation ?? props.relationsType ?? "direct";
  const classes = useListStyles();
  const { entity: groupEntity } = useEntity();
  const {
    metadata: { name: groupName, namespace: grpNamespace },
    spec: { profile }
  } = groupEntity;
  const catalogApi = useApi(catalogApiRef);
  const displayName = profile?.displayName ?? groupName;
  const groupNamespace = grpNamespace || DEFAULT_NAMESPACE;
  const [page, setPage] = useState(1);
  const pageChange = (_, pageIndex) => {
    setPage(pageIndex);
  };
  const [showAggregateMembers, setShowAggregateMembers] = useState(
    relationAggregation === "aggregated"
  );
  const { loading: loadingDescendantMembers, value: descendantMembers } = useAsync(async () => {
    if (!showAggregateMembers) {
      return [];
    }
    return await getAllDesendantMembersForGroupEntity(
      groupEntity,
      catalogApi,
      relationType
    );
  }, [catalogApi, groupEntity, showAggregateMembers]);
  const {
    loading,
    error,
    value: directMembers
  } = useAsync(async () => {
    const membersList = await catalogApi.getEntities({
      filter: {
        kind: "User",
        [`relations.${relationType.toLocaleLowerCase("en-US")}`]: [
          stringifyEntityRef({
            kind: "group",
            namespace: groupNamespace.toLocaleLowerCase("en-US"),
            name: groupName.toLocaleLowerCase("en-US")
          })
        ]
      }
    });
    return membersList.items;
  }, [catalogApi, groupEntity]);
  const members = removeDuplicateEntitiesFrom(
    [
      ...directMembers ?? [],
      ...descendantMembers && showAggregateMembers ? descendantMembers : []
    ].sort(
      (a, b) => stringifyEntityRef(a).localeCompare(stringifyEntityRef(b))
    )
  );
  if (loading) {
    return /* @__PURE__ */ jsx(Progress, {});
  } else if (error) {
    return /* @__PURE__ */ jsx(ResponseErrorPanel, { error });
  }
  const nbPages = Math.ceil((members?.length || 0) / pageSize);
  const paginationLabel = nbPages < 2 ? "" : t("membersListCard.paginationLabel", {
    page: String(page),
    nbPages: String(nbPages)
  });
  const pagination = /* @__PURE__ */ jsx(
    Pagination,
    {
      count: nbPages,
      page,
      onChange: pageChange,
      showFirstButton: true,
      showLastButton: true
    }
  );
  let memberList;
  if (members && members.length > 0) {
    memberList = /* @__PURE__ */ jsx(Box, { className: classes.memberList, children: members.slice(pageSize * (page - 1), pageSize * page).map((member) => /* @__PURE__ */ jsx(MemberComponent, { member }, stringifyEntityRef(member))) });
  } else {
    memberList = /* @__PURE__ */ jsx(Box, { p: 2, children: /* @__PURE__ */ jsx(Typography, { children: t("membersListCard.noMembersDescription") }) });
  }
  return /* @__PURE__ */ jsx(Grid, { item: true, className: classes.root, children: /* @__PURE__ */ jsxs(
    InfoCard,
    {
      title: `${memberDisplayTitle} (${members?.length || 0}${paginationLabel})`,
      subheader: t("membersListCard.subtitle", {
        groupName: displayName
      }),
      ...nbPages <= 1 ? {} : { actions: pagination },
      className: classes.root,
      cardClassName: classes.cardContent,
      children: [
        showAggregateMembersToggle && /* @__PURE__ */ jsxs(Fragment, { children: [
          t("membersListCard.aggregateMembersToggle.directMembers"),
          /* @__PURE__ */ jsx(
            Switch,
            {
              color: "primary",
              checked: showAggregateMembers,
              onChange: () => {
                setShowAggregateMembers(!showAggregateMembers);
              },
              inputProps: {
                "aria-label": t(
                  "membersListCard.aggregateMembersToggle.ariaLabel"
                )
              }
            }
          ),
          t("membersListCard.aggregateMembersToggle.aggregatedMembers")
        ] }),
        showAggregateMembers && loadingDescendantMembers ? /* @__PURE__ */ jsx(Progress, {}) : memberList
      ]
    }
  ) });
};

export { MembersListCard };
