// @backstage/plugin-scaffolder-backend/dist/ScaffolderPlugin.cjs.js
'use strict';

var backendPluginApi = require('@backstage/backend-plugin-api');
var integration = require('@backstage/integration');
var pluginCatalogNode = require('@backstage/plugin-catalog-node');
var pluginEventsNode = require('@backstage/plugin-events-node');
var alpha = require('@backstage/plugin-scaffolder-node/alpha');
var register = require('./scaffolder/actions/builtin/catalog/register.cjs.js');
var write = require('./scaffolder/actions/builtin/catalog/write.cjs.js');
var fetch = require('./scaffolder/actions/builtin/catalog/fetch.cjs.js');
var log = require('./scaffolder/actions/builtin/debug/log.cjs.js');
var wait = require('./scaffolder/actions/builtin/debug/wait.cjs.js');
var plain = require('./scaffolder/actions/builtin/fetch/plain.cjs.js');
var plainFile = require('./scaffolder/actions/builtin/fetch/plainFile.cjs.js');
var template = require('./scaffolder/actions/builtin/fetch/template.cjs.js');
var templateFile = require('./scaffolder/actions/builtin/fetch/templateFile.cjs.js');
var _delete = require('./scaffolder/actions/builtin/filesystem/delete.cjs.js');
var rename = require('./scaffolder/actions/builtin/filesystem/rename.cjs.js');
var read = require('./scaffolder/actions/builtin/filesystem/read.cjs.js');
require('@backstage/errors');
require('./scaffolder/tasks/DatabaseTaskStore.cjs.js');
require('@backstage/types');
require('zen-observable');
require('fs-extra');
require('@backstage/config');
require('lodash');
require('p-queue');
require('./scaffolder/tasks/NunjucksWorkflowRunner.cjs.js');
require('timers/promises');
var router = require('./service/router.cjs.js');
var loggerToWinstonLogger = require('./util/loggerToWinstonLogger.cjs.js');
var templating = require('./util/templating.cjs.js');
var alpha$1 = require('@backstage/backend-plugin-api/alpha');

const scaffolderPlugin = backendPluginApi.createBackendPlugin({
  pluginId: "scaffolder",
  register(env) {
    const addedActions = new Array();
    env.registerExtensionPoint(alpha.scaffolderActionsExtensionPoint, {
      addActions(...newActions) {
        addedActions.push(...newActions);
      }
    });
    let taskBroker;
    env.registerExtensionPoint(alpha.scaffolderTaskBrokerExtensionPoint, {
      setTaskBroker(newTaskBroker) {
        if (taskBroker) {
          throw new Error("Task broker may only be set once");
        }
        taskBroker = newTaskBroker;
      }
    });
    const additionalTemplateFilters = [];
    const additionalTemplateGlobals = [];
    env.registerExtensionPoint(alpha.scaffolderTemplatingExtensionPoint, {
      addTemplateFilters(newFilters) {
        additionalTemplateFilters.push(
          ...Array.isArray(newFilters) ? newFilters : Object.entries(newFilters).map(
            ([id, filter]) => alpha.createTemplateFilter({
              id,
              filter
            })
          )
        );
      },
      addTemplateGlobals(newGlobals) {
        additionalTemplateGlobals.push(
          ...Array.isArray(newGlobals) ? newGlobals : Object.entries(newGlobals).map(
            ([id, global]) => typeof global === "function" ? alpha.createTemplateGlobalFunction({ id, fn: global }) : alpha.createTemplateGlobalValue({ id, value: global })
          )
        );
      }
    });
    const autocompleteHandlers = {};
    env.registerExtensionPoint(alpha.scaffolderAutocompleteExtensionPoint, {
      addAutocompleteProvider(provider) {
        autocompleteHandlers[provider.id] = provider.handler;
      }
    });
    const additionalWorkspaceProviders = {};
    env.registerExtensionPoint(alpha.scaffolderWorkspaceProviderExtensionPoint, {
      addProviders(provider) {
        Object.assign(additionalWorkspaceProviders, provider);
      }
    });
    env.registerInit({
      deps: {
        logger: backendPluginApi.coreServices.logger,
        config: backendPluginApi.coreServices.rootConfig,
        lifecycle: backendPluginApi.coreServices.rootLifecycle,
        reader: backendPluginApi.coreServices.urlReader,
        permissions: backendPluginApi.coreServices.permissions,
        database: backendPluginApi.coreServices.database,
        auth: backendPluginApi.coreServices.auth,
        httpRouter: backendPluginApi.coreServices.httpRouter,
        httpAuth: backendPluginApi.coreServices.httpAuth,
        auditor: backendPluginApi.coreServices.auditor,
        catalog: pluginCatalogNode.catalogServiceRef,
        events: pluginEventsNode.eventsServiceRef,
        actionsRegistry: alpha$1.actionsServiceRef
      },
      async init({
        logger,
        config,
        lifecycle,
        reader,
        database,
        auth,
        httpRouter,
        httpAuth,
        catalog,
        permissions,
        events,
        auditor,
        actionsRegistry
      }) {
        const log$1 = loggerToWinstonLogger.loggerToWinstonLogger(logger);
        const integrations = integration.ScmIntegrations.fromConfig(config);
        const templateExtensions = {
          additionalTemplateFilters: templating.convertFiltersToRecord(
            additionalTemplateFilters
          ),
          additionalTemplateGlobals: templating.convertGlobalsToRecord(
            additionalTemplateGlobals
          )
        };
        const actions = [
          // actions provided from other modules
          ...addedActions,
          // built-in actions for the scaffolder
          plain.createFetchPlainAction({
            reader,
            integrations
          }),
          plainFile.createFetchPlainFileAction({
            reader,
            integrations
          }),
          template.createFetchTemplateAction({
            integrations,
            reader,
            ...templateExtensions
          }),
          templateFile.createFetchTemplateFileAction({
            integrations,
            reader,
            ...templateExtensions
          }),
          log.createDebugLogAction(),
          wait.createWaitAction(),
          // todo(blam): maybe these should be a -catalog module?
          register.createCatalogRegisterAction({ catalog, integrations }),
          fetch.createFetchCatalogEntityAction({ catalog }),
          write.createCatalogWriteAction(),
          _delete.createFilesystemDeleteAction(),
          rename.createFilesystemRenameAction(),
          read.createFilesystemReadDirAction()
        ];
        const actionIds = actions.map((action) => action.id).join(", ");
        log$1.info(
          `Starting scaffolder with the following actions enabled ${actionIds}`
        );
        const router$1 = await router.createRouter({
          logger,
          config,
          database,
          catalog,
          lifecycle,
          actions,
          taskBroker,
          additionalTemplateFilters,
          additionalTemplateGlobals,
          auth,
          httpAuth,
          permissions,
          autocompleteHandlers,
          additionalWorkspaceProviders,
          events,
          auditor,
          actionsRegistry
        });
        httpRouter.use(router$1);
      }
    });
  }
});

exports.scaffolderPlugin = scaffolderPlugin;
