// @backstage/plugin-proxy-backend/dist/plugin.cjs.js
'use strict';

var backendPluginApi = require('@backstage/backend-plugin-api');
var router = require('./service/router.cjs.js');
var alpha = require('@backstage/plugin-proxy-node/alpha');

const proxyPlugin = backendPluginApi.createBackendPlugin({
  pluginId: "proxy",
  register(env) {
    const additionalEndpoints = {};
    env.registerExtensionPoint(alpha.proxyEndpointsExtensionPoint, {
      addProxyEndpoints(endpoints) {
        Object.assign(additionalEndpoints, endpoints);
      }
    });
    env.registerInit({
      deps: {
        config: backendPluginApi.coreServices.rootConfig,
        discovery: backendPluginApi.coreServices.discovery,
        logger: backendPluginApi.coreServices.logger,
        httpRouter: backendPluginApi.coreServices.httpRouter
      },
      async init({ config, discovery, logger, httpRouter }) {
        await router.createRouter({
          config,
          logger,
          httpRouterService: httpRouter,
          additionalEndpoints
        });
      }
    });
  }
});

exports.proxyPlugin = proxyPlugin;
