// @module-federation/enhanced/dist/src/wrapper/HoistContainerReferencesPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const normalize_webpack_path_1 = require("@module-federation/sdk/normalize-webpack-path");
const PLUGIN_NAME = 'HoistContainerReferencesPlugin';
class HoistContainerReferencesPlugin {
    constructor(name, entryFilePath, bundlerRuntimeDep) {
        this.containerName = name || 'no known chunk name';
        this.entryFilePath = entryFilePath;
        this.bundlerRuntimeDep = bundlerRuntimeDep;
        this.explanation =
            'Bundler runtime path module is required for proper functioning';
        this.name = PLUGIN_NAME;
    }
    apply(compiler) {
        process.env['FEDERATION_WEBPACK_PATH'] =
            process.env['FEDERATION_WEBPACK_PATH'] || (0, normalize_webpack_path_1.getWebpackPath)(compiler);
        const CoreHoistContainerReferencesPlugin = require('../lib/container/HoistContainerReferencesPlugin')
            .default;
        new CoreHoistContainerReferencesPlugin(this.containerName, this.entryFilePath, this.bundlerRuntimeDep).apply(compiler);
    }
}
exports.default = HoistContainerReferencesPlugin;
