// @backstage/plugin-scaffolder-react/dist/next/components/TemplateCard/CardHeader.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { makeStyles, useTheme } from '@material-ui/core/styles';
import { ItemCardHeader } from '@backstage/core-components';
import { FavoriteEntity } from '@backstage/plugin-catalog-react';
import { TemplateDetailButton } from './TemplateDetailButton.esm.js';

const useStyles = makeStyles(() => ({
  header: {
    backgroundImage: ({ cardBackgroundImage }) => cardBackgroundImage,
    color: ({ cardFontColor }) => cardFontColor
  },
  subtitleWrapper: {
    display: "flex",
    justifyContent: "space-between"
  },
  detailIcon: {
    color: ({ cardFontColor }) => cardFontColor
  }
}));
const CardHeader = (props) => {
  const {
    template: {
      metadata: { title, name },
      spec: { type }
    }
  } = props;
  const { getPageTheme } = useTheme();
  const themeForType = getPageTheme({ themeId: type });
  const styles = useStyles({
    cardFontColor: themeForType.fontColor,
    cardBackgroundImage: themeForType.backgroundImage
  });
  const SubtitleComponent = /* @__PURE__ */ jsxs("div", { className: styles.subtitleWrapper, children: [
    /* @__PURE__ */ jsx("div", { children: type }),
    /* @__PURE__ */ jsxs("div", { children: [
      /* @__PURE__ */ jsx(
        TemplateDetailButton,
        {
          className: styles.detailIcon,
          template: props.template
        }
      ),
      /* @__PURE__ */ jsx(
        FavoriteEntity,
        {
          entity: props.template,
          style: { padding: 0, marginLeft: 6 }
        }
      )
    ] })
  ] });
  return /* @__PURE__ */ jsx(
    ItemCardHeader,
    {
      title: title ?? name,
      subtitle: SubtitleComponent,
      classes: { root: styles.header }
    }
  );
};

export { CardHeader };
