// @backstage/plugin-scaffolder-react/dist/next/components/TemplateCard/TemplateCardLinks.esm.js
import { jsxs, Fragment, jsx } from 'react/jsx-runtime';
import { useApp } from '@backstage/core-plugin-api';
import Divider from '@material-ui/core/Divider';
import Grid from '@material-ui/core/Grid';
import { makeStyles } from '@material-ui/core/styles';
import LanguageIcon from '@material-ui/icons/Language';
import { CardLink } from './CardLink.esm.js';

const useStyles = makeStyles({});
const TemplateCardLinks = ({
  template,
  additionalLinks
}) => {
  const styles = useStyles();
  const app = useApp();
  const iconResolver = (key) => key ? app.getSystemIcon(key) ?? LanguageIcon : LanguageIcon;
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(Grid, { item: true, xs: 12, children: /* @__PURE__ */ jsx(Divider, { "data-testid": "template-card-separator--links" }) }),
    /* @__PURE__ */ jsx(Grid, { item: true, xs: 12, children: /* @__PURE__ */ jsxs(Grid, { container: true, spacing: 2, "data-testid": "template-card-links", children: [
      additionalLinks?.map(({ icon, text, url }, index) => /* @__PURE__ */ jsx(
        Grid,
        {
          className: styles.linkText,
          item: true,
          xs: 6,
          "data-testid": "template-card-links--item",
          children: /* @__PURE__ */ jsx(CardLink, { icon, text, url })
        },
        index
      )),
      template.metadata.links?.map(({ url, icon, title }, index) => /* @__PURE__ */ jsx(
        Grid,
        {
          className: styles.linkText,
          item: true,
          xs: 6,
          "data-testid": "template-card-links--metalink",
          children: /* @__PURE__ */ jsx(
            CardLink,
            {
              icon: iconResolver(icon),
              text: title || url,
              url
            }
          )
        },
        index
      ))
    ] }) })
  ] });
};

export { TemplateCardLinks };
