// @backstage/plugin-scaffolder-react/dist/next/components/TemplateCard/TemplateCardContent.esm.js
import { jsx } from 'react/jsx-runtime';
import { MarkdownContent } from '@backstage/core-components';
import { useTranslationRef } from '@backstage/frontend-plugin-api';
import Box from '@material-ui/core/Box';
import Grid from '@material-ui/core/Grid';
import { makeStyles } from '@material-ui/core/styles';
import { scaffolderReactTranslationRef } from '../../../translation.esm.js';

const useStyles = makeStyles(() => ({
  box: {
    overflow: "hidden",
    textOverflow: "ellipsis",
    display: "-webkit-box",
    "-webkit-line-clamp": 10,
    "-webkit-box-orient": "vertical"
  },
  markdown: {
    /** to make the styles for React Markdown not leak into the description */
    "& :first-child": {
      margin: 0
    }
  }
}));
const TemplateCardContent = ({ template }) => {
  const { t } = useTranslationRef(scaffolderReactTranslationRef);
  const styles = useStyles();
  return /* @__PURE__ */ jsx(Grid, { item: true, xs: 12, "data-testid": "template-card-content-grid", children: /* @__PURE__ */ jsx(Box, { className: styles.box, "data-testid": "template-card-content-container", children: /* @__PURE__ */ jsx(
    MarkdownContent,
    {
      className: styles.markdown,
      content: template.metadata.description ?? t("templateCard.noDescription")
    }
  ) }) });
};

export { TemplateCardContent };
